diff --git a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
index d9459229807..0d7bfd24dde 100644
--- a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
+++ b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
@@ -13,4 +13,4 @@
 
 
 
-<h1>Issue 3</h1>
+<h1><span style="text-decoration: line-through;">Issue 3</span> Patch MC-3-V2 (Version 2)</h1>
