diff --git a/vendor/magento/module-versions-cms/view/adminhtml/templates/page/tab/hierarchy.phtml b/vendor/magento/module-versions-cms/view/adminhtml/templates/page/tab/hierarchy.phtml
index 8b5aa1e0af5..ab2a3d8ee3b 100644
--- a/vendor/magento/module-versions-cms/view/adminhtml/templates/page/tab/hierarchy.phtml
+++ b/vendor/magento/module-versions-cms/view/adminhtml/templates/page/tab/hierarchy.phtml
@@ -4,8 +4,11 @@
  * See COPYING.txt for license details.
  */
 
-/* @var \Magento\VersionsCms\Block\Adminhtml\Cms\Page\Edit\Tab\Hierarchy $block */
-/** @var \Magento\Framework\View\Helper\SecureHtmlRenderer $secureRenderer */
+/**
+ * @var \Magento\VersionsCms\Block\Adminhtml\Cms\Page\Edit\Tab\Hierarchy $block
+ * @var \Magento\Framework\View\Helper\SecureHtmlRenderer $secureRenderer
+ * @var \Magento\Framework\Escaper $escaper
+ */
 
 $jsTreeExplore = $block->getCMSPageCategory();
 
@@ -15,17 +18,17 @@ $jsonNodesData = $block->getPrefilledData();
 
 ?>
 <label class="admin__field-label" for="tree-container">
-    <span><?= $block->escapeHtml(__('Parent page')) ?></span>
+    <span><?= $escaper->escapeHtml(__('Parent page')) ?></span>
 </label>
 
 
 <div class="admin__field-control">
     <input type="hidden" name="node_ids" id="node_ids" data-form-part="cms_page_form"
-           value="<?= $block->escapeHtmlAttr($block->getSelectedNodeIds()) ?>"/>
+           value="<?= $escaper->escapeHtmlAttr($block->getSelectedNodeIds()) ?>"/>
     <input type="hidden" name="website_root" id="website_root" data-form-part="cms_page_form"
-           value="<?= $block->escapeHtmlAttr($block->getPage()->getWebsiteRoot()) ?>"/>
+           value="<?= $escaper->escapeHtmlAttr($block->getPage()->getWebsiteRoot()) ?>"/>
     <input type="hidden" name="nodes_data" id="nodes_data"
-           value="<?= $block->escapeHtml($jsonNodesData) ?>" data-form-part="cms_page_form"/>
+           value="<?= $escaper->escapeHtmlAttr($jsonNodesData) ?>" data-form-part="cms_page_form"/>
     <input type="hidden" name="assign_to_root" id="assign_to_root" data-form-part="cms_page_form" value="">
     <div id="tree-container" class="cms-tree tree x-tree"></div>
     <div id="tree-div" class="tree-wrapper cms-tree tree x-tree">
@@ -41,7 +44,7 @@ $jsonNodesData = $block->getPrefilledData();
 $nodesJson = $block->getNodesJson();
 $currentPageJson = $block->getCurrentPageJson();
 $isAssignedToWebsiteRoot = $block->getPage()->getWebsiteRoot() ? 1 : 0;
-$websiteRootMessage = $block->escapeJs(__('Website Root'));
+$websiteRootMessage = $escaper->escapeJs(__('Website Root'));
 $scriptString = <<<script
     require(['jquery',
              'prototype',
