diff --git a/vendor/magento/module-sales-archive/Plugin/ArchivedEntitiesProcessorPlugin.php b/vendor/magento/module-sales-archive/Plugin/ArchivedEntitiesProcessorPlugin.php
index cca569ca8a1c..e18f81d0596c 100644
--- a/vendor/magento/module-sales-archive/Plugin/ArchivedEntitiesProcessorPlugin.php
+++ b/vendor/magento/module-sales-archive/Plugin/ArchivedEntitiesProcessorPlugin.php
@@ -47,12 +47,16 @@ public function __construct(
      * @param IdListBuilder $idListBuilder
      * @param string $mainTableName
      * @param string $gridTableName
+     * @param int|null $startEntityId
+     * @param int|null $endEntityId
      * @return array
      */
     public function beforeBuild(
         IdListBuilder $idListBuilder,
         string $mainTableName,
-        string $gridTableName
+        string $gridTableName,
+        ?int $startEntityId = null,
+        ?int $endEntityId = null
     ) : array {
         if ($this->config->isArchiveActive()) {
             $idListBuilder->resetAdditionalGridTable();
@@ -62,6 +66,6 @@ public function beforeBuild(
             }
         }
 
-        return [$mainTableName, $gridTableName];
+        return [$mainTableName, $gridTableName, $startEntityId, $endEntityId];
     }
 }
