diff --git a/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php b/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
index b6e67ec6ed6..803781722f5 100644
--- a/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
+++ b/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
@@ -1,7 +1,18 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * ADOBE CONFIDENTIAL
+ *
+ * Copyright 2020 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
  */
 declare(strict_types=1);
 
@@ -77,15 +88,15 @@ public function execute(int $customerId, array $args): array
         $currentPage = $args['currentPage'] ?? 1;
         $pageSize = $args['pageSize'] ?? 20;
 
-        $builderCount = $this->getFilters($customerId, true, $args);
+        $builderCount = $this->getFilters($customerId, $args);
         $countLists = $this->repository->getList($builderCount->create())->getTotalCount();
 
-        $builder = $this->getFilters($customerId, false, $args);
+        $builder = $this->getFilters($customerId, $args);
         $builder->setPageSize($pageSize)->setCurrentPage($currentPage);
         $lists = $this->repository->getList($builder->create())->getItems();
 
         $data = [
-            'total_count' => count($lists),
+            'total_count' => $countLists,
             'items' => [],
             'page_info' => [
                 'page_size' => $pageSize,
@@ -123,53 +134,48 @@ public function execute(int $customerId, array $args): array
      * Get Filters
      *
      * @param int $customerId
-     * @param bool $isTotal
      * @param array $args
      * @return SearchCriteriaBuilder
      *
      * @SuppressWarnings(PHPMD.CyclomaticComplexity)
      */
-    private function getFilters(int $customerId, bool $isTotal, array $args = [])
+    private function getFilters(int $customerId, array $args = [])
     {
         $filtersGiven = isset($args['filter']) ? $args['filter'] : [];
 
         $filters = [];
 
-        if (!$isTotal) {
-            if (isset($filtersGiven['uids'])) {
-                if (key($filtersGiven['uids']) == 'eq' || key($filtersGiven['uids']) == 'in') {
-                    $reqIds = [];
-                    if (is_array($filtersGiven['uids'][key($filtersGiven['uids'])])) {
-                        foreach ($filtersGiven['uids'][key($filtersGiven['uids'])] as $id) {
-                            $reqIds[] = $this->idEncoder->decode($id);
-                        }
-                    } else {
-                        $reqIds = $this->idEncoder->decode($filtersGiven['uids'][key($filtersGiven['uids'])]);
+        if (isset($filtersGiven['uids'])) {
+            if (key($filtersGiven['uids']) == 'eq' || key($filtersGiven['uids']) == 'in') {
+                $reqIds = [];
+                if (is_array($filtersGiven['uids'][key($filtersGiven['uids'])])) {
+                    foreach ($filtersGiven['uids'][key($filtersGiven['uids'])] as $id) {
+                        $reqIds[] = $this->idEncoder->decode($id);
                     }
-
-                    $filters[] = $this->filterBuilder
-                        ->setField(RequisitionListInterface::REQUISITION_LIST_ID)
-                        ->setConditionType(key($filtersGiven['uids']))
-                        ->setValue($reqIds)
-                        ->create();
+                } else {
+                    $reqIds = $this->idEncoder->decode($filtersGiven['uids'][key($filtersGiven['uids'])]);
                 }
+
+                $filters[] = $this->filterBuilder
+                    ->setField(RequisitionListInterface::REQUISITION_LIST_ID)
+                    ->setConditionType(key($filtersGiven['uids']))
+                    ->setValue($reqIds)
+                    ->create();
             }
+        }
 
-            if (isset($filtersGiven['name'])) {
-                if (key($filtersGiven['name']) == 'match') {
-                    $filters[] = $this->filterBuilder
-                        ->setField(RequisitionListInterface::NAME)
-                        ->setConditionType('like')
-                        ->setValue($filtersGiven['name'][key($filtersGiven['name'])])
-                        ->create();
-                }
+        if (isset($filtersGiven['name'])) {
+            if (key($filtersGiven['name']) == 'match') {
+                $filters[] = $this->filterBuilder
+                    ->setField(RequisitionListInterface::NAME)
+                    ->setConditionType('like')
+                    ->setValue($filtersGiven['name'][key($filtersGiven['name'])])
+                    ->create();
             }
-            return $this->searchCriteriaBuilder
-                ->addFilter('customer_id', $customerId)
-                ->addFilters($filters);
-        } else {
-            return $this->searchCriteriaBuilder
-                ->addFilter('customer_id', $customerId);
         }
+
+        return $this->searchCriteriaBuilder
+            ->addFilter('customer_id', $customerId)
+            ->addFilters($filters);
     }
 }
