diff --git a/vendor/magento/module-quick-order/Plugin/AdvancedCheckout/AddExtraDataToCartItemForQuickOrderPlugin.php b/vendor/magento/module-quick-order/Plugin/AdvancedCheckout/AddExtraDataToCartItemForQuickOrderPlugin.php
index 4c0dd437e1..edda45c3e8 100644
--- a/vendor/magento/module-quick-order/Plugin/AdvancedCheckout/AddExtraDataToCartItemForQuickOrderPlugin.php
+++ b/vendor/magento/module-quick-order/Plugin/AdvancedCheckout/AddExtraDataToCartItemForQuickOrderPlugin.php
@@ -130,7 +130,7 @@ class AddExtraDataToCartItemForQuickOrderPlugin
             $item['qty'] = null;
         }
         $item['name'] = $product->getName();
-        $item['price'] = $this->taxHelper->getTaxPrice($product, $product->getFinalPrice($item['qty']), true);
+        $item['price'] = $this->taxHelper->getTaxPrice($product, $product->getFinalPrice($item['qty']));
         $item['thumbnail_url'] = $this->getProductThumbnailUrl($product);
         $item['url'] = $product->getProductUrl();
 
