diff --git a/vendor/magento/module-customer-balance/view/adminhtml/templates/sales/creditmemo/refund.phtml b/vendor/magento/module-customer-balance/view/adminhtml/templates/sales/creditmemo/refund.phtml
index ef8ac66338b..abd5f3d4e4e 100644
--- a/vendor/magento/module-customer-balance/view/adminhtml/templates/sales/creditmemo/refund.phtml
+++ b/vendor/magento/module-customer-balance/view/adminhtml/templates/sales/creditmemo/refund.phtml
@@ -6,6 +6,7 @@
 
 /**
  * @var \Magento\CustomerBalance\Block\Adminhtml\Sales\Order\Creditmemo\Controls $block
+ * @var \Magento\Framework\Escaper $escaper
  * @var \Magento\Framework\View\Helper\SecureHtmlRenderer $secureRenderer
  */
 $customerBalanceData = $block->getCustomerBalanceData();
@@ -26,7 +27,7 @@ $customerBalanceHelper = $block->getData('customerBalanceHelper');
                    checked="checked"
                 <?php endif; ?> />
         <label for="creditmemo_customerbalance_return_enable" class="admin__field-label">
-            <span><?= $block->escapeHtml(__('Refund to Store Credit')); ?></span>
+            <span><?= $escaper->escapeHtml(__('Refund to Store Credit')); ?></span>
         </label>
         <input type="text"
                name="creditmemo[refund_customerbalance_return]"
@@ -49,9 +50,11 @@ $customerBalanceHelper = $block->getData('customerBalanceHelper');
                         disableElements('refund');
                     } else {
                         refundReturn.prop('disabled', true);
-                        enableElements('refund');
                         if (creditMemoTotal === 0) {
+                            disableElements('refund');
                             enableElements('update-totals-button');
+                        } else {
+                            enableElements('refund');
                         }
                     }
                 };
