diff --git a/vendor/magento/module-reward/Model/Plugin/CustomerRegister.php b/vendor/magento/module-reward/Model/Plugin/CustomerRegister.php
index 0a0c0313092..2eb1c400035 100644
--- a/vendor/magento/module-reward/Model/Plugin/CustomerRegister.php
+++ b/vendor/magento/module-reward/Model/Plugin/CustomerRegister.php
@@ -5,57 +5,57 @@
  */
 namespace Magento\Reward\Model\Plugin;
 
+use Magento\Customer\Api\Data\CustomerInterface;
+use Magento\Customer\Model\AccountManagement;
+use Magento\Customer\Model\CustomerRegistry;
+use Magento\Reward\Helper\Data as RewardData;
+use Magento\Reward\Model\RewardFactory;
+use Magento\Store\Model\StoreManagerInterface;
+use Psr\Log\LoggerInterface;
+
 /**
  * Plugin for saving reward notification attributes on customer register
  */
 class CustomerRegister
 {
     /**
-     * @var \Psr\Log\LoggerInterface
+     * @var LoggerInterface
      */
     protected $_logger;
 
     /**
-     * Reward factory
-     *
-     * @var \Magento\Reward\Model\RewardFactory
+     * @var RewardFactory
      */
     protected $_rewardFactory;
 
     /**
-     * Core model store manager interface
-     *
-     * @var \Magento\Store\Model\StoreManagerInterface
+     * @var StoreManagerInterface
      */
     protected $_storeManager;
 
     /**
-     * Reward helper
-     *
-     * @var \Magento\Reward\Helper\Data
+     * @var RewardData
      */
     protected $_rewardData;
 
     /**
-     * Customer registry
-     *
-     * @var \Magento\Customer\Model\CustomerRegistry
+     * @var CustomerRegistry
      */
     private $customerRegistry;
 
     /**
-     * @param \Magento\Reward\Helper\Data $rewardData
-     * @param \Magento\Store\Model\StoreManagerInterface $storeManager
-     * @param \Magento\Reward\Model\RewardFactory $rewardFactory
-     * @param \Psr\Log\LoggerInterface $logger
-     * @param \Magento\Customer\Model\CustomerRegistry $customerRegistry
+     * @param RewardData $rewardData
+     * @param StoreManagerInterface $storeManager
+     * @param RewardFactory $rewardFactory
+     * @param LoggerInterface $logger
+     * @param CustomerRegistry $customerRegistry
      */
     public function __construct(
-        \Magento\Reward\Helper\Data $rewardData,
-        \Magento\Store\Model\StoreManagerInterface $storeManager,
-        \Magento\Reward\Model\RewardFactory $rewardFactory,
-        \Psr\Log\LoggerInterface $logger,
-        \Magento\Customer\Model\CustomerRegistry $customerRegistry
+        RewardData $rewardData,
+        StoreManagerInterface $storeManager,
+        RewardFactory $rewardFactory,
+        LoggerInterface $logger,
+        CustomerRegistry $customerRegistry
     ) {
         $this->_rewardData = $rewardData;
         $this->_storeManager = $storeManager;
@@ -67,15 +67,15 @@ class CustomerRegister
     /**
      * Save reward notification attributes and reward after customer account create
      *
-     * @param \Magento\Customer\Model\AccountManagement $subject
-     * @param \Magento\Customer\Api\Data\CustomerInterface $customer
+     * @param AccountManagement $subject
+     * @param CustomerInterface $customer
      *
-     * @return \Magento\Customer\Api\Data\CustomerInterface
+     * @return CustomerInterface
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function afterCreateAccountWithPasswordHash(
-        \Magento\Customer\Model\AccountManagement $subject,
-        \Magento\Customer\Api\Data\CustomerInterface $customer
+    public function afterCreateAccount(
+        AccountManagement $subject,
+        CustomerInterface $customer
     ) {
         if (!$this->_rewardData->isEnabledOnFront()) {
             return $customer;
diff --git a/vendor/magento/module-reward/etc/di.xml b/vendor/magento/module-reward/etc/di.xml
index c8731777c4c..d26229337f3 100644
--- a/vendor/magento/module-reward/etc/di.xml
+++ b/vendor/magento/module-reward/etc/di.xml
@@ -65,6 +65,9 @@
     <type name="Magento\Sales\Api\OrderRepositoryInterface">
         <plugin name="rewardCreditmemoUpdater" type="Magento\Reward\Model\Plugin\OrderRepository"/>
     </type>
+    <type name="Magento\Customer\Model\AccountManagement">
+        <plugin name="customerRegister" type="Magento\Reward\Model\Plugin\CustomerRegister"/>
+    </type>
     <type name="Magento\Framework\EntityManager\Operation\ExtensionPool">
         <arguments>
             <argument name="extensionActions" xsi:type="array">
diff --git a/vendor/magento/module-reward/etc/frontend/di.xml b/vendor/magento/module-reward/etc/frontend/di.xml
index bbf5417f932..f460296dff2 100644
--- a/vendor/magento/module-reward/etc/frontend/di.xml
+++ b/vendor/magento/module-reward/etc/frontend/di.xml
@@ -6,9 +6,6 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
-    <type name="Magento\Customer\Model\AccountManagement">
-        <plugin name="customerRegister" type="Magento\Reward\Model\Plugin\CustomerRegister"/>
-    </type>
     <type name="Magento\Checkout\Model\CompositeConfigProvider">
         <arguments>
             <argument name="configProviders" xsi:type="array">
