diff --git a/vendor/magento/module-admin-gws/Model/Models.php b/vendor/magento/module-admin-gws/Model/Models.php
index d39e459a0a4..903e15613ea 100644
--- a/vendor/magento/module-admin-gws/Model/Models.php
+++ b/vendor/magento/module-admin-gws/Model/Models.php
@@ -570,7 +570,7 @@ class Models extends \Magento\AdminGws\Model\Observer\AbstractObserver implement
      */
     public function customerLoadAfter($model)
     {
-        if (!$this->_role->hasStoreAccess($model->getStoreId())) {
+        if (!$this->_role->hasWebsiteAccess($model->getWebsiteId())) {
             $this->_throwLoad();
         }
     }
