diff --git a/vendor/magento/module-gift-wrapping/Model/Observer/ExtendOrderAttributes.php b/vendor/magento/module-gift-wrapping/Model/Observer/ExtendOrderAttributes.php
index 3b681d34fe8c..8bde0ddee15f 100644
--- a/vendor/magento/module-gift-wrapping/Model/Observer/ExtendOrderAttributes.php
+++ b/vendor/magento/module-gift-wrapping/Model/Observer/ExtendOrderAttributes.php
@@ -57,6 +57,9 @@ public function execute(\Magento\Framework\Event\Observer $observer)
         $shippingAddress = $quote->getShippingAddress();
 
         foreach ($this->attributes as $attribute) {
+            if ($attribute === 'gw_id' && $quote->getData('gw_id') === null) {
+                continue;
+            }
             if ($shippingAddress->hasData($attribute)) {
                 $order->setData($attribute, $shippingAddress->getData($attribute));
             }
diff --git a/vendor/magento/module-gift-wrapping-graph-ql/Model/Resolver/Cart/SetGiftOptionsOnCart.php b/vendor/magento/module-gift-wrapping-graph-ql/Model/Resolver/Cart/SetGiftOptionsOnCart.php
index 436a2e8adbda..ff57fc68128d 100644
--- a/vendor/magento/module-gift-wrapping-graph-ql/Model/Resolver/Cart/SetGiftOptionsOnCart.php
+++ b/vendor/magento/module-gift-wrapping-graph-ql/Model/Resolver/Cart/SetGiftOptionsOnCart.php
@@ -133,7 +133,9 @@ public function resolve(
         /** @var StoreInterface $store */
         $store = (int)$context->getExtensionAttributes()->getStore()->getStoreId();
 
-        if (isset($args['input']['gift_message'])) {
+        if ($args['input']['gift_message'] === null) {
+            $cart->setGiftMessageId(null);
+        } else {
             $this->saveGiftMessage($args['input']['gift_message'], $cart);
         }
 
