diff --git a/vendor/magento/module-purchase-order/Plugin/Quote/Model/QuotePlugin.php b/vendor/magento/module-purchase-order/Plugin/Quote/Model/QuotePlugin.php
index b783dc3e5adb..2e230c870bef 100644
--- a/vendor/magento/module-purchase-order/Plugin/Quote/Model/QuotePlugin.php
+++ b/vendor/magento/module-purchase-order/Plugin/Quote/Model/QuotePlugin.php
@@ -40,7 +40,7 @@ public function __construct(
      */
     public function aroundCollectTotals(Quote $subject, \Closure $proceed)
     {
-        if ($this->isPurchaseOrderQuote($subject)) {
+        if ($this->isPurchaseOrderQuote($subject) && !$subject->getIsVirtual()) {
             return $subject;
         }
         return $proceed();

