diff --git a/vendor/magento/module-checkout-address-search/Plugin/Customer/Model/Address/CustomerAddressDataProvider.php b/vendor/magento/module-checkout-address-search/Plugin/Customer/Model/Address/CustomerAddressDataProvider.php
new file mode 100644
index 000000000000..eb9c31c32a08
--- /dev/null
+++ b/vendor/magento/module-checkout-address-search/Plugin/Customer/Model/Address/CustomerAddressDataProvider.php
@@ -0,0 +1,61 @@
+<?php
+/**
+ * Copyright 2023 Adobe
+ * All Rights Reserved.
+ *
+ *  ADOBE CONFIDENTIAL
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\CheckoutAddressSearch\Plugin\Customer\Model\Address;
+
+use Magento\Customer\Model\Address\CustomerAddressDataProvider as AddressDataProvider;
+use Magento\Customer\Api\Data\CustomerInterface;
+use Magento\CheckoutAddressSearch\Model\Config as CustomerAddressSearchConfig;
+
+class CustomerAddressDataProvider
+{
+    /**
+     * @var CustomerAddressSearchConfig
+     */
+    private CustomerAddressSearchConfig $config;
+
+    /**
+     * @param CustomerAddressSearchConfig $config
+     */
+    public function __construct(CustomerAddressSearchConfig $config)
+    {
+        $this->config = $config;
+    }
+
+    /**
+     * If address search is enabled we should limit the number of addresses required
+     *
+     * @param AddressDataProvider $subject
+     * @param CustomerInterface $customer
+     * @param int|null $addressLimit
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeGetAddressDataByCustomer(
+        AddressDataProvider $subject,
+        CustomerInterface   $customer,
+        ?int                $addressLimit = null
+    ): array {
+        if ($this->config->isEnabledAddressSearch()) {
+            $addressLimit = $this->config->getSearchLimit();
+        }
+
+        return [$customer, $addressLimit];
+    }
+}
diff --git a/vendor/magento/module-checkout-address-search/etc/frontend/di.xml b/vendor/magento/module-checkout-address-search/etc/frontend/di.xml
index 0632a7808a33..bb5a52526556 100644
--- a/vendor/magento/module-checkout-address-search/etc/frontend/di.xml
+++ b/vendor/magento/module-checkout-address-search/etc/frontend/di.xml
@@ -14,4 +14,7 @@
             </argument>
         </arguments>
     </type>
+    <type name="Magento\Customer\Model\Address\CustomerAddressDataProvider">
+        <plugin name="customer_address_provider" type="Magento\CheckoutAddressSearch\Plugin\Customer\Model\Address\CustomerAddressDataProvider" sortOrder="1" disabled="false" />
+    </type>
 </config>
