diff --git a/vendor/magento/module-target-rule/Model/ResourceModel/Index.php b/vendor/magento/module-target-rule/Model/ResourceModel/Index.php
index b1b8eb3204a5..be81f7aaad82 100644
--- a/vendor/magento/module-target-rule/Model/ResourceModel/Index.php
+++ b/vendor/magento/module-target-rule/Model/ResourceModel/Index.php
@@ -365,7 +365,9 @@ protected function _prepareRuleActionSelectBind(TargetRuleIndex $object, array $
                 $v = $this->extractCategoryIdsFromRule($rule);
             }
             if (!$v) {
-                $v = $object->getProduct()->getData($bindData['field']);
+                $product = $object->getProduct();
+                $v = $product->getData($bindData['field']) ??
+                    $product->getDataUsingMethod($bindData['field']);
             }
 
             if (!empty($bindData['callback'])) {
