diff --git a/vendor/magento/module-gift-registry/Block/Email/Items.php b/vendor/magento/module-gift-registry/Block/Email/Items.php
index 3cbacb020bc2..0b570ddb212a 100644
--- a/vendor/magento/module-gift-registry/Block/Email/Items.php
+++ b/vendor/magento/module-gift-registry/Block/Email/Items.php
@@ -58,7 +58,7 @@ public function getItems()
      */
     public function getQtyOrdered($item)
     {
-        $updatedQty = $this->getEntity()->getUpdatedQty();
+        $updatedQty = $this->getEntity()?->getUpdatedQty();
         if (is_array($updatedQty) && !empty($updatedQty[$item->getId()]['ordered'])) {
             return $updatedQty[$item->getId()]['ordered'] * 1;
         }
@@ -100,7 +100,7 @@ public function getQty($item)
      */
     public function getQtyFulfilled($item)
     {
-        $updatedQty = $this->getEntity()->getUpdatedQty();
+        $updatedQty = $this->getEntity()?->getUpdatedQty();
         if (is_array($updatedQty) && !empty($updatedQty[$item->getId()]['fulfilled'])) {
             return $updatedQty[$item->getId()]['fulfilled'] * 1;
         }
