diff --git a/vendor/magento/module-login-as-customer-logging/Observer/LogSaveCustomerObserver.php b/vendor/magento/module-login-as-customer-logging/Observer/LogSaveCustomerObserver.php
index beac2bf1813..e64d57304d7 100644
--- a/vendor/magento/module-login-as-customer-logging/Observer/LogSaveCustomerObserver.php
+++ b/vendor/magento/module-login-as-customer-logging/Observer/LogSaveCustomerObserver.php
@@ -120,7 +120,8 @@ class LogSaveCustomerObserver implements ObserverInterface
         $this->eventResource->save($event);

         $customerData = $this->getFilteredData($observer->getEvent()->getCustomerDataObject()->__toArray());
-        $customerOrigData = $this->getFilteredData($observer->getEvent()->getOrigCustomerDataObject()->__toArray());
+        $customerOrigData = $observer->getEvent()->getOrigCustomerDataObject() ?
+            $this->getFilteredData($observer->getEvent()->getOrigCustomerDataObject()->__toArray()) : [];

         $this->processChanges($customerData, $customerOrigData, (int)$event->getId());
     }
