diff --git a/vendor/magento/module-multicoupon/Model/CouponManagement.php b/vendor/magento/module-multicoupon/Model/CouponManagement.php
index 7f37d005afa..d9162e2a65f 100644
--- a/vendor/magento/module-multicoupon/Model/CouponManagement.php
+++ b/vendor/magento/module-multicoupon/Model/CouponManagement.php
@@ -151,8 +151,8 @@ class CouponManagement implements CouponManagementInterface
         }
 
         $invalidCouponCodes = array_diff(
-            $couponCodes,
-            array_values($this->validateCouponCode->execute($couponCodes, $customerId))
+            array_map('strtolower', $couponCodes),
+            array_map('strtolower', array_values($this->validateCouponCode->execute($couponCodes, $customerId)))
         );
 
         if (!empty($invalidCouponCodes)) {
diff --git a/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderApplyCoupons.php b/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderApplyCoupons.php
index dfe3eda9d4e..dfe583bcd0a 100644
--- a/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderApplyCoupons.php
+++ b/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderApplyCoupons.php
@@ -163,8 +163,8 @@ class AdminCreateOrderApplyCoupons
         }
 
         $invalidCouponCodes = array_diff(
-            [$appendCouponCode],
-            array_values($this->validateCouponCode->execute([$appendCouponCode], $customerId))
+            [strtolower($appendCouponCode)],
+            array_map('strtolower', array_values($this->validateCouponCode->execute([$appendCouponCode], $customerId)))
         );
         if (!empty($invalidCouponCodes)) {
             return false;
diff --git a/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderValidateCoupon.php b/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderValidateCoupon.php
index 9aaceb9179d..d5f2e29febf 100644
--- a/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderValidateCoupon.php
+++ b/vendor/magento/module-multicoupon-ui/Plugin/AdminCreateOrderValidateCoupon.php
@@ -73,13 +73,13 @@ class AdminCreateOrderValidateCoupon
             return;
         }
 
-        if (in_array($code, $quote->getExtensionAttributes()->getCouponCodes())) {
+        if (in_array(strtolower($code), array_map('strtolower', $quote->getExtensionAttributes()->getCouponCodes()))) {
             $this->messageManager->addSuccessMessage(__('The coupon code has been accepted.'));
             return;
         }
 
         if (count($quote->getExtensionAttributes()->getCouponCodes())
-            >= $this->config->getMaximumNumberOfCoupons()
+            > $this->config->getMaximumNumberOfCoupons()
         ) {
             $this->messageManager->addErrorMessage(
                 __('Maximum allowed number of applied coupons was exceeded.')
