diff --git a/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/Errors/AbstractErrors.php b/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/Errors/AbstractErrors.php
index a3ae0c309b00..763123c2275e 100644
--- a/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/Errors/AbstractErrors.php
+++ b/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/Errors/AbstractErrors.php
@@ -166,6 +166,8 @@ public function getCart()
     {
         if (!isset($this->_cart)) {
             $this->_cart = $this->_cartFactory->create();
+            $this->_cart->setCurrentStore($this->getStore());
+            $this->_cart->getQuote()->setStoreId($this->_storeManager->getStore($this->getStore())->getId());
         }
         return $this->_cart;
     }
diff --git a/vendor/magento/module-advanced-checkout/Controller/Adminhtml/Index.php b/vendor/magento/module-advanced-checkout/Controller/Adminhtml/Index.php
index 007eba916a31..1384e2b77e24 100644
--- a/vendor/magento/module-advanced-checkout/Controller/Adminhtml/Index.php
+++ b/vendor/magento/module-advanced-checkout/Controller/Adminhtml/Index.php
@@ -147,6 +147,7 @@ protected function _initData($useRedirects = true)
         } else {
             // try to find quote for selected store
             $cart->setStoreId($storeId);
+            $cart->setCurrentStore($storeId);
         }
 
         $quote = $cart->getQuote();

