diff --git a/vendor/magento/module-async-order/Model/CartRepository.php b/vendor/magento/module-async-order/Model/CartRepository.php
index 3f6946c483ce..59c39a567b98 100644
--- a/vendor/magento/module-async-order/Model/CartRepository.php
+++ b/vendor/magento/module-async-order/Model/CartRepository.php
@@ -9,6 +9,7 @@
 namespace Magento\AsyncOrder\Model;
 
 use Magento\Framework\Api\SearchCriteriaInterface;
+use Magento\Framework\App\DeploymentConfig;
 use Magento\Quote\Api\CartRepositoryInterface;
 use Magento\Quote\Api\Data\CartInterface;
 use Magento\Quote\Model\QuoteRepository;
@@ -23,15 +24,23 @@ class CartRepository implements CartRepositoryInterface
      */
     private $quoteRepository;
 
+    /**
+     * @var DeploymentConfig
+     */
+    private DeploymentConfig $deploymentConfig;
+
     /**
      * Constructor
      *
      * @param QuoteRepository $quoteRepository
+     * @param DeploymentConfig $deploymentConfig
      */
     public function __construct(
-        QuoteRepository $quoteRepository
+        QuoteRepository $quoteRepository,
+        DeploymentConfig $deploymentConfig
     ) {
         $this->quoteRepository = $quoteRepository;
+        $this->deploymentConfig = $deploymentConfig;
     }
 
     /**
@@ -55,6 +64,9 @@ public function getForCustomer($customerId, array $sharedStoreIds = [])
      */
     public function getActive($cartId, array $sharedStoreIds = [])
     {
+        if (!$this->deploymentConfig->get(OrderManagement::ASYNC_ORDER_OPTION_PATH)) {
+            return $this->quoteRepository->getActive($cartId, $sharedStoreIds);
+        }
         return $this->get($cartId, $sharedStoreIds);
     }
 
