diff --git a/vendor/magento/module-negotiable-quote/Cron/SendEmails.php b/vendor/magento/module-negotiable-quote/Cron/SendEmails.php
index c7a737614797..055e3731cd10 100644
--- a/vendor/magento/module-negotiable-quote/Cron/SendEmails.php
+++ b/vendor/magento/module-negotiable-quote/Cron/SendEmails.php
@@ -1,10 +1,25 @@
 <?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+/************************************************************************
+ * ADOBE CONFIDENTIAL
+ *
+ * Copyright 2016 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ * ************************************************************************
  */
+declare(strict_types=1);
+
 namespace Magento\NegotiableQuote\Cron;
 
+use Magento\NegotiableQuote\Api\Data\NegotiableQuoteInterface;
 use Magento\Store\Model\ScopeInterface;
 
 /**
@@ -15,27 +30,27 @@ class SendEmails
     /**#@+
      * Configuration paths for email templates.
      */
-    const CONFIG_QUOTE_EMAIL_NOTIFICATIONS_ENABLED = 'sales_email/quote/enabled';
-    const EXPIRE_ONE_DAY_TEMPLATE = 'sales_email/quote/expire_one_day_template';
-    const EXPIRE_TWO_DAYS_TEMPLATE = 'sales_email/quote/expire_two_days_template';
-    const EXPIRED_TEMPLATE = 'sales_email/quote/expire_occur_template';
+    public const CONFIG_QUOTE_EMAIL_NOTIFICATIONS_ENABLED = 'sales_email/quote/enabled';
+    public const EXPIRE_ONE_DAY_TEMPLATE = 'sales_email/quote/expire_one_day_template';
+    public const EXPIRE_TWO_DAYS_TEMPLATE = 'sales_email/quote/expire_two_days_template';
+    public const EXPIRED_TEMPLATE = 'sales_email/quote/expire_occur_template';
     /**#@-*/
 
     /**#@+
      * Email notification statuses.
      */
-    const EMAIL_IS_NOT_SENT_COUNTER = 0;
-    const EMAIL_SENT_TWO_DAYS_COUNTER = 1;
-    const EMAIL_SENT_ONE_DAY_COUNTER = 2;
-    const EMAIL_SENT_ZERO_DAY_COUNTER = 3;
+    public const EMAIL_IS_NOT_SENT_COUNTER = 0;
+    public const EMAIL_SENT_TWO_DAYS_COUNTER = 1;
+    public const EMAIL_SENT_ONE_DAY_COUNTER = 2;
+    public const EMAIL_SENT_ZERO_DAY_COUNTER = 3;
     /**#@-*/
 
     /**#@+
      * Timing for sending notifications.
      */
-    const EMAIL_NOTIFICATION_TWO_DAYS = '+ 2 day';
-    const EMAIL_NOTIFICATION_ONE_DAY = '+ 1 day';
-    const EMAIL_NOTIFICATION_ZERO_DAY = '- 1 day';
+    public const EMAIL_NOTIFICATION_TWO_DAYS = '+ 2 day';
+    public const EMAIL_NOTIFICATION_ONE_DAY = '+ 1 day';
+    public const EMAIL_NOTIFICATION_ZERO_DAY = '- 1 day';
     /**#@-*/
 
     /**
@@ -232,6 +247,15 @@ private function getQuotes($statusEmailNotification, $days = '')
                     ->create(),
             ]
         );
+        $this->searchCriteriaBuilder->addFilters(
+            [
+                $this->filterBuilder
+                    ->setField('extension_attribute_negotiable_quote.status')
+                    ->setConditionType('neq')
+                    ->setValue(NegotiableQuoteInterface::STATUS_ORDERED)
+                    ->create(),
+            ]
+        );
         $searchCriteria = $this->searchCriteriaBuilder->create();
         $quotes = $this->negotiableQuoteRepository->getList($searchCriteria)->getItems();
 
