diff --git a/vendor/magento/module-negotiable-quote/view/frontend/web/js/action/place-order-negotiable-quote.js b/vendor/magento/module-negotiable-quote/view/frontend/web/js/action/place-order-negotiable-quote.js
index fe61e14946c7..1001ab82386e 100644
--- a/vendor/magento/module-negotiable-quote/view/frontend/web/js/action/place-order-negotiable-quote.js
+++ b/vendor/magento/module-negotiable-quote/view/frontend/web/js/action/place-order-negotiable-quote.js
@@ -7,8 +7,9 @@ define([
     'Magento_Checkout/js/model/quote',
     'Magento_Checkout/js/model/url-builder',
     'Magento_Checkout/js/model/error-processor',
-    'Magento_Checkout/js/model/full-screen-loader'
-], function (storage, quote, urlBuilder, errorProcessor, fullScreenLoader) {
+    'Magento_Checkout/js/model/full-screen-loader',
+    'Magento_Checkout/js/model/place-order'
+], function (storage, quote, urlBuilder, errorProcessor, fullScreenLoader, placeOrderService) {
     'use strict';
 
     return function (paymentData, messageContainer) {
@@ -26,12 +27,6 @@ define([
 
         fullScreenLoader.startLoader();
 
-        return storage.post(
-            serviceUrl,
-            JSON.stringify(payload)
-        ).fail(function (response) {
-            errorProcessor.process(response, messageContainer);
-            fullScreenLoader.stopLoader();
-        });
+        return placeOrderService(serviceUrl, payload, messageContainer);
     };
 });
