diff --git a/vendor/magento/module-purchase-order-rule/Model/Rule/CreateCondition.php b/vendor/magento/module-purchase-order-rule/Model/Rule/CreateCondition.php
index 814e4d59e2f2..c6ab167fb7c2 100644
--- a/vendor/magento/module-purchase-order-rule/Model/Rule/CreateCondition.php
+++ b/vendor/magento/module-purchase-order-rule/Model/Rule/CreateCondition.php
@@ -88,6 +88,7 @@ private function getCondition(array $conditionData): ConditionInterface
         $value = $conditionData['value']
             ?? $conditionData['amount']['value']
             ?? $conditionData['quantity']['value']
+            ?? $conditionData['quantity']
             ?? null;
         $currencyCode = $conditionData['currency_code']
             ?? $conditionData['amount']['currency']

