diff --git a/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php b/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php
index f54dd685ed4..6390bc8a00c 100644
--- a/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php
+++ b/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php
@@ -97,8 +97,8 @@ class CollectionFilter
     private function filterCollection(AbstractCollection $collection)
     {
         if (!$this->role->getIsAll() && !$collection->getFlag(self::FILTERED_FLAG_NAME)) {
-            if (method_exists($collection, 'addStoreFilter')) {
-                $collection->addStoreFilter($this->getStoreIds());
+            if (method_exists($collection, 'addStoresFilter')) {
+                $collection->addStoresFilter($this->getStoreIds());
                 $collection->setFlag(self::FILTERED_FLAG_NAME, true);
             } elseif (isset($collection->getSelect()->getPart(Select::FROM)['main_table'])) {
                 $this->tableBasedFilter($collection);
