diff --git a/vendor/magento/module-gift-wrapping/view/adminhtml/templates/order/create/info.phtml b/vendor/magento/module-gift-wrapping/view/adminhtml/templates/order/create/info.phtml
index d38324b93dea..b5e898217d83 100644
--- a/vendor/magento/module-gift-wrapping/view/adminhtml/templates/order/create/info.phtml
+++ b/vendor/magento/module-gift-wrapping/view/adminhtml/templates/order/create/info.phtml
@@ -140,6 +140,6 @@ giftWrappingOrder.setCardPrice(null, {$wrappingPrintedCardValue});
 //]]>
 
 });
-script
+script;
 ?>
 <?= /* @noEscape */ $secureRenderer->renderTag('script', [], $scriptString, false); ?>
