diff --git a/vendor/magento/module-reward/Model/Total/Creditmemo/Reward.php b/vendor/magento/module-reward/Model/Total/Creditmemo/Reward.php
index c1e68c86b7d6..dd6c363c6419 100644
--- a/vendor/magento/module-reward/Model/Total/Creditmemo/Reward.php
+++ b/vendor/magento/module-reward/Model/Total/Creditmemo/Reward.php
@@ -50,8 +50,8 @@ public function collect(Creditmemo $creditmemo)
             $order->getBaseRwrdCrrncyAmntRefnded();
         if ($order->getBaseRewardCurrencyAmount() && $baseRewardCurrencyAmountLeft > 0) {
             if ($baseRewardCurrencyAmountLeft >= $creditmemo->getBaseGrandTotal()) {
-                $rewardCurrencyAmountLeft = $creditmemo->getGrandTotal();
-                $baseRewardCurrencyAmountLeft = $creditmemo->getBaseGrandTotal();
+                $rewardCurrencyAmountLeft = $creditmemo->getGrandTotal() ?: $rewardCurrencyAmountLeft;
+                $baseRewardCurrencyAmountLeft = $creditmemo->getBaseGrandTotal() ?: $baseRewardCurrencyAmountLeft;
                 $creditmemo->setGrandTotal(0);
                 $creditmemo->setBaseGrandTotal(0);
                 $creditmemo->setAllowZeroGrandTotal(true);
