diff --git a/vendor/magento/module-catalog-permissions/Model/Indexer/Category/Action/Rows.php b/vendor/magento/module-catalog-permissions/Model/Indexer/Category/Action/Rows.php
index 3ff1c34d8486..fd54653fe665 100644
--- a/vendor/magento/module-catalog-permissions/Model/Indexer/Category/Action/Rows.php
+++ b/vendor/magento/module-catalog-permissions/Model/Indexer/Category/Action/Rows.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2025 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\CatalogPermissions\Model\Indexer\Category\Action;
 
@@ -136,7 +136,6 @@ public function execute(array $entityIds = [], $useIndexTempTable = false)
                 } else {
                     $this->removeObsoleteIndexData();
                 }
-                $this->reindex();
                 $this->connection->commit();
             } catch (\Exception $exception) {
                 $this->connection->rollBack();
@@ -144,6 +143,7 @@ public function execute(array $entityIds = [], $useIndexTempTable = false)
             }
         }
 
+        $this->reindex();
         $this->cleanCache();
     }
 
