diff --git a/vendor/magento/module-google-tag-manager/Plugin/Helper/Data.php b/vendor/magento/module-google-tag-manager/Plugin/Helper/Data.php
index 480409289c85..aff3a7291c55 100644
--- a/vendor/magento/module-google-tag-manager/Plugin/Helper/Data.php
+++ b/vendor/magento/module-google-tag-manager/Plugin/Helper/Data.php
@@ -36,6 +36,6 @@ public function __construct(
      */
     public function afterIsTagManagerAvailable(Helper $subject, $result)
     {
-        return $this->tagManagerConfig->isTagManagerAvailable() ?? $result;
+        return $this->tagManagerConfig->isTagManagerAvailable() ?: $result;
     }
 }
