diff --git a/vendor/magento/module-gift-card/Model/GiftCardItemEmail.php b/vendor/magento/module-gift-card/Model/GiftCardItemEmail.php
index c3364677db6a..c669225fc26c 100644
--- a/vendor/magento/module-gift-card/Model/GiftCardItemEmail.php
+++ b/vendor/magento/module-gift-card/Model/GiftCardItemEmail.php
@@ -128,7 +128,7 @@ public function send(
         $transport = $this->transportBuilder->setTemplateIdentifier($template)
             ->setTemplateOptions($templateOptions)
             ->setTemplateVars($templateData)
-            ->setFrom($emailIdentity)
+            ->setFromByScope($emailIdentity, $storeId)
             ->addTo($recipientAddress, $recipientName)
             ->getTransport();
         $transport->sendMessage();
