diff --git a/vendor/magento/module-company/view/frontend/web/js/hierarchy-tree.js b/vendor/magento/module-company/view/frontend/web/js/hierarchy-tree.js
index 7b24de54f1dd..157715269e7b 100755
--- a/vendor/magento/module-company/view/frontend/web/js/hierarchy-tree.js
+++ b/vendor/magento/module-company/view/frontend/web/js/hierarchy-tree.js
@@ -809,7 +809,7 @@ define([
                 self._filterRoles(name, value);
             }

-            this.options.popup.find('form [id="' + id + '"]').val(value);
+            this.options.popups.user.find('form [id="' + id + '"]').val(value);
         },

         /**
@@ -826,7 +826,7 @@ define([
                 self._filterRoles(name, value);
             }

-            this.options.popup.find('form [name="' + name + '"]').val(selectValues);
+            this.options.popups.user.find('form [name="' + name + '"]').val(selectValues);
         },

         /**
