diff --git a/vendor/magento/module-company-graph-ql/Model/Resolver/Company/Structure.php b/vendor/magento/module-company-graph-ql/Model/Resolver/Company/Structure.php
index 8626d75c772b..1c68ee70e7ef 100644
--- a/vendor/magento/module-company-graph-ql/Model/Resolver/Company/Structure.php
+++ b/vendor/magento/module-company-graph-ql/Model/Resolver/Company/Structure.php
@@ -118,7 +118,6 @@ public function resolve(
                 $structureItems = null;
             } else {
                 $this->treeManagement->addDataToTree($tree);
-                $this->treeManagement->filterTree($tree, 'is_active', true);
                 $structureItems = $structure->getStructureItems($tree, (int)$args['depth']);
             }
         } catch (\Exception $e) {
