diff --git a/vendor/magento/module-reward/Model/Reward.php b/vendor/magento/module-reward/Model/Reward.php
index 476b9b9592c4..731d4ed3ad0c 100644
--- a/vendor/magento/module-reward/Model/Reward.php
+++ b/vendor/magento/module-reward/Model/Reward.php
@@ -10,8 +10,6 @@
 use Magento\Customer\Model\Customer;
 
 /**
- * Reward model
- *
  * @method int getCustomerId()
  * @method \Magento\Reward\Model\Reward setCustomerId(int $value)
  * @method \Magento\Reward\Model\Reward setWebsiteId(int $value)
@@ -27,37 +25,37 @@
  */
 class Reward extends \Magento\Framework\Model\AbstractModel
 {
-    const XML_PATH_BALANCE_UPDATE_TEMPLATE = 'magento_reward/notification/balance_update_template';
+    public const XML_PATH_BALANCE_UPDATE_TEMPLATE = 'magento_reward/notification/balance_update_template';
 
-    const XML_PATH_BALANCE_WARNING_TEMPLATE = 'magento_reward/notification/expiry_warning_template';
+    public const XML_PATH_BALANCE_WARNING_TEMPLATE = 'magento_reward/notification/expiry_warning_template';
 
-    const XML_PATH_EMAIL_IDENTITY = 'magento_reward/notification/email_sender';
+    public const XML_PATH_EMAIL_IDENTITY = 'magento_reward/notification/email_sender';
 
-    const XML_PATH_MIN_POINTS_BALANCE = 'magento_reward/general/min_points_balance';
+    public const XML_PATH_MIN_POINTS_BALANCE = 'magento_reward/general/min_points_balance';
 
-    const REWARD_ACTION_ADMIN = 0;
+    public const REWARD_ACTION_ADMIN = 0;
 
-    const REWARD_ACTION_ORDER = 1;
+    public const REWARD_ACTION_ORDER = 1;
 
-    const REWARD_ACTION_REGISTER = 2;
+    public const REWARD_ACTION_REGISTER = 2;
 
-    const REWARD_ACTION_NEWSLETTER = 3;
+    public const REWARD_ACTION_NEWSLETTER = 3;
 
-    const REWARD_ACTION_INVITATION_CUSTOMER = 4;
+    public const REWARD_ACTION_INVITATION_CUSTOMER = 4;
 
-    const REWARD_ACTION_INVITATION_ORDER = 5;
+    public const REWARD_ACTION_INVITATION_ORDER = 5;
 
-    const REWARD_ACTION_REVIEW = 6;
+    public const REWARD_ACTION_REVIEW = 6;
 
-    const REWARD_ACTION_ORDER_EXTRA = 8;
+    public const REWARD_ACTION_ORDER_EXTRA = 8;
 
-    const REWARD_ACTION_CREDITMEMO = 9;
+    public const REWARD_ACTION_CREDITMEMO = 9;
 
-    const REWARD_ACTION_SALESRULE = 10;
+    public const REWARD_ACTION_SALESRULE = 10;
 
-    const REWARD_ACTION_REVERT = 11;
+    public const REWARD_ACTION_REVERT = 11;
 
-    const REWARD_ACTION_CREDITMEMO_VOID = 12;
+    public const REWARD_ACTION_CREDITMEMO_VOID = 12;
 
     /**
      * Model is loaded by customer
@@ -74,8 +72,6 @@ class Reward extends \Magento\Framework\Model\AbstractModel
     protected static $_actionModelClasses = [];
 
     /**
-     * Rates
-     *
      * @var array
      */
     protected $_rates = [];
@@ -88,22 +84,16 @@ class Reward extends \Magento\Framework\Model\AbstractModel
     protected $_rewardPointsUpdated = false;
 
     /**
-     * Reward data
-     *
      * @var \Magento\Reward\Helper\Data
      */
     protected $_rewardData = null;
 
     /**
-     * Reward customer
-     *
      * @var \Magento\Reward\Helper\Customer
      */
     protected $_rewardCustomer = null;
 
     /**
-     * Core model store manager interface
-     *
      * @var \Magento\Store\Model\StoreManagerInterface
      */
     protected $_storeManager;
@@ -114,8 +104,6 @@ class Reward extends \Magento\Framework\Model\AbstractModel
     protected $_localeCurrency;
 
     /**
-     * Reward history factory
-     *
      * @var \Magento\Reward\Model\Reward\HistoryFactory
      */
     protected $_historyFactory;
@@ -237,13 +225,13 @@ protected function _construct()
      * @param int $actionId
      * @param string $actionModelClass
      * @return void
-     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws LocalizedException
      */
     // phpcs:ignore
     public static function setActionModelClass($actionId, $actionModelClass)
     {
         if (!is_int($actionId)) {
-            throw new \Magento\Framework\Exception\LocalizedException(
+            throw new LocalizedException(
                 __('The action ID you enter must be a number.')
             );
         }
@@ -811,8 +799,9 @@ public function sendBalanceUpdateNotification()
                 'update_message' => $this->getHistory()->getMessage(),
                 'update_comment' => $history->getComment(),
             ]
-        )->setFrom(
-            $from
+        )->setFromByScope(
+            $from,
+            $store->getStoreId()
         )->addTo(
             $this->getCustomer()->getEmail()
         );
