diff --git a/vendor/magento/module-catalog-permissions/Model/Indexer/Plugin/Import.php b/vendor/magento/module-catalog-permissions/Model/Indexer/Plugin/Import.php
index ff4413f6e473..725230b88e15 100644
--- a/vendor/magento/module-catalog-permissions/Model/Indexer/Plugin/Import.php
+++ b/vendor/magento/module-catalog-permissions/Model/Indexer/Plugin/Import.php
@@ -5,6 +5,9 @@
  */
 namespace Magento\CatalogPermissions\Model\Indexer\Plugin;
 
+use Magento\CatalogPermissions\Model\Indexer\Category as CategoryIndexer;
+use Magento\CatalogPermissions\Model\Indexer\Product as ProductIndexer;
+
 class Import
 {
     /**
@@ -40,8 +43,14 @@ public function __construct(
     public function afterImportSource(\Magento\ImportExport\Model\Import $subject, $import)
     {
         if ($this->config->isEnabled()) {
-            $this->indexerRegistry->get(\Magento\CatalogPermissions\Model\Indexer\Category::INDEXER_ID)->invalidate();
-            $this->indexerRegistry->get(\Magento\CatalogPermissions\Model\Indexer\Product::INDEXER_ID)->invalidate();
+            $categoryIndexer = $this->indexerRegistry->get(CategoryIndexer::INDEXER_ID);
+            if (!$categoryIndexer->isScheduled()) {
+                $categoryIndexer->invalidate();
+            }
+            $productIndexer = $this->indexerRegistry->get(ProductIndexer::INDEXER_ID);
+            if (!$productIndexer->isScheduled()) {
+                $productIndexer->invalidate();
+            }
         }
         return $import;
     }
