diff --git a/vendor/magento/module-advanced-checkout/Plugin/Escaper.php b/vendor/magento/module-advanced-checkout/Plugin/Escaper.php
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/vendor/magento/module-advanced-checkout/view/adminhtml/web/addbysku.js b/vendor/magento/module-advanced-checkout/view/adminhtml/web/addbysku.js
index 3caf47812723..f3485cad2b10 100644
--- a/vendor/magento/module-advanced-checkout/view/adminhtml/web/addbysku.js
+++ b/vendor/magento/module-advanced-checkout/view/adminhtml/web/addbysku.js
@@ -466,7 +466,7 @@ define([
          */
         configure: function (id, sku) {
             // eslint-disable-next-line jquery-no-input-event-shorthand
-            var productRow = $('sku_errors_table').select('div[id=sku_' + sku + ']')[0],
+            var productRow = $('sku_errors_table').select('div[id="sku_' + sku + '"]')[0],
                 noticeElement = productRow.select('.message-notice'),
                 productQtyElement = productRow.up('tr').select('input[name=qty]')[0],
                 productPriceElement = productRow.up('tr').select('td.col-price')[0];
