diff --git a/vendor/magento/module-negotiable-quote/Plugin/Sales/Block/Order/Info/CreationInfoPlugin.php b/vendor/magento/module-negotiable-quote/Plugin/Sales/Block/Order/Info/CreationInfoPlugin.php
index 76d267c99c8c..837d621b1854 100644
--- a/vendor/magento/module-negotiable-quote/Plugin/Sales/Block/Order/Info/CreationInfoPlugin.php
+++ b/vendor/magento/module-negotiable-quote/Plugin/Sales/Block/Order/Info/CreationInfoPlugin.php
@@ -87,7 +87,7 @@ public function afterGetCreationInfo(CreationInfo $subject, string $result): str
             $customerId = $this->getCustomerId($orderId);
             $order = $this->orderRepository->get($orderId);
 
-            if ($order && $order->getEntityId() && $order->getCustomerId() === $customerId) {
+            if ($order && $order->getEntityId() && (int)$order->getCustomerId() === $customerId) {
                 return $result;
             }
         }
