diff --git a/vendor/magento/module-gift-card-account/Model/Plugin/CreditmemoRepository.php b/vendor/magento/module-gift-card-account/Model/Plugin/CreditmemoRepository.php
index 4e1c9a14b325..c1ee04482c9a 100644
--- a/vendor/magento/module-gift-card-account/Model/Plugin/CreditmemoRepository.php
+++ b/vendor/magento/module-gift-card-account/Model/Plugin/CreditmemoRepository.php
@@ -167,7 +167,7 @@ public function afterSave(CreditmemoRepositoryInterface $subject, CreditmemoInte
         $order = ($entity->getOrder()) ?: $this->getOrderById((int)$entity->getOrderId());
         $orderItems = $this->getOrderItems($order);
         if ($orderItems) {
-            foreach ($entity->getItems() as $creditMemoItem) {
+            foreach ($entity->getAllItems() as $creditMemoItem) {
                 $orderItem = $orderItems[$creditMemoItem->getOrderItemId()];
                 if ($orderItem->getProductOptionByCode('giftcard_created_codes')) {
                     $this->removeGiftCardAccount(
