diff --git a/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php b/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php
index f5032017822..4b833afaa67 100644
--- a/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php
+++ b/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php
@@ -364,6 +364,6 @@ class SqlBuilder
      */
     protected function normalizeConditionValue($condition)
     {
-        return $condition->getValue();
+        return $condition->getValueParsed();
     }
 }
