diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml b/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml
index 7824cbd9118..210da89863c 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml
@@ -23,6 +23,7 @@
                     <argument name="code" xsi:type="string">Magento\Customer\Model\Customer</argument>
                 </action>
                 <arguments>
+                    <argument name="edit_account_button_view_model" xsi:type="object">Magento\Customer\ViewModel\EditAccountButton</argument>
                     <argument name="view_model" xsi:type="object">CustomerFileAttribute</argument>
                 </arguments>
             </block>
diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml
index a8aa1667d47..5cb2769b92d 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml
@@ -11,6 +11,7 @@ use Magento\Customer\Block\Widget\Gender;
 use Magento\Customer\Block\Widget\Name;
 use Magento\Customer\Block\Widget\Taxvat;
 
+$editAccountButtonViewModel = $block->getData('edit_account_button_view_model');
 ?>
 <form class="form form-edit-account" action="<?= /* @noEscape */ $block->getUrl('customer/account/editPost') ?>"
       method="post" id="form-validate" enctype="multipart/form-data"
@@ -126,7 +127,9 @@ use Magento\Customer\Block\Widget\Taxvat;
 
     <div class="actions-toolbar">
         <div class="primary">
-            <button type="submit" class="action save primary" title="<?= $block->escapeHtmlAttr(__('Save')) ?>">
+            <button type="submit" class="action save primary" title="<?= $block->escapeHtmlAttr(__('Save')) ?>"
+                    id="send2"
+                <?php if ($editAccountButtonViewModel->disabled()): ?> disabled="disabled" <?php endif; ?>>
                 <span><?= $block->escapeHtml(__('Save')) ?></span>
             </button>
         </div>
diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
index 74e8ca58a6a..ec69945183f 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
@@ -21,6 +21,8 @@
 $addressHelper = $block->getData('addressHelper');
 /** @var \Magento\Directory\Helper\Data $directoryHelper */
 $directoryHelper = $block->getData('directoryHelper');
+/** @var \Magento\Customer\ViewModel\CreateAccountButton $createAccountButtonViewModel */
+$createAccountButtonViewModel = $block->getData('create_account_button_view_model');
 ?>
 <form class="form create account form-create-account"
       action="<?= $block->escapeUrl($block->getPostActionUrl()) ?>"
@@ -294,7 +296,9 @@ $directoryHelper = $block->getData('directoryHelper');
         <div class="primary">
             <button type="submit"
                     class="action submit primary"
-                    title="<?= $block->escapeHtmlAttr(__('Create an Account')) ?>">
+                    title="<?= $block->escapeHtmlAttr(__('Create an Account')) ?>"
+                    id="send2"
+                <?php if ($createAccountButtonViewModel->disabled()): ?> disabled="disabled" <?php endif; ?>>
                 <span><?= $block->escapeHtml(__('Create an Account')) ?></span>
             </button>
         </div>
@@ -310,6 +314,7 @@ $directoryHelper = $block->getData('directoryHelper');
     <?php
     $displayAllRegions = $block->getConfig("general/region/display_all") ? "true" : "false";
     ?>
+<?php // phpcs:ignore Magento2.Legacy.PhtmlTemplate ?>
 <script type="text/x-magento-init">
     {
         "#country": {
@@ -327,6 +332,7 @@ $directoryHelper = $block->getData('directoryHelper');
     }
 </script>
 <?php endif; ?>
+<?php // phpcs:ignore Magento2.Legacy.PhtmlTemplate ?>
 <script type="text/x-magento-init">
     {
         "*": {
