diff --git a/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/AbstractSku.php b/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/AbstractSku.php
index 06f702316482..6fe209f6fac8 100644
--- a/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/AbstractSku.php
+++ b/vendor/magento/module-advanced-checkout/Block/Adminhtml/Sku/AbstractSku.php
@@ -1,11 +1,24 @@
 <?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+/************************************************************************
+ *
+ * ADOBE CONFIDENTIAL
+ * ___________________
+ *
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ * ************************************************************************
  */
+
 namespace Magento\AdvancedCheckout\Block\Adminhtml\Sku;
-use Magento\Framework\App\ObjectManager;
-use Magento\Framework\View\Helper\SecureHtmlRenderer;
 
 /**
  * Admin Checkout main form container
@@ -22,7 +35,7 @@ abstract class AbstractSku extends \Magento\Backend\Block\Template
     /**
      * List type of current block
      */
-    const LIST_TYPE = 'add_by_sku';
+    public const LIST_TYPE = 'add_by_sku';
 
     /**
      * @var string
@@ -34,27 +47,19 @@ abstract class AbstractSku extends \Magento\Backend\Block\Template
      */
     protected $_jsonEncoder;
 
-    /**
-     * @var SecureHtmlRenderer
-     */
-    private $secureRenderer;
-
     /**
      * @codeCoverageIgnore
      * @param \Magento\Backend\Block\Template\Context $context
      * @param \Magento\Framework\Json\EncoderInterface $jsonEncoder
      * @param array $data
-     * @param SecureHtmlRenderer|null $secureRenderer
      * @codeCoverageIgnore
      */
     public function __construct(
         \Magento\Backend\Block\Template\Context $context,
         \Magento\Framework\Json\EncoderInterface $jsonEncoder,
-        array $data = [],
-        SecureHtmlRenderer $secureRenderer = null
+        array $data = []
     ) {
         $this->_jsonEncoder = $jsonEncoder;
-        $this->secureRenderer = $secureRenderer ?? ObjectManager::getInstance()->get(SecureHtmlRenderer::class);
         parent::__construct($context, $data);
     }
 
@@ -79,33 +84,10 @@ protected function _construct()
      */
     protected function _prepareLayout()
     {
-        //Delete button will be copied for each row so we need a listener that will work for duplicates.
-        $deleteButtonId = $this->mathRandom->getRandomString('8');
-        $deleteButtonClass = 'admin-checkout-sku-delete-button-' . $deleteButtonId;
-        $deleteFunctionName = 'skuDeleteButtonListener' . $deleteButtonId;
-        $deleteActionScript = <<<SCRIPT
-            if (typeof($deleteFunctionName) == "undefined") {
-                $deleteFunctionName = function (event) {
- 		            addBySku.del(event.target);
- 	 		    };
- 	 	 	    require(['jquery'], function($){
- 	 	 	        $("body").on("click", ".$deleteButtonClass", $deleteFunctionName);
- 	 	 	    });
- 	 	 	}
-SCRIPT;
         $this->addChild(
             'deleteButton',
             \Magento\Backend\Block\Widget\Button::class,
-            [
-                'label' => '',
-                'class' => 'action-delete ' . $deleteButtonClass,
-                'before_html' => $this->secureRenderer->renderTag(
-                    'script',
-                    ['type' => 'text/javascript'],
-                    $deleteActionScript,
-                    false
-                ),
-            ]
+            ['label' => '', 'class' => 'action-delete']
         );
 
         $this->addChild(
diff --git a/vendor/magento/module-advanced-checkout/view/adminhtml/templates/sku/add.phtml b/vendor/magento/module-advanced-checkout/view/adminhtml/templates/sku/add.phtml
index c84b601f5df9..b85a7f5254b6 100644
--- a/vendor/magento/module-advanced-checkout/view/adminhtml/templates/sku/add.phtml
+++ b/vendor/magento/module-advanced-checkout/view/adminhtml/templates/sku/add.phtml
@@ -1,12 +1,28 @@
 <?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+/************************************************************************
+ *
+ * ADOBE CONFIDENTIAL
+ * ___________________
+ *
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ * ************************************************************************
  */
 // phpcs:disable Generic.Files.LineLength
 
 /**
  * @var \Magento\Framework\View\Helper\SecureHtmlRenderer $secureRenderer
+ * @var \Magento\Framework\Escaper $escaper
+ * @var \Magento\AdvancedCheckout\Block\Adminhtml\Sku\AbstractSku $block
  */
 ?>
 <div class="add-by-sku-wrapper">
@@ -18,8 +34,8 @@
                     class="admin__control-table">
                     <thead>
                     <tr class="headings">
-                        <th class="col-sku"><?= $block->escapeHtml(__('SKU number')) ?></th>
-                        <th class="col-qty"><?= $block->escapeHtml(__('Qty')) ?></th>
+                        <th class="col-sku"><?= $escaper->escapeHtml(__('SKU number')) ?></th>
+                        <th class="col-qty"><?= $escaper->escapeHtml(__('Qty')) ?></th>
                         <th class="col-actions last">&nbsp;</th>
                     </tr>
                     </thead>
@@ -51,10 +67,10 @@
         </div>
     </div>
     <div class="admin__field add-sku-csv">
-        <label class="admin__field-label" for="sku_upload"><?= $block->escapeHtml(__('Upload File')) ?></label>
+        <label class="admin__field-label" for="sku_upload"><?= $escaper->escapeHtml(__('Upload File')) ?></label>
         <div class="admin__field-control">
             <div class="admin__field-note">
-                <span><?= $block->escapeHtml(__('Your csv file must include "sku" and "qty" columns.')) ?></span>
+                <span><?= $escaper->escapeHtml(__('Your csv file must include "sku" and "qty" columns.')) ?></span>
             </div>
             <?php $inputName = \Magento\AdvancedCheckout\Model\Import::FIELD_NAME_SOURCE_FILE; ?>
             <input
@@ -65,7 +81,7 @@
                 value=""  />
             <button type="button"
                     class="action-default action-reset">
-                <span><?= $block->escapeHtml(__('Reset')) ?></span>
+                <span><?= $escaper->escapeHtml(__('Reset')) ?></span>
             </button>
         </div>
         <?= /* @noEscape */
@@ -79,7 +95,7 @@
 </div>
 
 <?php
-$adminCheckoutInstanceName = $block->escapeJs($block->getJsOrderObject());
+$adminCheckoutInstanceName = $escaper->escapeJs($block->getJsOrderObject());
 $addBySkuData = $block->getAddBySkuDataJson();
 $additionalJavaScript = $block->getAdditionalJavascript();
 $contextSpecificJs = $block->getContextSpecificJs();
@@ -115,7 +131,7 @@ require([
     }
 
     window.initSku = initSku;
-    $('sku_table').on('click', "[data-ui-id='sku-accordion-deletebutton']", function(event) {
+    jQuery('#sku_table').on('click', '.action-delete', function(event) {
         addBySku.del(event.target);
     });
 

