diff --git a/vendor/magento/module-shared-catalog/Observer/UpdateConfig.php b/vendor/magento/module-shared-catalog/Observer/UpdateConfig.php
index c1aa03c34852..c092e24be8cc 100644
--- a/vendor/magento/module-shared-catalog/Observer/UpdateConfig.php
+++ b/vendor/magento/module-shared-catalog/Observer/UpdateConfig.php
@@ -18,9 +18,12 @@
 use Magento\Store\Model\ScopeInterface;
 use Magento\SharedCatalog\Model\CatalogPermissionManagement;
 use Magento\Framework\App\Config\ConfigResource\ConfigInterface as ConfigResource;
+use Magento\Store\Model\ScopeTypeNormalizer;

 /**
  * Additional actions after saving data to system config.
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  */
 class UpdateConfig implements ObserverInterface
 {
@@ -59,6 +62,11 @@ class UpdateConfig implements ObserverInterface
      */
     private $permissionsSynchronizer;

+    /**
+     * @var ScopeTypeNormalizer
+     */
+    private $scopeTypeNormalizer;
+
     /**
      * @param StatusServiceInterface $companyStatusService
      * @param SharedCatalogModuleConfig $sharedCatalogModuleConfig
@@ -67,6 +75,7 @@ class UpdateConfig implements ObserverInterface
      * @param ConfigResource $configResource
      * @param ScopeResolverPool $scopeResolverPool
      * @param Synchronizer $permissionsSynchronizer
+     * @param ScopeTypeNormalizer $scopeTypeNormalizer
      */
     public function __construct(
         StatusServiceInterface $companyStatusService,
@@ -75,7 +84,8 @@ public function __construct(
         CatalogPermissionManagement $catalogPermissionsManagement,
         ConfigResource $configResource,
         ScopeResolverPool $scopeResolverPool,
-        Synchronizer $permissionsSynchronizer
+        Synchronizer $permissionsSynchronizer,
+        ScopeTypeNormalizer $scopeTypeNormalizer
     ) {
         $this->companyStatusService = $companyStatusService;
         $this->sharedCatalogModuleConfig = $sharedCatalogModuleConfig;
@@ -84,6 +94,7 @@ public function __construct(
         $this->configResource = $configResource;
         $this->scopeResolverPool = $scopeResolverPool;
         $this->permissionsSynchronizer = $permissionsSynchronizer;
+        $this->scopeTypeNormalizer = $scopeTypeNormalizer;
     }

     /**
@@ -114,7 +125,7 @@ public function execute(Observer $observer)
             $this->configResource->saveConfig(
                 SharedCatalogModuleConfig::CONFIG_SHARED_CATALOG,
                 0,
-                $scope->getScopeType(),
+                $this->scopeTypeNormalizer->normalize($scope->getScopeType()),
                 $scope->getId()
             );

