diff --git a/vendor/magento/module-admin-gws/Plugin/ConfigurableProduct/Block/Adminhtml/Product/Steps/SelectAttributes.php b/vendor/magento/module-admin-gws/Plugin/ConfigurableProduct/Block/Adminhtml/Product/Steps/SelectAttributes.php
new file mode 100644
index 000000000000..11bd8c3934db
--- /dev/null
+++ b/vendor/magento/module-admin-gws/Plugin/ConfigurableProduct/Block/Adminhtml/Product/Steps/SelectAttributes.php
@@ -0,0 +1,61 @@
+<?php
+/**
+ * ADOBE CONFIDENTIAL
+ *
+ * Copyright 2025 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ */
+declare(strict_types=1);
+
+namespace Magento\AdminGws\Plugin\ConfigurableProduct\Block\Adminhtml\Product\Steps;
+
+use Magento\AdminGws\Model\Role;
+use Magento\ConfigurableProduct\Block\Adminhtml\Product\Steps\SelectAttributes as SelectAttributesBlock;
+
+/**
+ * Check if certain admin is allowed to create new product attribute.
+ */
+class SelectAttributes
+{
+    /**
+     * @var Role
+     */
+    private $role;
+
+    /**
+     * @param Role $role
+     */
+    public function __construct(Role $role)
+    {
+        $this->role = $role;
+    }
+
+    /**
+     * Around plugin to determine if admin is allowed to create new product attribute.
+     *
+     * @param SelectAttributesBlock $subject
+     * @param \Closure $proceed
+     * @param string $dataProvider
+     * @return string
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundGetAddNewAttributeButton(
+        SelectAttributesBlock $subject,
+        \Closure $proceed,
+        $dataProvider = ''
+    ): string {
+        if (!$this->role->getIsAll()) {
+            return '';
+        }
+        return $proceed($dataProvider);
+    }
+}
diff --git a/vendor/magento/module-admin-gws/etc/adminhtml/di.xml b/vendor/magento/module-admin-gws/etc/adminhtml/di.xml
index 9075bc3b1c2c..bcf4f9b6373d 100644
--- a/vendor/magento/module-admin-gws/etc/adminhtml/di.xml
+++ b/vendor/magento/module-admin-gws/etc/adminhtml/di.xml
@@ -122,4 +122,7 @@
     <type name="Magento\Store\Model\StoreSwitcher\ManageStoreCookie">
         <plugin name="store_switcher_manage_store_cookie_plugin" type="Magento\AdminGws\Model\Plugin\Store\StoreSwitcher\ManageStoreCookie"/>
     </type>
+    <type name="Magento\ConfigurableProduct\Block\Adminhtml\Product\Steps\SelectAttributes">
+        <plugin name="is_customer_allowed_to_create_new_attribute" type="Magento\AdminGws\Plugin\ConfigurableProduct\Block\Adminhtml\Product\Steps\SelectAttributes"/>
+    </type>
 </config>
