diff --git a/vendor/magento/module-catalog-rule-staging/Model/CatalogRuleApplier.php b/vendor/magento/module-catalog-rule-staging/Model/CatalogRuleApplier.php
index e62cd775e8b3..8ee4258caa18 100644
--- a/vendor/magento/module-catalog-rule-staging/Model/CatalogRuleApplier.php
+++ b/vendor/magento/module-catalog-rule-staging/Model/CatalogRuleApplier.php
@@ -9,10 +9,8 @@
 use Magento\Staging\Model\StagingApplierInterface;
 use Magento\Framework\App\ObjectManager;
 use Magento\Framework\Indexer\IndexerRegistry;
+use Magento\Catalog\Model\Indexer\Product\Price\Processor as PriceProcessor;
 
-/**
- * Class CatalogRuleApplier
- */
 class CatalogRuleApplier implements StagingApplierInterface
 {
     /**
@@ -36,11 +34,12 @@ public function __construct(
         IndexerRegistry $indexerRegistry = null
     ) {
         $this->ruleProductProcessor = $ruleProductProcessor;
-        $this->indexerRegistry = $indexerRegistry
-            ?: ObjectManager::getInstance()->get(IndexerRegistry::class);
+        $this->indexerRegistry = $indexerRegistry ?: ObjectManager::getInstance()->get(IndexerRegistry::class);
     }
 
     /**
+     * Invalidate related indexers
+     *
      * @param array $entityIds
      * @return void
      */
@@ -48,10 +47,7 @@ public function execute(array $entityIds)
     {
         if (!empty($entityIds)) {
             $this->ruleProductProcessor->markIndexerAsInvalid();
-            $this->indexerRegistry->get(\Magento\CatalogRule\Model\Indexer\Product\ProductRuleProcessor::INDEXER_ID)
-                ->invalidate();
-            $this->indexerRegistry->get(\Magento\Catalog\Model\Indexer\Product\Price\Processor::INDEXER_ID)
-                ->invalidate();
+            $this->indexerRegistry->get(PriceProcessor::INDEXER_ID)->invalidate();
         }
     }
 }
