diff --git a/vendor/magento/module-visual-merchandiser/Model/Product/DataProvider.php b/vendor/magento/module-visual-merchandiser/Model/Product/DataProvider.php
index 78a5b3530617..b266f4c6df6c 100755
--- a/vendor/magento/module-visual-merchandiser/Model/Product/DataProvider.php
+++ b/vendor/magento/module-visual-merchandiser/Model/Product/DataProvider.php
@@ -99,11 +99,6 @@ public function __construct(
         $this->collection->setStoreId($this->request->getParam('store_id', "0"));
         $this->collection = $this->quantityAndStock->joinStock($this->collection);
 
-        $this->collection->getSelect()->group('e.entity_id');
-        $subSelect = clone($this->collection->getSelect());
-        $this->collection->getSelect()->reset()->from(['e' => $subSelect]);
-        $this->collection->addStaticField('stock');
-
         $this->prepareUpdateUrl();
     }
 
