diff --git a/vendor/magento/module-company/Model/Authorization/PermissionProvider.php b/vendor/magento/module-company/Model/Authorization/PermissionProvider.php
index 7e73896917e5..079c4c84d624 100755
--- a/vendor/magento/module-company/Model/Authorization/PermissionProvider.php
+++ b/vendor/magento/module-company/Model/Authorization/PermissionProvider.php
@@ -5,32 +5,32 @@
  */
 namespace Magento\Company\Model\Authorization;
 
-/**
- * Class PermissionProvider.
- */
+use Magento\Company\Model\ResourceModel\Permission\CollectionFactory;
+use Magento\Company\Model\ResourcePool;
+
 class PermissionProvider
 {
     /**
-     * \Magento\Company\Model\ResourceModel\Permission\Collection
+     * @var CollectionFactory
      */
-    private $permissionCollection;
+    private $permissionCollectionFactory;
 
     /**
-     * @var \Magento\Company\Model\ResourcePool
+     * @var ResourcePool
      */
     private $resourcePool;
 
     /**
      * PermissionProvider constructor.
      *
-     * @param \Magento\Company\Model\ResourceModel\Permission\Collection $permissionCollection
-     * @param \Magento\Company\Model\ResourcePool $resourcePool
+     * @param CollectionFactory $permissionCollectionFactory
+     * @param ResourcePool $resourcePool
      */
     public function __construct(
-        \Magento\Company\Model\ResourceModel\Permission\Collection $permissionCollection,
-        \Magento\Company\Model\ResourcePool $resourcePool
+        CollectionFactory $permissionCollectionFactory,
+        ResourcePool      $resourcePool
     ) {
-        $this->permissionCollection = $permissionCollection;
+        $this->permissionCollectionFactory = $permissionCollectionFactory;
         $this->resourcePool = $resourcePool;
     }
 
@@ -42,7 +42,7 @@ public function __construct(
      */
     public function retrieveRolePermissions($roleId)
     {
-        return $this->permissionCollection
+        return $this->permissionCollectionFactory->create()
             ->addFieldToFilter('role_id', ['eq' => $roleId])
             ->toOptionHash('resource_id', 'permission');
     }
