diff --git a/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php b/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
index e5e5e44014cc..b6e67ec6ed61 100755
--- a/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
+++ b/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
@@ -81,6 +81,7 @@ public function execute(int $customerId, array $args): array
         $countLists = $this->repository->getList($builderCount->create())->getTotalCount();
 
         $builder = $this->getFilters($customerId, false, $args);
+        $builder->setPageSize($pageSize)->setCurrentPage($currentPage);
         $lists = $this->repository->getList($builder->create())->getItems();
 
         $data = [
@@ -145,7 +146,7 @@ private function getFilters(int $customerId, bool $isTotal, array $args = [])
                     } else {
                         $reqIds = $this->idEncoder->decode($filtersGiven['uids'][key($filtersGiven['uids'])]);
                     }
-                    
+
                     $filters[] = $this->filterBuilder
                         ->setField(RequisitionListInterface::REQUISITION_LIST_ID)
                         ->setConditionType(key($filtersGiven['uids']))
