diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml b/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml
index 7824cbd91186..6b2b4869c893 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/layout/customer_account_edit.xml
@@ -23,6 +23,7 @@
                     <argument name="code" xsi:type="string">Magento\Customer\Model\Customer</argument>
                 </action>
                 <arguments>
+                    <argument name="button_lock_manager" xsi:type="object">Magento\Framework\View\Element\ButtonLockManager</argument>
                     <argument name="view_model" xsi:type="object">CustomerFileAttribute</argument>
                 </arguments>
             </block>
diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml
index 7d2543c73854..97bb8fe08912 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/edit.phtml
@@ -129,7 +129,10 @@ use Magento\Customer\Block\Widget\Taxvat;
 
     <div class="actions-toolbar">
         <div class="primary">
-            <button type="submit" class="action save primary" title="<?= $block->escapeHtmlAttr(__('Save')) ?>">
+            <button type="submit" class="action save primary" title="<?= $block->escapeHtmlAttr(__('Save')) ?>"
+                <?php if ($block->getButtonLockManager()->isDisabled('customer_edit_form_submit')): ?>
+                    disabled="disabled"
+                <?php endif; ?>>
                 <span><?= $block->escapeHtml(__('Save')) ?></span>
             </button>
         </div>
diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
index 70d10a3f1d16..30a495f6676a 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
@@ -25,8 +25,6 @@ $directoryHelper = $block->getData('directoryHelper');
 $formData = $block->getFormData();
 /** @var \Magento\CustomerCustomAttributes\ViewModel\Address\RegionProvider $regionProvider */
 $regionProvider = $block->getRegionProvider();
-/** @var \Magento\Customer\ViewModel\CreateAccountButton $createAccountButtonViewModel */
-$createAccountButtonViewModel = $block->getData('create_account_button_view_model');
 ?>
 <form class="form create account form-create-account"
       action="<?= $block->escapeUrl($block->getPostActionUrl()) ?>"
@@ -322,7 +320,9 @@ $createAccountButtonViewModel = $block->getData('create_account_button_view_mode
                     class="action submit primary"
                     title="<?= $block->escapeHtmlAttr(__('Create an Account')) ?>"
                     id="send2"
-                <?php if ($createAccountButtonViewModel->disabled()): ?> disabled="disabled" <?php endif; ?>>
+                <?php if ($block->getButtonLockManager()->isDisabled('customer_create_form_submit')): ?>
+                    disabled="disabled"
+                <?php endif; ?>>
                 <span><?= $block->escapeHtml(__('Create an Account')) ?></span>
             </button>
         </div>
diff --git a/vendor/magento/module-invitation/view/frontend/layout/magento_invitation_customer_account_create.xml b/vendor/magento/module-invitation/view/frontend/layout/magento_invitation_customer_account_create.xml
index 6b7bba3fc6f9..00549d670255 100644
--- a/vendor/magento/module-invitation/view/frontend/layout/magento_invitation_customer_account_create.xml
+++ b/vendor/magento/module-invitation/view/frontend/layout/magento_invitation_customer_account_create.xml
@@ -12,7 +12,7 @@
             <block class="Magento\Invitation\Block\Customer\Form\Register" name="customer_form_register_invitation" template="Magento_Customer::form/register.phtml" cacheable="false">
                 <container name="form.additional.info" as="form_additional_info"/>
                 <arguments>
-                    <argument name="create_account_button_view_model" xsi:type="object">Magento\Customer\ViewModel\CreateAccountButton</argument>
+                    <argument name="button_lock_manager" xsi:type="object">Magento\Framework\View\Element\ButtonLockManager</argument>
                 </arguments>
             </block>
             <block class="Magento\Cookie\Block\RequireCookie" name="require-cookie" template="Magento_Cookie::require_cookie.phtml">
