diff --git a/vendor/magento/module-company/Controller/Account/CreatePost.php b/vendor/magento/module-company/Controller/Account/CreatePost.php
index 98c9befe59bd..aac210f0642c 100644
--- a/vendor/magento/module-company/Controller/Account/CreatePost.php
+++ b/vendor/magento/module-company/Controller/Account/CreatePost.php
@@ -25,7 +25,6 @@
 use Magento\Company\Model\CompanyUser;
 use Magento\Company\Model\Create\Session;
 use Magento\Company\Model\Customer\Company;
-use Magento\Company\Model\Customer\IsSuperUser;
 use Magento\Customer\Api\AccountManagementInterface;
 use Magento\Customer\Api\CustomerRepositoryInterface;
 use Magento\Customer\Api\Data\CustomerInterface;
@@ -43,6 +42,7 @@
 use Magento\Framework\Exception\NoSuchEntityException;
 use Magento\Company\Model\Customer\AccountManagement\CompanyRequestHydrator;
 use Psr\Log\LoggerInterface;
+use Magento\Framework\App\Http\Context as HttpContext;
 
 /**
  * Create company account action.
@@ -118,9 +118,9 @@ class CreatePost extends Action implements HttpPostActionInterface
     private $customerRepository;
 
     /**
-     * @var IsSuperUser
+     * @var HttpContext
      */
-    private $isSuperUser;
+    private $httpContext;
 
     /**
      * @var CompanyRequestHydrator
@@ -141,8 +141,8 @@ class CreatePost extends Action implements HttpPostActionInterface
      * @param CustomerExtractor|null $customerExtractor
      * @param Company|null $customerCompany
      * @param CustomerRepositoryInterface|null $customerRepository
-     * @param IsSuperUser|null $isSuperUser
      * @param CompanyRequestHydrator|null $companyRequestHydrator
+     * @param HttpContext|null $httpContext
      */
     public function __construct(
         Context                               $context,
@@ -158,8 +158,8 @@ public function __construct(
         ?CustomerExtractor                    $customerExtractor = null,
         ?Company                              $customerCompany = null,
         ?CustomerRepositoryInterface          $customerRepository = null,
-        ?IsSuperUser                          $isSuperUser = null,
-        ?CompanyRequestHydrator               $companyRequestHydrator = null
+        ?CompanyRequestHydrator               $companyRequestHydrator = null,
+        ?HttpContext                          $httpContext = null
     ) {
         parent::__construct($context);
         $this->userContext = $userContext;
@@ -178,10 +178,10 @@ public function __construct(
             ?: ObjectManager::getInstance()->get(Company::class);
         $this->customerRepository = $customerRepository
             ?: ObjectManager::getInstance()->get(CustomerRepositoryInterface::class);
-        $this->isSuperUser = $isSuperUser
-            ?: ObjectManager::getInstance()->get(IsSuperUser::class);
         $this->companyRequestHydrator = $companyRequestHydrator
             ?: ObjectManager::getInstance()->get(CompanyRequestHydrator::class);
+        $this->httpContext = $httpContext
+            ?: ObjectManager::getInstance()->get(HttpContext::class);
     }
 
     /**
@@ -240,6 +240,8 @@ public function execute()
      */
     private function createCompanyAndReturnSuperUser(CustomerInterface $customerData): CustomerInterface
     {
+        $this->httpContext->unsValue(\Magento\Company\Model\CompanyContextInterface::CONTEXT_COMPANY_ID);
+
         try {
             $customer = $this->customerRepository->get($customerData->getEmail());
         } catch (NoSuchEntityException $exception) {

