diff --git a/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php b/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php
index 6390bc8a00c..7677ea811d1 100755
--- a/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php
+++ b/vendor/magento/module-admin-gws/Plugin/CollectionFilter.php
@@ -100,6 +100,9 @@ class CollectionFilter
             if (method_exists($collection, 'addStoresFilter')) {
                 $collection->addStoresFilter($this->getStoreIds());
                 $collection->setFlag(self::FILTERED_FLAG_NAME, true);
+            } elseif (method_exists($collection, 'addStoreFilter')) {
+                $collection->addStoreFilter($this->getStoreIds());
+                $collection->setFlag(self::FILTERED_FLAG_NAME, true);
             } elseif (isset($collection->getSelect()->getPart(Select::FROM)['main_table'])) {
                 $this->tableBasedFilter($collection);
                 $collection->setFlag(self::FILTERED_FLAG_NAME, true);
