diff --git a/vendor/magento/module-negotiable-quote/Model/QuoteUpdatesInfo.php b/vendor/magento/module-negotiable-quote/Model/QuoteUpdatesInfo.php
index 886d17218ba8..8c32900b6bb7 100644
--- a/vendor/magento/module-negotiable-quote/Model/QuoteUpdatesInfo.php
+++ b/vendor/magento/module-negotiable-quote/Model/QuoteUpdatesInfo.php
@@ -229,7 +229,7 @@ private function getAllUpdatedItems(\Magento\Quote\Api\Data\CartInterface $quote
                 'url' => $this->getProductUrlByItem($item),
                 'sku' => $item->getProduct()->getSku(),
                 'cost' => $this->formatPrice($quote, $this->getQuoteTotals($quote)->getItemCost($item)),
-                'stock' => number_format($this->quoteHelper->getStockForProduct($item)),
+                'stock' => number_format($this->quoteHelper->getStockForProduct($item) ?? 0),
                 'qty' => $item->getQty(),
                 'subtotal' => $this->formatPrice($quote, $item->getBaseRowTotal() - $item->getBaseDiscountAmount()),
                 'tax' => $this->formatPrice($quote, $item->getBaseTaxAmount()),
diff --git a/vendor/magento/module-negotiable-quote/view/adminhtml/templates/quote/view/items/grid.phtml b/vendor/magento/module-negotiable-quote/view/adminhtml/templates/quote/view/items/grid.phtml
index f224ca070f1b..3f9bc027e915 100644
--- a/vendor/magento/module-negotiable-quote/view/adminhtml/templates/quote/view/items/grid.phtml
+++ b/vendor/magento/module-negotiable-quote/view/adminhtml/templates/quote/view/items/grid.phtml
@@ -142,7 +142,8 @@ $disabled = !$block->canEdit() ? '_disabled' : '';
                     </div>
                 </td>
                 <td class="col-stock">
-                    <?= /* @noEscape */ number_format($quoteHelper->getStockForProduct($item)) ?>
+                    <?= /* @noEscape */ $quoteHelper->getStockForProduct($item) !== null ?
+                        number_format($quoteHelper->getStockForProduct($item)) : 0 ?>
                 </td>
                 <td class="col-cost">
                     <?= /* @noEscape */ $block->getFormattedCost($item) ?>
diff --git a/vendor/magento/module-negotiable-quote-weee/view/adminhtml/templates/quote/view/items/grid.phtml b/vendor/magento/module-negotiable-quote-weee/view/adminhtml/templates/quote/view/items/grid.phtml
index 256ec85b4b76..1088ad553e4f 100644
--- a/vendor/magento/module-negotiable-quote-weee/view/adminhtml/templates/quote/view/items/grid.phtml
+++ b/vendor/magento/module-negotiable-quote-weee/view/adminhtml/templates/quote/view/items/grid.phtml
@@ -148,7 +148,10 @@ $disabled = !$block->canEdit() ? '_disabled' : '';
                         <?php endif; ?>
                     </div>
                 </td>
-                <td class="col-stock"><?= /* @noEscape */ number_format($quoteHelper->getStockForProduct($item)) ?></td>
+                <td class="col-stock">
+                    <?= /* @noEscape */ $quoteHelper->getStockForProduct($item) !== null ?
+                        number_format($quoteHelper->getStockForProduct($item)) : 0 ?>
+                </td>
                 <td class="col-cost"><?= /* @noEscape */ $block->getFormattedCost($item) ?></td>
                 <td class="col-price"><?= /* @noEscape */ $block->getFormattedCatalogPrice($item) ?></td>
                 <td class="col-cart-price"><?= /* @noEscape */ $block->getFormattedCartPrice($item) ?></td>
