diff --git a/vendor/magento/module-catalog-staging/Observer/UpdateProductDateAttributes.php b/vendor/magento/module-catalog-staging/Observer/UpdateProductDateAttributes.php
index bce6f35835e7..c4511b3e092b 100644
--- a/vendor/magento/module-catalog-staging/Observer/UpdateProductDateAttributes.php
+++ b/vendor/magento/module-catalog-staging/Observer/UpdateProductDateAttributes.php
@@ -10,8 +10,11 @@
 use Magento\Catalog\Model\Attribute\ScopeOverriddenValue;
 use Magento\Catalog\Model\Product;
 use Magento\CatalogStaging\Model\Product\DateAttributesMetadata;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\App\ScopeInterface;
 use Magento\Framework\Event\Observer;
 use Magento\Framework\Event\ObserverInterface;
+use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\Stdlib\DateTime;
 use Magento\Framework\Stdlib\DateTime\TimezoneInterface;
 use Magento\Store\Model\Store;
@@ -32,7 +35,7 @@ class UpdateProductDateAttributes implements ObserverInterface
     public function __construct(
         private TimezoneInterface $localeDate,
         private DateAttributesMetadata $dateAttributesMetadata,
-        private ScopeOverriddenValue $scopeOverriddenValue,
+        private ScopeOverriddenValue $scopeOverriddenValue
     ) {
     }
 
@@ -45,6 +48,7 @@ public function __construct(
      *
      * @param Observer $observer
      * @return void
+     * @throws LocalizedException
      */
     public function execute(Observer $observer)
     {
@@ -53,7 +57,11 @@ public function execute(Observer $observer)
         $this->updateAttributes(
             $product,
             $this->dateAttributesMetadata->getStartDateAttributes(),
-            $this->localeDate->date()->format(DateTime::DATETIME_PHP_FORMAT),
+            $this->localeDate->scopeDate(
+                ScopeInterface::SCOPE_DEFAULT,
+                null,
+                true
+            )->format(DateTime::DATETIME_PHP_FORMAT),
         );
         $this->updateAttributes(
             $product,
