diff --git a/vendor/magento/module-company-graph-ql/Model/Resolver/UpdateCompanyRole.php b/vendor/magento/module-company-graph-ql/Model/Resolver/UpdateCompanyRole.php
index 5d52f53758..35a1c5c64d 100644
--- a/vendor/magento/module-company-graph-ql/Model/Resolver/UpdateCompanyRole.php
+++ b/vendor/magento/module-company-graph-ql/Model/Resolver/UpdateCompanyRole.php
@@ -13,6 +13,7 @@ use Magento\CompanyGraphQl\Model\Company\Role\PermissionsFormatter;
 use Magento\CompanyGraphQl\Model\Company\Role\UpdateRole;
 use Magento\Framework\GraphQl\Config\Element\Field;
 use Magento\Framework\GraphQl\Exception\GraphQlInputException;
+use Magento\Framework\GraphQl\Exception\GraphQlAuthenticationException;
 use Magento\Framework\GraphQl\Query\Resolver\ContextInterface;
 use Magento\Framework\GraphQl\Query\ResolverInterface;
 use Magento\Framework\GraphQl\Query\Uid;
@@ -84,6 +85,12 @@ class UpdateCompanyRole implements ResolverInterface
      */
     public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
     {
+        if (false === $context->getExtensionAttributes()->getIsCustomer()) {
+            throw new GraphQlAuthenticationException(
+                __('The current customer isn\'t authorized.')
+            );
+        }
+
         $this->resolverAccess->isAllowed($this->allowedResources);
 
         if (empty($args['input']) || !is_array($args['input'])) {
