diff --git a/vendor/magento/module-customer-custom-attributes/Plugin/Catalog/UpdateMultiselectAttributes.php b/vendor/magento/module-customer-custom-attributes/Plugin/Catalog/UpdateMultiselectAttributes.php
deleted file mode 100644
index 7ce34b6a894..00000000000
--- a/vendor/magento/module-customer-custom-attributes/Plugin/Catalog/UpdateMultiselectAttributes.php
+++ /dev/null
@@ -1,39 +0,0 @@
-<?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
- */
-declare(strict_types=1);
-
-namespace Magento\CustomerCustomAttributes\Plugin\Catalog;
-
-use Magento\Catalog\Api\Data\ProductInterface;
-use Magento\Catalog\Controller\Adminhtml\Product\Initialization\Helper;
-
-/**
- * Updates multiselect attributes for product data
- */
-class UpdateMultiselectAttributes
-{
-    /**
-     * Update empty multiselect attributes for product data
-     *
-     * @param Helper $subject
-     * @param ProductInterface $product
-     * @return ProductInterface
-     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     */
-    public function afterInitializeFromData(Helper $subject, ProductInterface $product): ProductInterface
-    {
-        $productData = $product->getData();
-        $attributes = $product->getAttributes();
-        foreach ($attributes as $attrKey => $attribute) {
-            if ($attribute->getFrontendInput() === 'multiselect') {
-                if (array_key_exists($attrKey, $productData) && $productData[$attrKey] == null) {
-                    $product->setData($attrKey, '');
-                }
-            }
-        }
-        return $product;
-    }
-}
diff --git a/vendor/magento/module-customer-custom-attributes/etc/adminhtml/di.xml b/vendor/magento/module-customer-custom-attributes/etc/adminhtml/di.xml
index 4f2c3c80cdf..1ae00b2ce28 100644
--- a/vendor/magento/module-customer-custom-attributes/etc/adminhtml/di.xml
+++ b/vendor/magento/module-customer-custom-attributes/etc/adminhtml/di.xml
@@ -62,7 +62,4 @@
             </argument>
         </arguments>
     </type>
-    <type name="Magento\Catalog\Controller\Adminhtml\Product\Initialization\Helper">
-        <plugin name="update_multiselect_attributes" type="Magento\CustomerCustomAttributes\Plugin\Catalog\UpdateMultiselectAttributes"/>
-    </type>
 </config>
