diff --git a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
index 56bc725c4dd2..41a4e14dabd3 100644
--- a/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
+++ b/vendor/magento/module-customer-custom-attributes/view/frontend/templates/customer/form/register.phtml
@@ -25,6 +25,8 @@ $directoryHelper = $block->getData('directoryHelper');
 $formData = $block->getFormData();
 /** @var \Magento\CustomerCustomAttributes\ViewModel\Address\RegionProvider $regionProvider */
 $regionProvider = $block->getRegionProvider();
+/** @var \Magento\Customer\ViewModel\CreateAccountButton $createAccountButtonViewModel */
+$createAccountButtonViewModel = $block->getData('create_account_button_view_model');
 ?>
 <form class="form create account form-create-account"
       action="<?= $block->escapeUrl($block->getPostActionUrl()) ?>"
@@ -318,7 +320,9 @@ $regionProvider = $block->getRegionProvider();
         <div class="primary">
             <button type="submit"
                     class="action submit primary"
-                    title="<?= $block->escapeHtmlAttr(__('Create an Account')) ?>">
+                    title="<?= $block->escapeHtmlAttr(__('Create an Account')) ?>"
+                    id="send2"
+                <?php if ($createAccountButtonViewModel->disabled()): ?> disabled="disabled" <?php endif; ?>>
                 <span><?= $block->escapeHtml(__('Create an Account')) ?></span>
             </button>
         </div>
@@ -334,6 +338,7 @@ $regionProvider = $block->getRegionProvider();
     <?php
     $displayAllRegions = $block->getConfig("general/region/display_all") ? "true" : "false";
     ?>
+<?php // phpcs:ignore Magento2.Legacy.PhtmlTemplate ?>
 <script type="text/x-magento-init">
     {
         "#country": {
@@ -351,6 +356,7 @@ $regionProvider = $block->getRegionProvider();
     }
 </script>
 <?php endif; ?>
+<?php // phpcs:ignore Magento2.Legacy.PhtmlTemplate ?>
 <script type="text/x-magento-init">
     {
         "*": {
