diff --git a/vendor/magento/module-gift-registry/Block/View.php b/vendor/magento/module-gift-registry/Block/View.php
index 58e77dc21f0d..04159bc2a531 100644
--- a/vendor/magento/module-gift-registry/Block/View.php
+++ b/vendor/magento/module-gift-registry/Block/View.php
@@ -1,7 +1,21 @@
 <?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+/************************************************************************
+ *
+ * ADOBE CONFIDENTIAL
+ * ___________________
+ *
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ * ************************************************************************
  */
 namespace Magento\GiftRegistry\Block;

@@ -68,7 +82,7 @@ public function getEntity()
     public function getFormattedDate($date)
     {
         if ($date) {
-            return $this->formatDate($date, \IntlDateFormatter::MEDIUM);
+            return $this->formatDate($date, \IntlDateFormatter::MEDIUM, false, 'UTC');
         }
         return '';
     }
