diff --git a/vendor/magento/module-company-graph-ql/Model/Company/Structure.php b/vendor/magento/module-company-graph-ql/Model/Company/Structure.php
index 6f9cbc215c..d72fc49fa8 100644
--- a/vendor/magento/module-company-graph-ql/Model/Company/Structure.php
+++ b/vendor/magento/module-company-graph-ql/Model/Company/Structure.php
@@ -163,6 +163,7 @@ class Structure
                 'type' => HierarchyInterface::TYPE_TEAM,
                 'id' => $this->idEncoder->encode($tree->getData(TeamInterface::TEAM_ID)),
                 'name' => $tree->getData(TeamInterface::NAME),
+                'structure_id' => $this->idEncoder->encode($tree->getData(StructureInterface::STRUCTURE_ID)),
                 'description' => $tree->getData(TeamInterface::DESCRIPTION)
             ];
         }
