diff --git a/vendor/magento/module-customer-segment/Model/Customer.php b/vendor/magento/module-customer-segment/Model/Customer.php
index 9ac45f6facc8..5a115fc97cd7 100644
--- a/vendor/magento/module-customer-segment/Model/Customer.php
+++ b/vendor/magento/module-customer-segment/Model/Customer.php
@@ -349,7 +349,7 @@ public function addCustomerToWebsiteSegments(int $customerId, int $websiteId, ar
         );
 
         $value = array_filter($this->_customerWebsiteSegments[$websiteId][$customerId]);
-        $this->_httpContext->setValue(Data::CONTEXT_SEGMENT, $value, $value);
+        $this->_httpContext->setValue(Data::CONTEXT_SEGMENT, $value, []);
 
         $visitorCustomerSegmentIds = $this->_customerSession->getCustomerSegmentIds();
 
