diff --git a/vendor/magento/module-elasticsearch-catalog-permissions/Model/Adapter/FieldMapper/Product/FieldProvider/FieldName/Resolver/CategoryPermission.php b/vendor/magento/module-elasticsearch-catalog-permissions/Model/Adapter/FieldMapper/Product/FieldProvider/FieldName/Resolver/CategoryPermission.php
index 32995f440474..145d1c3d0dea 100644
--- a/vendor/magento/module-elasticsearch-catalog-permissions/Model/Adapter/FieldMapper/Product/FieldProvider/FieldName/Resolver/CategoryPermission.php
+++ b/vendor/magento/module-elasticsearch-catalog-permissions/Model/Adapter/FieldMapper/Product/FieldProvider/FieldName/Resolver/CategoryPermission.php
@@ -21,7 +21,7 @@ class CategoryPermission implements ResolverInterface
     public function getFieldName(AttributeAdapter $attribute, $context = []): ?string
     {
         if ($attribute->getAttributeCode() === 'category_permission') {
-            return "category_permission_${context['storeId']}_${context['customerGroupId']}";
+            return "category_permission_{$context['storeId']}_{$context['customerGroupId']}";
         }
 
         return null;
