diff --git a/vendor/magento/module-multicoupon-ui/Observer/ResetShippingDataObserver.php b/vendor/magento/module-multicoupon-ui/Observer/ResetShippingDataObserver.php
new file mode 100644
index 000000000000..eb93c2e7965a
--- /dev/null
+++ b/vendor/magento/module-multicoupon-ui/Observer/ResetShippingDataObserver.php
@@ -0,0 +1,46 @@
+<?php
+/**
+ * Copyright 2025 Adobe
+ * All Rights Reserved.
+ */
+declare(strict_types=1);
+
+namespace Magento\MulticouponUi\Observer;
+
+use Magento\Framework\Event\Observer;
+use Magento\Multicoupon\Model\Config\Config;
+use Magento\Framework\Event\ObserverInterface;
+
+/**
+ * Class for resetting shipping method for multi coupons.
+ */
+class ResetShippingDataObserver implements ObserverInterface
+{
+    /**
+     * @param Config $multiCouponConfig
+     */
+    public function __construct(private readonly Config $multiCouponConfig)
+    {
+    }
+
+    /**
+     * Check if config have multi coupons enabled and reset the shipping method if needed.
+     *
+     * @param Observer $observer
+     * @return $this
+     */
+    public function execute(Observer $observer)
+    {
+        $request = $observer->getEvent()->getRequest();
+        if (array_key_exists('order', $request)) {
+            $isResetShippingFlag = array_key_exists('reset_shipping', $request)
+                && (bool) $request['reset_shipping'];
+            $noOfCouponsAvailable = $this->multiCouponConfig->getMaximumNumberOfCoupons();
+            if ($noOfCouponsAvailable > 1 && $isResetShippingFlag) {
+                    $observer->getEvent()->getOrderCreateModel()
+                        ->resetShippingMethod(true);
+            }
+        }
+        return $this;
+    }
+}
diff --git a/vendor/magento/module-multicoupon-ui/etc/adminhtml/events.xml b/vendor/magento/module-multicoupon-ui/etc/adminhtml/events.xml
new file mode 100644
index 000000000000..e6ddfb92ad0d
--- /dev/null
+++ b/vendor/magento/module-multicoupon-ui/etc/adminhtml/events.xml
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
+<!--
+/**
+ * Copyright 2025 Adobe
+ * All rights reserved.
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
+    <event name="adminhtml_sales_order_create_process_data">
+        <observer name="magento_multi_coupon_shipping_method" instance="Magento\MulticouponUi\Observer\ResetShippingDataObserver"/>
+    </event>
+</config>
