diff --git a/vendor/magento/module-banner/Model/Banner/Data.php b/vendor/magento/module-banner/Model/Banner/Data.php
index 271fb8d5befc..cbc136116d3f 100644
--- a/vendor/magento/module-banner/Model/Banner/Data.php
+++ b/vendor/magento/module-banner/Model/Banner/Data.php
@@ -290,7 +290,7 @@ private function getBannerIdsByCatalogRules()
     private function getProductRelatedBannerIds(int $productId): array
     {
         $result = $this->catalogRule->getRulesFromProduct(
-            $this->dateTime->scopeDate($this->storeManager->getStore()->getId()),
+            $this->dateTime->scopeDate($this->storeManager->getStore()->getId(), includeTime: true)->getTimestamp(),
             $this->storeManager->getWebsite()->getId(),
             $this->httpContext->getValue(\Magento\Customer\Model\Context::CONTEXT_GROUP),
             $productId
