diff --git a/vendor/magento/module-negotiable-quote/Controller/Adminhtml/Quote/Update.php b/vendor/magento/module-negotiable-quote/Controller/Adminhtml/Quote/Update.php
index 2f870f27b51a..780fd39edf76 100644
--- a/vendor/magento/module-negotiable-quote/Controller/Adminhtml/Quote/Update.php
+++ b/vendor/magento/module-negotiable-quote/Controller/Adminhtml/Quote/Update.php
@@ -94,9 +94,10 @@ public function execute()
         }

         $data = $this->getQuoteData();
+
         /** @var \Magento\Framework\Controller\Result\Json $response */
         $response = $this->getResultJson();
-        $response->setJsonData(json_encode($data, JSON_NUMERIC_CHECK));
+        $response->setJsonData(json_encode($data));
         return $response;
     }

