diff --git a/vendor/magento/module-catalog-staging/Plugin/Catalog/Model/Category/DataProvider/TrimData.php b/vendor/magento/module-catalog-staging/Plugin/Catalog/Model/Category/DataProvider/TrimData.php
deleted file mode 100644
index 8a0b9f21e5f..00000000000
--- a/vendor/magento/module-catalog-staging/Plugin/Catalog/Model/Category/DataProvider/TrimData.php
+++ /dev/null
@@ -1,24 +0,0 @@
-<?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
- */
-namespace Magento\CatalogStaging\Plugin\Catalog\Model\Category\DataProvider;
-
-class TrimData
-{
-    /**
-     * @param \Magento\Catalog\Model\Category\DataProvider $subject
-     * @param array $data
-     * @return array
-     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     */
-    public function afterGetData(\Magento\Catalog\Model\Category\DataProvider $subject, array $data)
-    {
-        foreach ($data as &$categoryData) {
-            unset($categoryData['updated_in']);
-            unset($categoryData['created_in']);
-        }
-        return $data;
-    }
-}
diff --git a/vendor/magento/module-catalog-staging/etc/di.xml b/vendor/magento/module-catalog-staging/etc/di.xml
index 2c88e223216..2476f260e3d 100644
--- a/vendor/magento/module-catalog-staging/etc/di.xml
+++ b/vendor/magento/module-catalog-staging/etc/di.xml
@@ -333,9 +333,6 @@
     <type name="Magento\Framework\Pricing\Render\PriceBox">
         <plugin name="price_box_staging_plugin" type="Magento\CatalogStaging\Plugin\Catalog\Pricing\Render\PriceBox" />
     </type>
-    <type name="Magento\Catalog\Model\Category\DataProvider">
-        <plugin name="catalogCategoryDataProviderTrimData" type="Magento\CatalogStaging\Plugin\Catalog\Model\Category\DataProvider\TrimData" />
-    </type>
     <type name="Magento\Staging\Model\Entity\Builder">
         <arguments>
             <argument name="strategies" xsi:type="array">
