diff --git a/vendor/magento/module-catalog-permissions-graph-ql/Model/Resolver/Categories/DataProvider/Category/CollectionProcessor/ApplyPermissionsOnCategory.php b/vendor/magento/module-catalog-permissions-graph-ql/Model/Resolver/Categories/DataProvider/Category/CollectionProcessor/ApplyPermissionsOnCategory.php
index 5a0dfe112fa9..95715954549c 100644
--- a/vendor/magento/module-catalog-permissions-graph-ql/Model/Resolver/Categories/DataProvider/Category/CollectionProcessor/ApplyPermissionsOnCategory.php
+++ b/vendor/magento/module-catalog-permissions-graph-ql/Model/Resolver/Categories/DataProvider/Category/CollectionProcessor/ApplyPermissionsOnCategory.php
@@ -1,7 +1,21 @@
 <?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+/************************************************************************
+ *
+ * ADOBE CONFIDENTIAL
+ * ___________________
+ *
+ * Copyright 2020 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ * ************************************************************************
  */
 declare(strict_types=1);
 
@@ -118,6 +132,10 @@ public function process(
         }
 
         foreach ($collection as $key => $category) {
+            if ($category->getLevel() <= 1) {
+                continue;
+            }
+
             $this->applyPermissionsOnCategory($category, $customerGroupId, $storeId);
 
             /** Filter out hidden items */
