diff -Nuar a/vendor/magento/module-rma/Block/Adminhtml/Rma/Edit/Tab/General/Shippingmethod.php b/vendor/magento/module-rma/Block/Adminhtml/Rma/Edit/Tab/General/Shippingmethod.php
--- a/vendor/magento/module-rma/Block/Adminhtml/Rma/Edit/Tab/General/Shippingmethod.php
+++ b/vendor/magento/module-rma/Block/Adminhtml/Rma/Edit/Tab/General/Shippingmethod.php
@@ -8,6 +8,7 @@
 use Magento\Framework\Pricing\PriceCurrencyInterface;
 use Magento\Framework\App\ObjectManager;
 use Magento\Framework\Serialize\Serializer\Json;
+use Magento\Shipping\Helper\Carrier;

 /**
  * Shipping Method Block at RMA page
@@ -84,6 +85,7 @@ class Shippingmethod extends \Magento\Rma\Block\Adminhtml\Rma\Edit\Tab\General\A
      * @param PriceCurrencyInterface $priceCurrency
      * @param array $data
      * @param Json|null $json
+     * @param Carrier|null $carrierHelper
      */
     public function __construct(
         \Magento\Backend\Block\Template\Context $context,
@@ -93,13 +95,15 @@ public function __construct(
         \Magento\Rma\Model\ShippingFactory $shippingFactory,
         PriceCurrencyInterface $priceCurrency,
         array $data = [],
-        Json $json = null
+        Json $json = null,
+        Carrier $carrierHelper = null
     ) {
         $this->priceCurrency = $priceCurrency;
         $this->_taxData = $taxData;
         $this->_rmaData = $rmaData;
         $this->_shippingFactory = $shippingFactory;
         $this->json = $json ?: ObjectManager::getInstance()->get(Json::class);
+        $data['carrierHelper'] = $carrierHelper ?? ObjectManager::getInstance()->get(Carrier::class);
         parent::__construct($context, $registry, $data);
     }

