diff -Naur a/vendor/magento/module-advanced-checkout/Model/Cart.php b/vendor/magento/module-advanced-checkout/Model/Cart.php
--- a/vendor/magento/module-advanced-checkout/Model/Cart.php
+++ a/vendor/magento/module-advanced-checkout/Model/Cart.php
@@ -33,6 +33,7 @@
     /**
      * Context of the cart - admin order
      */
+    // MCTEST-1001: EE change
     const CONTEXT_ADMIN_ORDER = 'admin_order';
 
     /**
@@ -1011,7 +1012,7 @@
                 ? false
                 : clone $this->productsConfig[$sku];
         }
-        
+
         return $this->productsConfig[$configSkuKey];
     }
 
