diff --git a/vendor/magento/module-gift-card-graph-ql/Model/Resolver/Product/Options/GiftCardOptions.php b/vendor/magento/module-gift-card-graph-ql/Model/Resolver/Product/Options/GiftCardOptions.php
index 7fd1021c81e..fb9ddf2e173 100644
--- a/vendor/magento/module-gift-card-graph-ql/Model/Resolver/Product/Options/GiftCardOptions.php
+++ b/vendor/magento/module-gift-card-graph-ql/Model/Resolver/Product/Options/GiftCardOptions.php
@@ -85,6 +85,9 @@ private function getCustomOptionsData(Product $product): array
             $customAmountOption = $this->getCustomAmountOptions($product);
             $customOptions = array_merge($customOptions, $customAmountOption);
         }
+        foreach ($customOptions as &$customOption) {
+            $customOption['uid'] = $customOption['value']['uid'];
+        }

         return $customOptions;
     }

