diff --git a/vendor/magento/module-shared-catalog/Model/TierPriceManagement.php b/vendor/magento/module-shared-catalog/Model/TierPriceManagement.php
index e3f609e80..a3128fb72 100644
--- a/vendor/magento/module-shared-catalog/Model/TierPriceManagement.php
+++ b/vendor/magento/module-shared-catalog/Model/TierPriceManagement.php
@@ -72,7 +72,7 @@ class TierPriceManagement
     ) {
         $groupCodes[] = $this->getCustomerGroupCode($sharedCatalog->getCustomerGroupId());
         if ($sharedCatalog->getType() == SharedCatalogInterface::TYPE_PUBLIC) {
-            $groupCodes[] = ProductItemManagementInterface::CUSTOMER_GROUP_NOT_LOGGED_IN;
+            $groupCodes[] = $this->getCustomerGroupCode(ProductItemManagementInterface::CUSTOMER_GROUP_NOT_LOGGED_IN);
         }
 
         while (count($skus)) {
