diff -Nuar a/vendor/magento/module-catalog-permissions/Plugin/UpdateCachePlugin.php b/vendor/magento/module-catalog-permissions/Plugin/UpdateCachePlugin.php
index e1b677405f1..e00f3a603fc 100644
--- a/vendor/magento/module-catalog-permissions/Plugin/UpdateCachePlugin.php
+++ b/vendor/magento/module-catalog-permissions/Plugin/UpdateCachePlugin.php
@@ -10,19 +10,15 @@ namespace Magento\CatalogPermissions\Plugin;
 use Magento\Customer\Model\Session;
 use Magento\CatalogPermissions\App\ConfigInterface;
 use Magento\Framework\App\Http\Context as HttpContext;
-use Magento\Framework\Registry;
 
 /**
  * Class UpdateCachePlugin to update Context with data
+ *
+ * @SuppressWarnings(PHPMD.CookieAndSessionMisuse)
  */
 class UpdateCachePlugin
 {
     /**
-     * @var Registry
-     */
-    private $coreRegistry;
-
-    /**
      * @var Session
      */
     private $customerSession;
@@ -33,16 +29,13 @@ class UpdateCachePlugin
     private $permissionsConfig;
 
     /**
-     * @param Registry $coreRegistry
      * @param Session $customerSession
      * @param ConfigInterface $permissionsConfig
      */
     public function __construct(
-        Registry $coreRegistry,
         Session $customerSession,
         ConfigInterface $permissionsConfig
     ) {
-        $this->coreRegistry = $coreRegistry;
         $this->customerSession = $customerSession;
         $this->permissionsConfig = $permissionsConfig;
     }
@@ -62,17 +55,11 @@ class UpdateCachePlugin
             return $result;
         }
 
-        $category = $this->coreRegistry->registry('current_category');
         $customerGroupId = $this->customerSession->getCustomerGroupId();
-
         if ($customerGroupId) {
             $result['customer_group'] = $customerGroupId;
         }
 
-        if ($category && $category->getId()) {
-            $result['category'] = $category->getId();
-        }
-
         return $result;
     }
 }
