diff --git a/vendor/magento/module-advanced-checkout/Model/Cart.php b/vendor/magento/module-advanced-checkout/Model/Cart.php
index ada42020107..2a0ae12a6a3 100644
--- a/vendor/magento/module-advanced-checkout/Model/Cart.php
+++ b/vendor/magento/module-advanced-checkout/Model/Cart.php
@@ -1242,7 +1242,7 @@ class Cart extends \Magento\Framework\DataObject implements \Magento\Checkout\Mo
         $checkedItems = $this->areProductsSalable($checkedItems);
         $products = $this->preloadProducts($checkedItems);
         foreach ($checkedItems as $sku => &$item) {
-            $itemProduct = $products[$sku] ?? null;
+            $itemProduct = $products[mb_strtolower($sku)] ?? null;
             $item = $this->checkItem(
                 $sku,
                 $item['qty'],
@@ -1942,7 +1942,7 @@ class Cart extends \Magento\Framework\DataObject implements \Magento\Checkout\Mo
                     $isSalable = false;
                 }
                 $product->setIsSalable($isSalable);
-                $products[$sku] = $product;
+                $products[mb_strtolower($sku)] = $product;
                 $this->addProductToLocalCache($product, $product->getStoreId());
             }
         }
