diff --git a/vendor/magento/module-staging/Model/ResourceModel/Update.php b/vendor/magento/module-staging/Model/ResourceModel/Update.php
index 03baeb38c501..ca22d26413b1 100644
--- a/vendor/magento/module-staging/Model/ResourceModel/Update.php
+++ b/vendor/magento/module-staging/Model/ResourceModel/Update.php
@@ -20,22 +20,21 @@
 class Update extends AbstractDb
 {
     /**
-     * Use is object new method for save of object
+     * The property value used to check the model for save. If it is not new it requires additional preparation to
+     * save the model properly
      *
      * @var bool
      */
     protected $_useIsObjectNew = true;

     /**
-     * Primary key auto increment flag
+     * The primary key auto increment flag. In this model the auto increment is not used for the primary key.
      *
      * @var bool
      */
     protected $_isPkAutoIncrement = false;

     /**
-     * Event prefix
-     *
      * @var string
      */
     protected $_eventPrefix = 'staging_update_resource';
@@ -108,6 +107,7 @@ public function getMaxIdByTime($timestamp)
             $select = $this->getConnection()->select()
                 ->from($this->getMainTable())
                 ->where('start_time <= ?', $date->format('Y-m-d H:i:s'))
+                ->where('moved_to IS NULL')
                 ->order(['id ' . \Magento\Framework\DB\Select::SQL_DESC])
                 ->limit(1);
             $this->versionCache[$timestamp] = $this->getConnection()->fetchOne($select);
