diff -Nuar a/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php b/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php
index 9f307ff9620..89ab29aea4a 100644
--- a/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php
+++ b/vendor/magento/module-target-rule/Model/Rule/Condition/Product/Attributes/SqlBuilder.php
@@ -219,7 +219,7 @@ class SqlBuilder
                 $this->indexResource->getOperatorCondition(
                     $valueExpr,
                     $condition->getOperator(),
-                    $condition->getValue()
+                    $condition->getValueParsed()
                 )
             );
         }
@@ -286,7 +286,7 @@ class SqlBuilder
                 $this->indexResource->getOperatorCondition(
                     'table.value',
                     $condition->getOperator(),
-                    $condition->getValue()
+                    $condition->getValueParsed()
                 )
             );
         }
@@ -314,7 +314,7 @@ class SqlBuilder
                 $this->indexResource->getOperatorCondition(
                     'table.value',
                     $condition->getOperator(),
-                    $condition->getValue()
+                    $condition->getValueParsed()
                 )
             );
         }
