diff --git a/vendor/magento/module-advanced-checkout/Model/Cart.php b/vendor/magento/module-advanced-checkout/Model/Cart.php
index ada42020107..a980b68d24d 100644
--- a/vendor/magento/module-advanced-checkout/Model/Cart.php
+++ b/vendor/magento/module-advanced-checkout/Model/Cart.php
@@ -1938,7 +1938,9 @@ class Cart extends \Magento\Framework\DataObject implements \Magento\Checkout\Mo
             foreach ($collection as $product) {
                 $sku = $product->getSku();
                 $isSalable = true;
-                if ($itemsLowerCase[mb_strtolower($sku)]['code'] === Data::ADD_ITEM_STATUS_FAILED_OUT_OF_STOCK) {
+                if (!isset($itemsLowerCase[mb_strtolower($sku)]) ||
+                    $itemsLowerCase[mb_strtolower($sku)]['code'] === Data::ADD_ITEM_STATUS_FAILED_OUT_OF_STOCK
+                ) {
                     $isSalable = false;
                 }
                 $product->setIsSalable($isSalable);
