diff --git a/vendor/magento/module-customer-balance/Model/Plugin/CollectQuoteTotalsPlugin.php b/vendor/magento/module-customer-balance/Model/Plugin/CollectQuoteTotalsPlugin.php
new file mode 100644
index 00000000000..7148ee02e4e
--- /dev/null
+++ b/vendor/magento/module-customer-balance/Model/Plugin/CollectQuoteTotalsPlugin.php
@@ -0,0 +1,28 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\CustomerBalance\Model\Plugin;
+
+use Magento\Quote\Model\Quote;
+
+class CollectQuoteTotalsPlugin
+{
+    /**
+     * Reset quote used customer balance amount
+     *
+     * @param \Magento\Quote\Model\Quote\TotalsCollector $subject
+     * @param Quote $quote
+     *
+     * @return void
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeCollectQuoteTotals(
+        \Magento\Quote\Model\Quote\TotalsCollector $subject,
+        Quote $quote
+    ) {
+        $quote->setBaseCustomerBalAmountUsed(0);
+        $quote->setCustomerBalanceAmountUsed(0);
+    }
+}
\ No newline at end of file
diff --git a/vendor/magento/module-customer-balance/etc/graphql/di.xml b/vendor/magento/module-customer-balance/etc/graphql/di.xml
new file mode 100644
index 00000000000..aee150823f4
--- /dev/null
+++ b/vendor/magento/module-customer-balance/etc/graphql/di.xml
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
+<!--
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
+    <type name ="Magento\Quote\Model\Quote\TotalsCollector">
+        <plugin name="quoteCollectTotalsPlugin" type="Magento\CustomerBalance\Model\Plugin\CollectQuoteTotalsPlugin" />
+    </type>
+</config>
\ No newline at end of file
