diff --git a/vendor/magento/module-page-builder/Controller/Adminhtml/ContentType/Image/Upload.php b/vendor/magento/module-page-builder/Controller/Adminhtml/ContentType/Image/Upload.php
index 40489a496..340f92de8 100644
--- a/vendor/magento/module-page-builder/Controller/Adminhtml/ContentType/Image/Upload.php
+++ b/vendor/magento/module-page-builder/Controller/Adminhtml/ContentType/Image/Upload.php
@@ -8,7 +8,9 @@ namespace Magento\PageBuilder\Controller\Adminhtml\ContentType\Image;
 use Magento\Framework\App\Action\HttpPostActionInterface;

 /**
- * Class Upload
+ * Image upload controller
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  */
 class Upload extends \Magento\Backend\App\Action implements HttpPostActionInterface
 {
@@ -101,7 +103,7 @@ class Upload extends \Magento\Backend\App\Action implements HttpPostActionInterf
             }

             $result = $fileUploader->save($this->getUploadDir());
-            $baseUrl = $this->storeManager->getStore()->getBaseUrl(\Magento\Framework\UrlInterface::URL_TYPE_MEDIA);
+            $baseUrl = $this->_backendUrl->getBaseUrl(['_type' => \Magento\Framework\UrlInterface::URL_TYPE_MEDIA]);
             $result['id'] = $this->cmsWysiwygImages->idEncode($result['file']);
             $result['url'] = $baseUrl . $this->getFilePath(self::UPLOAD_DIR, $result['file']);
         } catch (\Exception $e) {
diff --git a/vendor/magento/module-page-builder/Model/Stage/Config.php b/vendor/magento/module-page-builder/Model/Stage/Config.php
index 7535b138e..66cc38b3a 100644
--- a/vendor/magento/module-page-builder/Model/Stage/Config.php
+++ b/vendor/magento/module-page-builder/Model/Stage/Config.php
@@ -165,7 +165,7 @@ class Config
             'menu_sections' => $this->getMenuSections(),
             'content_types' => $this->getContentTypes(),
             'stage_config' => $this->data,
-            'media_url' => $this->frontendUrlBuilder->getBaseUrl(['_type' => UrlInterface::URL_TYPE_MEDIA]),
+            'media_url' => $this->urlBuilder->getBaseUrl(['_type' => UrlInterface::URL_TYPE_MEDIA]),
             'preview_url' => $this->urlBuilder->getUrl('pagebuilder/stage/preview'),
             'render_url' => $this->urlBuilder->getUrl('pagebuilder/stage/render'),
             'template_save_url' => $this->urlBuilder->getUrl('pagebuilder/template/save'),
