diff --git a/vendor/magento/module-customer-balance/Observer/RevertStoreCreditForOrder.php b/vendor/magento/module-customer-balance/Observer/RevertStoreCreditForOrder.php
index 3d038b7233c..fcd0c865b79 100644
--- a/vendor/magento/module-customer-balance/Observer/RevertStoreCreditForOrder.php
+++ b/vendor/magento/module-customer-balance/Observer/RevertStoreCreditForOrder.php
@@ -46,12 +46,14 @@ class RevertStoreCreditForOrder
             return $this;
         }
 
+        $customerAmountDelta = $order->getBaseCustomerBalanceAmount() - $order->getBaseCustomerBalanceInvoiced();
+
         $this->_balanceFactory->create()->setCustomerId(
             $order->getCustomerId()
         )->setWebsiteId(
             $this->_storeManager->getStore($order->getStoreId())->getWebsiteId()
         )->setAmountDelta(
-            $order->getBaseCustomerBalanceAmount()
+            $customerAmountDelta
         )->setHistoryAction(
             \Magento\CustomerBalance\Model\Balance\History::ACTION_REVERTED
         )->setOrder(
