diff --git a/vendor/magento/module-versions-cms/Controller/Adminhtml/Cms/Page/InlineEdit/Plugin.php b/vendor/magento/module-versions-cms/Controller/Adminhtml/Cms/Page/InlineEdit/Plugin.php
index 3e39b1045ec8..8f4bef329e26 100644
--- a/vendor/magento/module-versions-cms/Controller/Adminhtml/Cms/Page/InlineEdit/Plugin.php
+++ b/vendor/magento/module-versions-cms/Controller/Adminhtml/Cms/Page/InlineEdit/Plugin.php
@@ -5,6 +5,8 @@
  */
 namespace Magento\VersionsCms\Controller\Adminhtml\Cms\Page\InlineEdit;

+use Magento\VersionsCms\Model\Hierarchy\Node;
+
 /**
  * Plugin for cms page grid inline edit that adds information about hierarchy nodes
  */
@@ -55,6 +57,9 @@ public function beforeSetCmsPageData(
         $nodeCollection = $this->getNodesCollection()->joinPageExistsNodeInfo($page);
         $nodeCollectionData = $nodeCollection->getData();
         foreach ($nodeCollectionData as $node) {
+            if ($node['level'] == Node::NODE_LEVEL_FAKE) {
+                continue;
+            }
             $nodesData[$node['node_id']] = [
                 'node_id' => $node['node_id'],
                 'page_id' => $node['page_id'],
