diff --git a/vendor/magento/module-scalable-checkout/Model/Merger.php b/vendor/magento/module-scalable-checkout/Model/Merger.php
index e2694a2cdbf..acfa1d46941 100644
--- a/vendor/magento/module-scalable-checkout/Model/Merger.php
+++ b/vendor/magento/module-scalable-checkout/Model/Merger.php
@@ -32,13 +32,14 @@ class Merger implements \Magento\Framework\MessageQueue\MergerInterface
     public function merge(array $messageList)
     {
         $result = [];
-
+        $originalMessagesIds = [];
         foreach ($messageList as $topicName => $topicMessages) {
             foreach ($topicMessages as $messageId => $message) {
+                $originalMessagesIds[] = $messageId;
                 $mergedMessage = $this->mergedMessageFactory->create(
                     [
                         'mergedMessage' => $message,
-                        'originalMessagesIds' => [$messageId]
+                        'originalMessagesIds' => $originalMessagesIds,
                     ]
                 );
                 $result[$topicName] = [$mergedMessage];
