diff --git a/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php b/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
index 0f37bdc77f..e5e5e44014 100755
--- a/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
+++ b/vendor/magento/module-requisition-list-graph-ql/Model/RequisitionList/GetRequisitionList.php
@@ -82,7 +82,7 @@ class GetRequisitionList
 
         $builder = $this->getFilters($customerId, false, $args);
         $lists = $this->repository->getList($builder->create())->getItems();
-   
+
         $data = [
             'total_count' => count($lists),
             'items' => [],
@@ -131,16 +131,25 @@ class GetRequisitionList
     private function getFilters(int $customerId, bool $isTotal, array $args = [])
     {
         $filtersGiven = isset($args['filter']) ? $args['filter'] : [];
-        
+
         $filters = [];
 
         if (!$isTotal) {
             if (isset($filtersGiven['uids'])) {
                 if (key($filtersGiven['uids']) == 'eq' || key($filtersGiven['uids']) == 'in') {
+                    $reqIds = [];
+                    if (is_array($filtersGiven['uids'][key($filtersGiven['uids'])])) {
+                        foreach ($filtersGiven['uids'][key($filtersGiven['uids'])] as $id) {
+                            $reqIds[] = $this->idEncoder->decode($id);
+                        }
+                    } else {
+                        $reqIds = $this->idEncoder->decode($filtersGiven['uids'][key($filtersGiven['uids'])]);
+                    }
+                    
                     $filters[] = $this->filterBuilder
                         ->setField(RequisitionListInterface::REQUISITION_LIST_ID)
                         ->setConditionType(key($filtersGiven['uids']))
-                        ->setValue((int)$filtersGiven['uids'][key($filtersGiven['uids'])])
+                        ->setValue($reqIds)
                         ->create();
                 }
             }
