diff --git a/vendor/magento/module-shared-catalog/Plugin/AdvancedCheckout/Model/HideProductsAbsentInSharedCatalogPlugin.php b/vendor/magento/module-shared-catalog/Plugin/AdvancedCheckout/Model/HideProductsAbsentInSharedCatalogPlugin.php
index 0984d5400ef..26540996acf 100644
--- a/vendor/magento/module-shared-catalog/Plugin/AdvancedCheckout/Model/HideProductsAbsentInSharedCatalogPlugin.php
+++ b/vendor/magento/module-shared-catalog/Plugin/AdvancedCheckout/Model/HideProductsAbsentInSharedCatalogPlugin.php
@@ -105,6 +105,6 @@ private function getUnavalableProducts(array $skus, $customerGroupId)
         $collection->addFieldToFilter(ProductItemInterface::CUSTOMER_GROUP_ID, $customerGroupId);
         $collection->getSelect()->where(\sprintf('%s IN (?)', ProductItemInterface::SKU), $skus);

-        return \array_diff($skus, $collection->getColumnValues('sku'));
+        return \array_udiff($skus, $collection->getColumnValues('sku'), 'strcasecmp');
     }
 }
