diff --git a/vendor/magento/module-reward-graph-ql/Plugin/Quote/Model/Quote/TotalsCollector/ResetQuoteRewardPointsAmountPlugin.php b/vendor/magento/module-reward-graph-ql/Plugin/Quote/Model/Quote/TotalsCollector/ResetQuoteRewardPointsAmountPlugin.php
new file mode 100644
index 00000000000..2bababe02ee
--- /dev/null
+++ b/vendor/magento/module-reward-graph-ql/Plugin/Quote/Model/Quote/TotalsCollector/ResetQuoteRewardPointsAmountPlugin.php
@@ -0,0 +1,34 @@
+<?php
+/**
+ * Copyright Â© Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\RewardGraphQl\Plugin\Quote\Model\Quote\TotalsCollector;
+
+use Magento\Quote\Model\Quote;
+use Magento\Quote\Model\Quote\TotalsCollector;
+
+/**
+ * Reset quote reward points before collect totals plugin.
+ */
+class ResetQuoteRewardPointsAmountPlugin
+{
+    /**
+     * Reset quote reward points amount for correct totals calculation.
+     *
+     * @param TotalsCollector $subject
+     * @param Quote $quote
+     * @return void
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeCollectQuoteTotals(
+        TotalsCollector $subject,
+        Quote $quote
+    ) {
+        $quote->setRewardPointsBalance(0);
+        $quote->setRewardCurrencyAmount(0);
+        $quote->setBaseRewardCurrencyAmount(0);
+    }
+}
diff --git a/vendor/magento/module-reward-graph-ql/etc/graphql/di.xml b/vendor/magento/module-reward-graph-ql/etc/graphql/di.xml
index 8e784521828..296b51a7cea 100644
--- a/vendor/magento/module-reward-graph-ql/etc/graphql/di.xml
+++ b/vendor/magento/module-reward-graph-ql/etc/graphql/di.xml
@@ -46,4 +46,7 @@
             </argument>
         </arguments>
     </type>
+    <type name="Magento\Quote\Model\Quote\TotalsCollector">
+        <plugin name="reset_quote_reward_points_amount_plugin" type="Magento\RewardGraphQl\Plugin\Quote\Model\Quote\TotalsCollector\ResetQuoteRewardPointsAmountPlugin" />
+    </type>
 </config>
