diff -Nuar a/vendor/magento/module-company/Model/Email/Sender.php b/vendor/magento/module-company/Model/Email/Sender.php
index fe31d20f8..18960965e 100644
--- a/vendor/magento/module-company/Model/Email/Sender.php
+++ b/vendor/magento/module-company/Model/Email/Sender.php
@@ -257,11 +257,6 @@ class Sender
         $mergedCustomerData = $this->customerData->getDataObjectByCustomer($customer, $companyId);
 
         if ($companySuperUser && $mergedCustomerData) {
-            $sender = [
-                'name' => $companySuperUser->getName(),
-                'email' => $companySuperUser->getEmail()
-            ];
-
             $mergedCustomerData->setData('companyAdminEmail', $companySuperUser->getEmail());
             $this->sendEmailTemplate(
                 $customer->getEmail(),
@@ -270,7 +265,7 @@ class Sender
                     ScopeInterface::SCOPE_STORE,
                     $customer->getStoreId()
                 ),
-                $sender,
+                $this->xmlPathRegisterEmailIdentity,
                 ['customer' => $mergedCustomerData],
                 $customer->getStoreId()
             );
@@ -308,10 +303,7 @@ class Sender
                 $recipient,
                 $toName,
                 $this->emailTemplateConfig->getCompanyCreateNotifyAdminTemplateId(),
-                [
-                    'email' => $customer->getEmail(),
-                    'name' => $this->customerViewHelper->getCustomerName($customer)
-                ],
+                $this->xmlPathRegisterEmailIdentity,
                 [
                     'customer' => $customer->getFirstname(),
                     'company' => $companyName,
