diff --git a/vendor/magento/module-checkout-staging/Model/PreviewQuotaManager.php b/vendor/magento/module-checkout-staging/Model/PreviewQuotaManager.php
index f6d232f999d..e0034a9fbf2 100644
--- a/vendor/magento/module-checkout-staging/Model/PreviewQuotaManager.php
+++ b/vendor/magento/module-checkout-staging/Model/PreviewQuotaManager.php
@@ -90,14 +90,14 @@ class PreviewQuotaManager
                 )
             );
 
-            $this->searchCriteriaBuilder->addFilter('entity_id', $ids, 'in');
+            $this->searchCriteriaBuilder->addFilter('main_table.entity_id', $ids, 'in');
             $this->searchCriteriaBuilder->addFilter(
-                CartInterface::KEY_STORE_ID,
+                'main_table.' . CartInterface::KEY_STORE_ID,
                 $storeId,
                 'eq'
             );
             $this->searchCriteriaBuilder->addFilter(
-                CartInterface::KEY_UPDATED_AT,
+                'main_table.' . CartInterface::KEY_UPDATED_AT,
                 $now->format('Y-m-d H:i:s'),
                 'to'
             );
