diff --git a/vendor/magento/module-customer-balance-graph-ql/Model/Resolver/RemoveStoreCreditFromCart.php b/vendor/magento/module-customer-balance-graph-ql/Model/Resolver/RemoveStoreCreditFromCart.php
index a0e23bc42fd..8f013ae51ce 100644
--- a/vendor/magento/module-customer-balance-graph-ql/Model/Resolver/RemoveStoreCreditFromCart.php
+++ b/vendor/magento/module-customer-balance-graph-ql/Model/Resolver/RemoveStoreCreditFromCart.php
@@ -54,6 +54,7 @@ class RemoveStoreCreditFromCart implements ResolverInterface
         $cart = $this->getCartForUser->execute($maskedCartId, $customerId, $storeId);
         $this->unapply($cart);
 
+        $cart = $this->getCartForUser->execute($maskedCartId, $customerId, $storeId);
         return [
             'cart' => [
                 'model' => $cart,
diff --git a/vendor/magento/module-gift-card-account-graph-ql/Model/Resolver/ApplyGiftCardToCart.php b/vendor/magento/module-gift-card-account-graph-ql/Model/Resolver/ApplyGiftCardToCart.php
index 1a6c2e1a17a..72c56e3d15f 100644
--- a/vendor/magento/module-gift-card-account-graph-ql/Model/Resolver/ApplyGiftCardToCart.php
+++ b/vendor/magento/module-gift-card-account-graph-ql/Model/Resolver/ApplyGiftCardToCart.php
@@ -88,6 +88,7 @@ class ApplyGiftCardToCart implements ResolverInterface
             throw new GraphQlInputException(__($e->getMessage()), $e);
         }
 
+        $cart = $this->getCartForUser->execute($maskedCartId, $currentUserId, $storeId);
         return [
             'cart' => [
                 'model' => $cart,
