diff --git a/vendor/magento/module-catalog-permissions-graph-ql/Model/Store/StoreProcessor.php b/vendor/magento/module-catalog-permissions-graph-ql/Model/Store/StoreProcessor.php
index 73a47265c4e..9a7ba28e4d4 100644
--- a/vendor/magento/module-catalog-permissions-graph-ql/Model/Store/StoreProcessor.php
+++ b/vendor/magento/module-catalog-permissions-graph-ql/Model/Store/StoreProcessor.php
@@ -20,7 +20,7 @@ class StoreProcessor
      * @param ContextInterface|null $context
      * @return int|null
      */
-    public function getStoreId(ContextInterface $context = null): int
+    public function getStoreId(ContextInterface $context = null) : ?int
     {
         $storeId = null;
         if ($context) {
