diff --git a/vendor/magento/module-quick-order/Controller/Ajax/Search.php b/vendor/magento/module-quick-order/Controller/Ajax/Search.php
index 9e2d1cd0ee..1557cfd4b0 100644
--- a/vendor/magento/module-quick-order/Controller/Ajax/Search.php
+++ b/vendor/magento/module-quick-order/Controller/Ajax/Search.php
@@ -62,6 +62,7 @@ class Search extends \Magento\QuickOrder\Controller\AbstractAction implements Ht
             );
         } else {
             $this->cart->setContext(Cart::CONTEXT_FRONTEND);
+            $this->cart->removeAllAffectedItems();
             $this->cart->prepareAddProductsBySku($items);
             $items = $this->cart->getAffectedItems();
         }
