diff --git a/vendor/magento/module-customer-segment/view/adminhtml/web/js/conditions.js b/vendor/magento/module-customer-segment/view/adminhtml/web/js/conditions.js
index 886db8107894..5b95f4cc3577 100644
--- a/vendor/magento/module-customer-segment/view/adminhtml/web/js/conditions.js
+++ b/vendor/magento/module-customer-segment/view/adminhtml/web/js/conditions.js
@@ -15,5 +15,6 @@ define([
         if (config.isReadonly !== '') {
             segmentConditionsFieldset.setReadonly(true);
         }
+        window[config.jsObjectName] = segmentConditionsFieldset;
     };
 });
