diff --git a/vendor/magento/module-company/Plugin/Company/Model/EmailNotification.php b/vendor/magento/module-company/Plugin/Company/Model/EmailNotification.php
index 55dc2458d68a..47d1d74953c7 100644
--- a/vendor/magento/module-company/Plugin/Company/Model/EmailNotification.php
+++ b/vendor/magento/module-company/Plugin/Company/Model/EmailNotification.php
@@ -69,7 +69,7 @@ public function afterSave(
         Save $subject,
         CompanyInterface $company
     ): CompanyInterface {
-        if ($company) {
+        if ($company && $company->isObjectNew()) {
             $customerData = $this->customerRepository
                 ->getById(
                     $company->getSuperUserId()
