diff -Nuar a/vendor/magento/module-negotiable-quote/Model/Plugin/Cart/ValidateQuoteCurrentStorePlugin.php b/vendor/magento/module-negotiable-quote/Model/Plugin/Cart/ValidateQuoteCurrentStorePlugin.php
new file mode 100644
index 000000000..a9518b809
--- /dev/null
+++ b/vendor/magento/module-negotiable-quote/Model/Plugin/Cart/ValidateQuoteCurrentStorePlugin.php
@@ -0,0 +1,39 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\NegotiableQuote\Model\Plugin\Cart;
+
+use Magento\AdvancedCheckout\Model\Cart;
+use Magento\Store\Model\Store;
+
+/**
+ * Class ValidateProductCurrentStore plugin
+ *
+ * Validate current store id with the quote store id
+ */
+class ValidateQuoteCurrentStorePlugin
+{
+    /**
+     * After get current store plugin
+     *
+     * @param Cart $cart
+     * @param Store $currentStore
+     * @return Store
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterGetCurrentStore(
+        Cart $cart,
+        Store $currentStore
+    ): Store {
+        $cartStore = $cart->getQuote()->getStore();
+        $quoteStoreId = $cartStore->getStoreId();
+        if ($quoteStoreId &&  $quoteStoreId !== $currentStore->getStoreId()) {
+            $currentStore = $cartStore;
+        }
+        return $currentStore;
+    }
+}
diff -Nuar a/vendor/magento/module-negotiable-quote/etc/adminhtml/di.xml b/vendor/magento/module-negotiable-quote/etc/adminhtml/di.xml
index 6c3c67a63..893c18cf6 100644
--- a/vendor/magento/module-negotiable-quote/etc/adminhtml/di.xml
+++ b/vendor/magento/module-negotiable-quote/etc/adminhtml/di.xml
@@ -27,4 +27,7 @@
             <argument name="customerGroupRetriever" xsi:type="object">\Magento\NegotiableQuote\Model\CustomerGroupRetriever</argument>
         </arguments>
     </type>
+    <type name="Magento\AdvancedCheckout\Model\Cart">
+        <plugin name="validateQuoteCurrentStorePlugin" type="Magento\NegotiableQuote\Model\Plugin\Cart\ValidateQuoteCurrentStorePlugin"/>
+    </type>
 </config>
