diff --git a/vendor/magento/module-page-builder/Ui/DataProvider/Product/ProductCollection.php b/vendor/magento/module-page-builder/Ui/DataProvider/Product/ProductCollection.php
index 87c4d2b78..1bdda9c00 100644
--- a/vendor/magento/module-page-builder/Ui/DataProvider/Product/ProductCollection.php
+++ b/vendor/magento/module-page-builder/Ui/DataProvider/Product/ProductCollection.php
@@ -29,4 +29,13 @@ class ProductCollection extends \Magento\Catalog\Model\ResourceModel\Product\Col
 
         return $this;
     }
+
+    /**
+     * @inheritdoc
+     */
+    protected function _productLimitationJoinPrice()
+    {
+        $this->_productLimitationFilters->setUsePriceIndex($this->getStoreId() !== Store::DEFAULT_STORE_ID);
+        return $this->_productLimitationPrice(false);
+    }
 }
