diff --git a/vendor/magento/module-requisition-list/Api/Data/RequisitionListItemInterface.php b/vendor/magento/module-requisition-list/Api/Data/RequisitionListItemInterface.php
index 5e7881cc95..18c8314480 100644
--- a/vendor/magento/module-requisition-list/Api/Data/RequisitionListItemInterface.php
+++ b/vendor/magento/module-requisition-list/Api/Data/RequisitionListItemInterface.php
@@ -53,7 +53,7 @@ interface RequisitionListItemInterface extends \Magento\Framework\Api\Extensible
     /**
      * Get Product SKU.
      *
-     * @return int
+     * @return string
      */
     public function getSku();
 
