diff --git a/vendor/magento/module-advanced-checkout/Model/Cart.php b/vendor/magento/module-advanced-checkout/Model/Cart.php
index 6322a62989b6..762757c8b5ee 100644
--- a/vendor/magento/module-advanced-checkout/Model/Cart.php
+++ b/vendor/magento/module-advanced-checkout/Model/Cart.php
@@ -33,17 +33,17 @@ class Cart extends \Magento\Framework\DataObject implements \Magento\Checkout\Mo
     /**
      * Context of the cart - admin order
      */
-    const CONTEXT_ADMIN_ORDER = 'admin_order';
+    public const CONTEXT_ADMIN_ORDER = 'admin_order';
 
     /**
      * Context of the cart - admin checkout
      */
-    const CONTEXT_ADMIN_CHECKOUT = 'admin_checkout';
+    public const CONTEXT_ADMIN_CHECKOUT = 'admin_checkout';
 
     /**
      * Context of the cart - frontend
      */
-    const CONTEXT_FRONTEND = 'frontend';
+    public const CONTEXT_FRONTEND = 'frontend';
 
     /**
      * Context of the cart
@@ -126,7 +126,7 @@ class Cart extends \Magento\Framework\DataObject implements \Magento\Checkout\Mo
     protected $quoteRepository;
 
     /**
-     * Wishlist factory
+     * Wishlist factory model
      *
      * @var \Magento\Wishlist\Model\WishlistFactory
      */
@@ -976,7 +976,7 @@ protected function _loadProductBySku($sku)
             try {
                 $product = $this->productRepository->get($sku, false, $storeId);
                 $this->addProductToLocalCache($product, $storeId);
-            } catch (NoSuchEntityException $e) {
+            } catch (\Magento\Framework\Exception\LocalizedException $e) {
                 return false;
             }
         }
