--- a/vendor/paypal/module-braintree-core/view/frontend/web/js/paypal/button.js	2026-03-24 00:04:41.716453736 -0700
+++ b/vendor/paypal/module-braintree-core/view/frontend/web/js/paypal/button.js	2026-03-24 00:04:41.736453720 -0700
@@ -180,6 +180,9 @@
              * @param buttonConfig
              */
             loadSDK: function (buttonConfig) {
+                if (!buttonConfig || !buttonConfig.clientToken) {
+                    return;
+                }
                 // Load SDK
                 braintree.create({
                     authorization: buttonConfig.clientToken
