diff --git a/vendor/magento/module-page-cache/Model/App/Request/Http/IdentifierForSave.php b/vendor/magento/module-page-cache/Model/App/Request/Http/IdentifierForSave.php
index cbb3c34db6cbc..09ce18fe731a4 100644
--- a/vendor/magento/module-page-cache/Model/App/Request/Http/IdentifierForSave.php
+++ b/vendor/magento/module-page-cache/Model/App/Request/Http/IdentifierForSave.php
@@ -13,6 +13,7 @@
 use Magento\Framework\App\PageCache\IdentifierInterface;
 use Magento\Framework\App\Request\Http;
 use Magento\Framework\Serialize\Serializer\Json;
+use Magento\Framework\App\Response\Http as ResponseHttp;
 
 /**
  * Page unique identifier
@@ -49,11 +50,10 @@ public function getValue()
         $data = [
             $this->request->isSecure(),
             preg_replace($pattern, $replace, (string)$this->request->getUriString()),
-            $this->context->getVaryString()
+            $this->request->get(ResponseHttp::COOKIE_VARY_STRING)
+                ?: $this->context->getVaryString()
         ];
-
         $data = $this->identifierStoreReader->getPageTagsWithStoreCacheTags($data);
-
         return sha1($this->serializer->serialize($data));
     }
 }
