diff --git a/vendor/magento/framework-amqp/Connection/Factory.php b/vendor/magento/framework-amqp/Connection/Factory.php
--- a/vendor/magento/framework-amqp/Connection/Factory.php
+++ b/vendor/magento/framework-amqp/Connection/Factory.php	(date 1750251863102)
@@ -39,6 +39,7 @@
         // Set SSL options if SSL is enabled
         if ($options->isSslEnabled()) {
             $config->setIsSecure(true);
+            $config->setNetworkProtocol(\Magento\Framework\Amqp\Config::SSL);
             $sslOptions = $options->getSslOptions();
             if ($sslOptions) {
                 if (isset($sslOptions['cafile'])) {
