diff --git a/vendor/magento/module-sales-graph-ql/Model/Formatter/Order.php b/vendor/magento/module-sales-graph-ql/Model/Formatter/Order.php
index 12b9a28c60755..418b996bfb664 100644
--- a/vendor/magento/module-sales-graph-ql/Model/Formatter/Order.php
+++ b/vendor/magento/module-sales-graph-ql/Model/Formatter/Order.php
@@ -49,7 +49,7 @@ public function format(OrderInterface $orderModel): array
             'increment_id' => $orderModel->getIncrementId(),
             'number' => $orderModel->getIncrementId(),
             'order_date' => $this->timezone->date($orderModel->getCreatedAt())
-                ->format(DateTime::DATETIME_SLASH_PHP_FORMAT),
+                ->format(DateTime::DATETIME_PHP_FORMAT),
             'order_number' => $orderModel->getIncrementId(),
             'status' => $orderModel->getStatusLabel(),
             'email' => $orderModel->getCustomerEmail(),
diff --git a/vendor/magento/module-sales-graph-ql/Model/Resolver/CreditMemo/CreditMemoComments.php b/vendor/magento/module-sales-graph-ql/Model/Resolver/CreditMemo/CreditMemoComments.php
index 4c9a60edcc22a..f243024174491 100644
--- a/vendor/magento/module-sales-graph-ql/Model/Resolver/CreditMemo/CreditMemoComments.php
+++ b/vendor/magento/module-sales-graph-ql/Model/Resolver/CreditMemo/CreditMemoComments.php
@@ -53,7 +53,7 @@ public function resolve(
                 $comments[] = [
                     'message' => $comment->getComment(),
                     'timestamp' => $this->timezone->date($comment->getCreatedAt())
-                        ->format(DateTime::DATETIME_SLASH_PHP_FORMAT)
+                        ->format(DateTime::DATETIME_PHP_FORMAT)
                 ];
             }
         }
diff --git a/vendor/magento/module-sales-graph-ql/Model/Resolver/Invoices.php b/vendor/magento/module-sales-graph-ql/Model/Resolver/Invoices.php
index fd0681bd05792..d08f2d6d1a8ba 100644
--- a/vendor/magento/module-sales-graph-ql/Model/Resolver/Invoices.php
+++ b/vendor/magento/module-sales-graph-ql/Model/Resolver/Invoices.php
@@ -74,7 +74,7 @@ private function getInvoiceComments(InvoiceInterface $invoice): array
             if ($comment->getIsVisibleOnFront()) {
                 $comments[] = [
                     'timestamp' => $this->timezone->date($comment->getCreatedAt())
-                        ->format(DateTime::DATETIME_SLASH_PHP_FORMAT),
+                        ->format(DateTime::DATETIME_PHP_FORMAT),
                     'message' => $comment->getComment()
                 ];
             }
diff --git a/vendor/magento/module-sales-graph-ql/Model/Resolver/Shipments.php b/vendor/magento/module-sales-graph-ql/Model/Resolver/Shipments.php
index b50c038370253..b923b1e9e2fc0 100644
--- a/vendor/magento/module-sales-graph-ql/Model/Resolver/Shipments.php
+++ b/vendor/magento/module-sales-graph-ql/Model/Resolver/Shipments.php
@@ -75,7 +75,7 @@ private function getShipmentComments(ShipmentInterface $shipment): array
             if ($comment->getIsVisibleOnFront()) {
                 $comments[] = [
                     'timestamp' => $this->timezone->date($comment->getCreatedAt())
-                        ->format(DateTime::DATETIME_SLASH_PHP_FORMAT),
+                        ->format(DateTime::DATETIME_PHP_FORMAT),
                     'message' => $comment->getComment()
                 ];
             }

 
