diff --git a/vendor/magento/module-paypal/Setup/Patch/Data/UpdateBmltoPayLater.php b/vendor/magento/module-paypal/Setup/Patch/Data/UpdateBmltoPayLater.php
--- a/vendor/magento/module-paypal/Setup/Patch/Data/UpdateBmltoPayLater.php
+++ b/vendor/magento/module-paypal/Setup/Patch/Data/UpdateBmltoPayLater.php
@@ -134,6 +134,9 @@ public function apply()
         foreach ($bmlSettings as $bmlPath => $bmlValue) {
             $setting = str_replace(self::BMLPATH, '', $bmlPath);
             $settingParts = explode('_', $setting);
+            if (count($settingParts) !== 2) {
+                continue;
+            }
             $page = $settingParts[0];
             $setting = $settingParts[1];
             $payLaterPage = $page === 'checkout' ? 'cartpage' : $page;