diff --git a/vendor/magento/magento2-base/lib/web/mage/requirejs/mixins.js b/vendor/magento/magento2-base/lib/web/mage/requirejs/mixins.js
index bebd10b1d6ee3..46647e66fbec2 100644
--- a/vendor/magento/magento2-base/lib/web/mage/requirejs/mixins.js
+++ b/vendor/magento/magento2-base/lib/web/mage/requirejs/mixins.js
@@ -227,8 +227,17 @@ require([
      * that way it is able to respect any changes done after mixins module has initialized.
      */
     defContext.configure = function (cfg) {
+        var unbundledCfg = {};
+
         originalContextConfigure(cfg);
-        unbundledContext.configure(cfg);
+
+        Object.keys(cfg || {}).forEach(function (key) {
+            if (key !== 'deps' && key !== 'callback') {
+                unbundledCfg[key] = cfg[key];
+            }
+        });
+
+        unbundledContext.configure(unbundledCfg);
     };
 
     /**
