diff --git a/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php b/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php
index 75097b36115ce..f64eda6a2e44b 100644
--- a/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php
+++ b/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php
@@ -142,8 +142,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $searchConfig = $this->searchConfigFactory->create();
             $this->cache->clean();
             $searchConfig->validateSearchEngine();
-            $installer->removeUnusedTriggers();
             $installer->installSchema($request);
+            $installer->removeUnusedTriggers();
             $installer->installDataFixtures($request, true);
 
             if ($this->deploymentConfig->isAvailable()) {
