diff --git a/vendor/magento/module-inventory-bundle-product/Plugin/InventorySales/Model/IsProductSalableCondition/GetIsQtySalableForBundleProduct.php b/vendor/magento/module-inventory-bundle-product/Plugin/InventorySales/Model/IsProductSalableCondition/GetIsQtySalableForBundleProduct.php
index 429fbf96c95..ad84b1889c7 100644
--- a/vendor/magento/module-inventory-bundle-product/Plugin/InventorySales/Model/IsProductSalableCondition/GetIsQtySalableForBundleProduct.php
+++ b/vendor/magento/module-inventory-bundle-product/Plugin/InventorySales/Model/IsProductSalableCondition/GetIsQtySalableForBundleProduct.php
@@ -53,8 +53,9 @@ class GetIsQtySalableForBundleProduct
         string $sku,
         int $stockId
     ): bool {
-        return $isSalable && $this->getProductTypesBySkus->execute([$sku])[$sku] === Type::TYPE_CODE
-            ? $this->isBundleProductChildrenSalable->execute($sku, $stockId)
-            : $isSalable;
+        if ($this->getProductTypesBySkus->execute([$sku])[$sku] === Type::TYPE_CODE) {
+            $isSalable = $this->isBundleProductChildrenSalable->execute($sku, $stockId);
+        }
+        return $isSalable;
     }
 }
