diff --git a/vendor/magento/module-checkout/Controller/Cart/Add.php b/vendor/magento/module-checkout/Controller/Cart/Add.php
index ea02c43acc0..59d672ccb6c 100644
--- a/vendor/magento/module-checkout/Controller/Cart/Add.php
+++ b/vendor/magento/module-checkout/Controller/Cart/Add.php
@@ -144,6 +144,11 @@ class Add extends \Magento\Checkout\Controller\Cart implements HttpPostActionInt
                         $product->getName()
                     );
                     $this->messageManager->addSuccessMessage($message);
+                } elseif ($this->cart->getQuote()->getHasError()) {
+                    $errors = $this->cart->getQuote()->getErrors();
+                    foreach ($errors as $error) {
+                        $this->messageManager->addErrorMessage($error->getText());
+                    }
                 } else {
                     $this->messageManager->addComplexSuccessMessage(
                         'addCartSuccessMessage',
@@ -153,12 +158,6 @@ class Add extends \Magento\Checkout\Controller\Cart implements HttpPostActionInt
                         ]
                     );
                 }
-                if ($this->cart->getQuote()->getHasError()) {
-                    $errors = $this->cart->getQuote()->getErrors();
-                    foreach ($errors as $error) {
-                        $this->messageManager->addErrorMessage($error->getText());
-                    }
-                }
                 return $this->goBack(null, $product);
             }
         } catch (\Magento\Framework\Exception\LocalizedException $e) {
