diff --git a/vendor/magento/module-multishipping/Plugin/MultishippingQuoteRepository.php b/vendor/magento/module-multishipping/Plugin/MultishippingQuoteRepository.php
index af19e4bc91f5..e5f26b46accc 100644
--- a/vendor/magento/module-multishipping/Plugin/MultishippingQuoteRepository.php
+++ b/vendor/magento/module-multishipping/Plugin/MultishippingQuoteRepository.php
@@ -144,6 +144,7 @@ private function getQuoteItems(Quote $quote, Quote\Address $address): array
             $quoteItem = $quote->getItemById($addressItem->getQuoteItemId());
             if ($quoteItem) {
                 $multishippingQuoteItem = clone $quoteItem;
+                $multishippingQuoteItem->setQuote($quoteItem->getQuote());
                 $qty = $addressItem->getQty();
                 $sku = $multishippingQuoteItem->getSku();
                 if (isset($quoteItems[$sku])) {
