diff --git a/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/QuoteItemQtyList.php b/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/QuoteItemQtyList.php
index 600bf9897a0..363f91916fb 100644
--- a/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/QuoteItemQtyList.php
+++ b/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/QuoteItemQtyList.php
@@ -33,7 +33,7 @@ class QuoteItemQtyList
     public function getQty($productId, $quoteItemId, $quoteId, $itemQty)
     {
         $qty = $itemQty;
-        if (isset($this->_checkedQuoteItems[$quoteId][$productId]['qty']) && !in_array(
+        if (isset($this->_checkedQuoteItems[$quoteId][$productId]['qty']) && $quoteItemId !== null && !in_array(
             $quoteItemId,
             $this->_checkedQuoteItems[$quoteId][$productId]['items']
         )
