diff --git a/vendor/magento/module-sales/view/frontend/templates/email/invoice/items.phtml b/vendor/magento/module-sales/view/frontend/templates/email/invoice/items.phtml
index e2efd650295..f2b2de505c3 100644
--- a/vendor/magento/module-sales/view/frontend/templates/email/invoice/items.phtml
+++ b/vendor/magento/module-sales/view/frontend/templates/email/invoice/items.phtml
@@ -4,26 +4,29 @@
  * See COPYING.txt for license details.
  */
 
+/** @var \Magento\Sales\Block\Order\Email\Invoice\Items $block */
+/** @var \Magento\Framework\Escaper $escaper */
+
 ?>
 <?php $_invoice = $block->getInvoice() ?>
 <?php $_order   = $block->getOrder() ?>
-<?php if ($_invoice && $_order) : ?>
+<?php if ($_invoice && $_order): ?>
     <table class="email-items">
         <thead>
             <tr>
                 <th class="item-info">
-                    <?= $block->escapeHtml(__('Items')) ?>
+                    <?= $escaper->escapeHtml(__('Items')) ?>
                 </th>
                 <th class="item-qty">
-                    <?= $block->escapeHtml(__('Qty')) ?>
+                    <?= $escaper->escapeHtml(__('Qty')) ?>
                 </th>
                 <th class="item-subtotal">
-                    <?= $block->escapeHtml(__('Subtotal')) ?>
+                    <?= $escaper->escapeHtml(__('Subtotal')) ?>
                 </th>
             </tr>
         </thead>
-        <?php foreach ($_invoice->getAllItems() as $_item) : ?>
-            <?php if (!$_item->getOrderItem()->getParentItem()) : ?>
+        <?php foreach ($_invoice->getAllItems() as $_item): ?>
+            <?php if (!$_item->getOrderItem()->getParentItem()): ?>
                 <tbody>
                     <?= $block->getItemHtml($_item) ?>
                 </tbody>
diff --git a/vendor/magento/module-sales/view/frontend/templates/email/items/creditmemo/default.phtml b/vendor/magento/module-sales/view/frontend/templates/email/items/creditmemo/default.phtml
index 6c304db4533..7c4b96ea550 100644
--- a/vendor/magento/module-sales/view/frontend/templates/email/items/creditmemo/default.phtml
+++ b/vendor/magento/module-sales/view/frontend/templates/email/items/creditmemo/default.phtml
@@ -4,28 +4,29 @@
  * See COPYING.txt for license details.
  */
 
+/** @var \Magento\Framework\Escaper $escaper */
 ?>
 <?php $_item = $block->getItem() ?>
 <?php $_order = $block->getItem()->getOrder(); ?>
 <tr>
     <td class="item-info<?= ($block->getItemOptions() ? ' has-extra' : '') ?>">
-        <p class="product-name"><?= $block->escapeHtml($_item->getName()) ?></p>
-        <p class="sku"><?= $block->escapeHtml(__('SKU')) ?>: <?= $block->escapeHtml($block->getSku($_item)) ?></p>
-        <?php if ($block->getItemOptions()) : ?>
+        <p class="product-name"><?= $escaper->escapeHtml($_item->getName()) ?></p>
+        <p class="sku"><?= $escaper->escapeHtml(__('SKU')) ?>: <?= $escaper->escapeHtml($block->getSku($_item)) ?></p>
+        <?php if ($block->getItemOptions()): ?>
             <dl>
-                <?php foreach ($block->getItemOptions() as $option) : ?>
-                    <dt><strong><em><?= $block->escapeHtml($option['label']) ?></em></strong></dt>
+                <?php foreach ($block->getItemOptions() as $option): ?>
+                    <dt><strong><em><?= $escaper->escapeHtml($option['label']) ?></em></strong></dt>
                     <dd>
-                        <?= /* @noEscape */  nl2br($block->escapeHtml($option['value'])) ?>
+                        <?= /* @noEscape */  nl2br($option['value']) ?>
                     </dd>
                 <?php endforeach; ?>
             </dl>
         <?php endif; ?>
         <?php $addInfoBlock = $block->getProductAdditionalInformationBlock(); ?>
-        <?php if ($addInfoBlock) :?>
+        <?php if ($addInfoBlock):?>
             <?= $addInfoBlock->setItem($_item->getOrderItem())->toHtml() ?>
         <?php endif; ?>
-        <?= $block->escapeHtml($_item->getDescription()) ?>
+        <?= $escaper->escapeHtml($_item->getDescription()) ?>
     </td>
     <td class="item-qty"><?= (float) $_item->getQty() ?></td>
     <td class="item-price">
diff --git a/vendor/magento/module-sales/view/frontend/templates/email/items/invoice/default.phtml b/vendor/magento/module-sales/view/frontend/templates/email/items/invoice/default.phtml
index 217cf85dde1..f53f527a491 100644
--- a/vendor/magento/module-sales/view/frontend/templates/email/items/invoice/default.phtml
+++ b/vendor/magento/module-sales/view/frontend/templates/email/items/invoice/default.phtml
@@ -3,28 +3,29 @@
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
  */
+/** @var \Magento\Framework\Escaper $escaper */
 ?>
 <?php $_item = $block->getItem() ?>
 <?php $_order = $block->getItem()->getOrder(); ?>
 <tr>
     <td class="item-info<?= ($block->getItemOptions() ? ' has-extra' : '') ?>">
-        <p class="product-name"><?= $block->escapeHtml($_item->getName()) ?></p>
-        <p class="sku"><?= $block->escapeHtml(__('SKU')) ?>: <?= $block->escapeHtml($block->getSku($_item)) ?></p>
-        <?php if ($block->getItemOptions()) : ?>
+        <p class="product-name"><?= $escaper->escapeHtml($_item->getName()) ?></p>
+        <p class="sku"><?= $escaper->escapeHtml(__('SKU')) ?>: <?= $escaper->escapeHtml($block->getSku($_item)) ?></p>
+        <?php if ($block->getItemOptions()): ?>
             <dl>
-                <?php foreach ($block->getItemOptions() as $option) : ?>
-                    <dt><strong><em><?= $block->escapeHtml($option['label']) ?></em></strong></dt>
+                <?php foreach ($block->getItemOptions() as $option): ?>
+                    <dt><strong><em><?= $escaper->escapeHtml($option['label']) ?></em></strong></dt>
                     <dd>
-                        <?= /* @noEscape */  nl2br($block->escapeHtml($option['value'])) ?>
+                        <?= /* @noEscape */  nl2br($option['value']) ?>
                     </dd>
                 <?php endforeach; ?>
             </dl>
         <?php endif; ?>
         <?php $addInfoBlock = $block->getProductAdditionalInformationBlock(); ?>
-        <?php if ($addInfoBlock) : ?>
+        <?php if ($addInfoBlock): ?>
             <?= $addInfoBlock->setItem($_item->getOrderItem())->toHtml() ?>
         <?php endif; ?>
-        <?= $block->escapeHtml($_item->getDescription()) ?>
+        <?= $escaper->escapeHtml($_item->getDescription()) ?>
     </td>
     <td class="item-qty"><?= (float) $_item->getQty() ?></td>
     <td class="item-price">
diff --git a/vendor/magento/module-sales/view/frontend/templates/email/items/order/default.phtml b/vendor/magento/module-sales/view/frontend/templates/email/items/order/default.phtml
index 37525c2fc62..9b9dd57d89f 100644
--- a/vendor/magento/module-sales/view/frontend/templates/email/items/order/default.phtml
+++ b/vendor/magento/module-sales/view/frontend/templates/email/items/order/default.phtml
@@ -5,32 +5,35 @@
  */
 
 // phpcs:disable Magento2.Templates.ThisInTemplate
+// phpcs:disable Magento2.Files.LineLength, Generic.Files.LineLength
+
 
 /** @var $block \Magento\Sales\Block\Order\Email\Items\DefaultItems */
 
 /** @var $_item \Magento\Sales\Model\Order\Item */
+/** @var \Magento\Framework\Escaper $escaper */
 $_item = $block->getItem();
 $_order = $_item->getOrder();
 ?>
 <tr>
     <td class="item-info<?= ($block->getItemOptions() ? ' has-extra' : '') ?>">
-        <p class="product-name"><?= $block->escapeHtml($_item->getName()) ?></p>
-        <p class="sku"><?= $block->escapeHtml(__('SKU')) ?>: <?= $block->escapeHtml($block->getSku($_item)) ?></p>
-        <?php if ($block->getItemOptions()) : ?>
+        <p class="product-name"><?= $escaper->escapeHtml($_item->getName()) ?></p>
+        <p class="sku"><?= $escaper->escapeHtml(__('SKU')) ?>: <?= $escaper->escapeHtml($block->getSku($_item)) ?></p>
+        <?php if ($block->getItemOptions()): ?>
             <dl class="item-options">
-            <?php foreach ($block->getItemOptions() as $option) : ?>
-                <dt><strong><em><?= $block->escapeHtml($option['label']) ?></em></strong></dt>
+            <?php foreach ($block->getItemOptions() as $option): ?>
+                <dt><strong><em><?= $escaper->escapeHtml($option['label']) ?></em></strong></dt>
                 <dd>
-                    <?= /* @noEscape */  nl2br($block->escapeHtml($option['value'])) ?>
+                    <?= /* @noEscape */  nl2br($option['value']) ?>
                 </dd>
             <?php endforeach; ?>
             </dl>
         <?php endif; ?>
         <?php $addInfoBlock = $block->getProductAdditionalInformationBlock(); ?>
-        <?php if ($addInfoBlock) :?>
+        <?php if ($addInfoBlock):?>
             <?= $addInfoBlock->setItem($_item)->toHtml() ?>
         <?php endif; ?>
-        <?= $block->escapeHtml($_item->getDescription()) ?>
+        <?= $escaper->escapeHtml($_item->getDescription()) ?>
     </td>
     <td class="item-qty"><?= (float) $_item->getQtyOrdered() ?></td>
     <td class="item-price">
@@ -40,17 +43,17 @@ $_order = $_item->getOrder();
 <?php if ($_item->getGiftMessageId()
     && $_giftMessage = $this->helper(\Magento\GiftMessage\Helper\Message::class)
         ->getGiftMessage($_item->getGiftMessageId())
-) : ?>
+): ?>
     <tr>
     <td colspan="3" class="item-extra">
         <table class="message-gift">
             <tr>
                 <td>
-                    <h3><?= $block->escapeHtml(__('Gift Message')) ?></h3>
-                    <strong><?= $block->escapeHtml(__('From:')) ?></strong> <?= $block->escapeHtml($_giftMessage->getSender()) ?>
-                    <br /><strong><?= $block->escapeHtml(__('To:')) ?></strong> <?= $block->escapeHtml($_giftMessage->getRecipient()) ?>
-                    <br /><strong><?= $block->escapeHtml(__('Message:')) ?></strong>
-                    <br /><?= $block->escapeHtml($_giftMessage->getMessage()) ?>
+                    <h3><?= $escaper->escapeHtml(__('Gift Message')) ?></h3>
+                    <strong><?= $escaper->escapeHtml(__('From:')) ?></strong> <?= $escaper->escapeHtml($_giftMessage->getSender()) ?>
+                    <br /><strong><?= $escaper->escapeHtml(__('To:')) ?></strong> <?= $escaper->escapeHtml($_giftMessage->getRecipient()) ?>
+                    <br /><strong><?= $escaper->escapeHtml(__('Message:')) ?></strong>
+                    <br /><?= $escaper->escapeHtml($_giftMessage->getMessage()) ?>
                 </td>
             </tr>
         </table>
diff --git a/vendor/magento/module-sales/view/frontend/templates/email/items/shipment/default.phtml b/vendor/magento/module-sales/view/frontend/templates/email/items/shipment/default.phtml
index 75962ff20ef..8e1cf70653f 100644
--- a/vendor/magento/module-sales/view/frontend/templates/email/items/shipment/default.phtml
+++ b/vendor/magento/module-sales/view/frontend/templates/email/items/shipment/default.phtml
@@ -4,16 +4,17 @@
  * See COPYING.txt for license details.
  */
 
-/** @var $_item \Magento\Sales\Model\Order\Item */
+/** @var \Magento\Framework\Escaper $escaper */
+/** @var \Magento\Sales\Model\Order\Item $_item */
 $_item = $block->getItem() ?>
 <tr>
     <td class="item-info<?= ($block->getItemOptions() ? ' has-extra' : '') ?>">
-        <p class="product-name"><?= $block->escapeHtml($_item->getName()) ?></p>
-        <p class="sku"><?= $block->escapeHtml(__('SKU')) ?>: <?= $block->escapeHtml($block->getSku($_item)) ?></p>
-        <?php if ($block->getItemOptions()) : ?>
+        <p class="product-name"><?= $escaper->escapeHtml($_item->getName()) ?></p>
+        <p class="sku"><?= $escaper->escapeHtml(__('SKU')) ?>: <?= $escaper->escapeHtml($block->getSku($_item)) ?></p>
+        <?php if ($block->getItemOptions()): ?>
             <dl class="item-options">
-                <?php foreach ($block->getItemOptions() as $option) : ?>
-                    <dt><strong><em><?= $block->escapeHtml($option['label']) ?></em></strong></dt>
+                <?php foreach ($block->getItemOptions() as $option): ?>
+                    <dt><strong><em><?= $escaper->escapeHtml($option['label']) ?></em></strong></dt>
                     <dd>
                         <?= /* @noEscape */  nl2br($option['value']) ?>
                     </dd>
@@ -21,10 +22,10 @@ $_item = $block->getItem() ?>
             </dl>
         <?php endif; ?>
         <?php $addInfoBlock = $block->getProductAdditionalInformationBlock(); ?>
-        <?php if ($addInfoBlock) : ?>
+        <?php if ($addInfoBlock): ?>
             <?= $addInfoBlock->setItem($_item->getOrderItem())->toHtml() ?>
         <?php endif; ?>
-        <?= $block->escapeHtml($_item->getDescription()) ?>
+        <?= $escaper->escapeHtml($_item->getDescription()) ?>
     </td>
     <td class="item-qty"><?= (float) $_item->getQty() ?></td>
 </tr>
