diff --git a/vendor/magento/module-catalog/Helper/Product/View.php b/vendor/magento/module-catalog/Helper/Product/View.php
index ca9b550ca31d..abf151c713e9 100644
--- a/vendor/magento/module-catalog/Helper/Product/View.php
+++ b/vendor/magento/module-catalog/Helper/Product/View.php
@@ -129,14 +129,7 @@ private function preparePageMetadata(ResultPage $resultPage, $product)
             $pageConfig->setKeywords($product->getName());
         }

-        $description = $product->getMetaDescription();
-        if ($description) {
-            $pageConfig->setDescription($description);
-        } else {
-            $productDescription = is_string($product->getDescription()) ?
-                $this->string->substr(strip_tags($product->getDescription()), 0, 255) : '';
-            $pageConfig->setDescription($productDescription);
-        }
+        $pageConfig->setDescription($product->getMetaDescription());

         if ($this->_catalogProduct->canUseCanonicalTag()) {
             $pageConfig->addRemotePageAsset(
