diff --git a/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockStatusToSelect.php b/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockStatusToSelect.php
index e736e64643ad..d549f24ae610 100644
--- a/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockStatusToSelect.php
+++ b/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockStatusToSelect.php
@@ -11,9 +11,6 @@
 use Magento\InventoryIndexer\Indexer\IndexStructure;
 use Magento\InventoryIndexer\Model\StockIndexTableNameResolverInterface;

-/**
- * Adapt adding stock status to select for Multi Stocks.
- */
 class AddStockStatusToSelect
 {
     /**
@@ -30,6 +27,8 @@ public function __construct(StockIndexTableNameResolverInterface $stockIndexTabl
     }

     /**
+     * Adapt adding stock status to select for Multi Stocks.
+     *
      * @param Select $select
      * @param int $stockId
      * @return void
@@ -41,7 +40,7 @@ public function execute(Select $select, int $stockId)
         $select->joinLeft(
             ['stock_status' => $tableName],
             'e.sku = stock_status.sku',
-            [IndexStructure::IS_SALABLE]
+            [IndexStructure::IS_SALABLE => IndexStructure::IS_SALABLE]
         );
     }
 }
