diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/MaskedCartId.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/MaskedCartId.php
index c607c77659dc0..ffbcfe8db8e2e 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/MaskedCartId.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/MaskedCartId.php
@@ -95,11 +95,7 @@ private function getQuoteMaskId(int $quoteId): string
      */
     private function ensureQuoteMaskExist(int $quoteId): string
     {
-        try {
-            $maskedId = $this->quoteIdToMaskedQuoteId->execute($quoteId);
-        } catch (NoSuchEntityException $e) {
-            $maskedId = '';
-        }
+        $maskedId = $this->quoteIdToMaskedQuoteId->execute($quoteId);
         if ($maskedId === '') {
             $quoteIdMask = $this->quoteIdMaskFactory->create();
             $quoteIdMask->setQuoteId($quoteId);
