diff --git a/vendor/magento/module-google-optimizer/Model/Plugin/Catalog/Category/DataProvider.php b/vendor/magento/module-google-optimizer/Model/Plugin/Catalog/Category/DataProvider.php
index 7c0330740a1..3aea6acb915 100644
--- a/vendor/magento/module-google-optimizer/Model/Plugin/Catalog/Category/DataProvider.php
+++ b/vendor/magento/module-google-optimizer/Model/Plugin/Catalog/Category/DataProvider.php
@@ -34,6 +34,8 @@ class DataProvider
     }
 
     /**
+     * Updates metadata.
+     *
      * @param \Magento\Catalog\Model\Category\DataProvider $subject
      * @param array $result
      * @return array
@@ -45,6 +47,7 @@ class DataProvider
             !$this->_helper->isGoogleExperimentActive();
         $result['category_view_optimization']['arguments']['data']['config']['componentType'] =
             \Magento\Ui\Component\Form\Fieldset::NAME;
+        $result['category_view_optimization']['arguments']['data']['config']['label'] = '';
 
         return $result;
     }
