diff --git a/vendor/magento/module-sales/Model/ResourceModel/Provider/Query/IdListBuilder.php b/vendor/magento/module-sales/Model/ResourceModel/Provider/Query/IdListBuilder.php
index bf624620c133..65b11e1129b3 100644
--- a/vendor/magento/module-sales/Model/ResourceModel/Provider/Query/IdListBuilder.php
+++ b/vendor/magento/module-sales/Model/ResourceModel/Provider/Query/IdListBuilder.php
@@ -54,6 +54,18 @@ public function addAdditionalGridTable(string $table): IdListBuilder
         return $this;
     }
 
+    /**
+     * Reset added additional grid table where entities may already exist.
+     *
+     * @return $this
+     */
+    public function resetAdditionalGridTable(): IdListBuilder
+    {
+        $this->additionalGridTables = [];
+
+        return $this;
+    }
+
     /**
      * Returns connection.
      *
