diff --git a/vendor/magento/module-url-rewrite/view/adminhtml/templates/categories.phtml b/vendor/magento/module-url-rewrite/view/adminhtml/templates/categories.phtml
index bccffcef8786..38ff5c5b1edd 100644
--- a/vendor/magento/module-url-rewrite/view/adminhtml/templates/categories.phtml
+++ b/vendor/magento/module-url-rewrite/view/adminhtml/templates/categories.phtml
@@ -5,23 +5,23 @@
  */
 
 /** @var \Magento\UrlRewrite\Block\Catalog\Category\Tree $block */
+$root = $block->getRoot(null, 0)
 ?>
     <fieldset class="admin__fieldset" data-ui-id="category-selector">
         <legend class="admin__legend"><span><?= $block->escapeHtml(__('Select Category')) ?></span></legend>
         <div class="content content-category-tree">
             <input type="hidden" name="categories" id="product_categories" value=""/>
-            <?php if ($block->getRoot()) : ?>
+            <?php if ($root): ?>
                 <div class="jstree-default"></div>
             <?php endif; ?>
         </div>
     </fieldset>
-<?php if ($block->getRoot()) : ?>
+<?php if ($root): ?>
     <script type="text/x-magento-init">
     {
         ".jstree-default": {
             "categoryTree": {
-                "data": <?= /* @noEscape */
-                $this->helper(\Magento\Framework\Json\Helper\Data::class)->jsonEncode($block->getTreeArray()); ?>,
+                "data": <?= /* @noEscape */ $block->getTreeArray(null, true); ?>,
                 "url": "<?= $block->escapeJs($block->escapeUrl($block->getLoadTreeUrl())); ?>"
             }
         }
