diff --git a/vendor/magento/module-sales-rule/Model/Quote/Discount.php b/vendor/magento/module-sales-rule/Model/Quote/Discount.php
index 19e9bdf377bf9..3b91d61f19f8f 100644
--- a/vendor/magento/module-sales-rule/Model/Quote/Discount.php
+++ b/vendor/magento/module-sales-rule/Model/Quote/Discount.php
@@ -159,7 +159,7 @@ public function collect(
         $address->setCartFixedRules([]);
         $quote->setCartFixedRules([]);
         foreach ($items as $item) {
-            $this->rulesApplier->setAppliedRuleIds($item, []);
+            $item->setAppliedRuleIds(null);
             if ($item->getExtensionAttributes()) {
                 $item->getExtensionAttributes()->setDiscounts(null);
             }
