diff --git a/vendor/magento/module-quote-graph-ql/etc/schema.graphqls b/vendor/magento/module-quote-graph-ql/etc/schema.graphqls
index 1dc66531fbac..85a8c9b8cbee 100644
--- a/vendor/magento/module-quote-graph-ql/etc/schema.graphqls
+++ b/vendor/magento/module-quote-graph-ql/etc/schema.graphqls
@@ -123,6 +123,7 @@ input CartAddressInput @doc(description: "Defines the billing or shipping addres
     postcode: String @doc(description: "The ZIP or postal code of the billing or shipping address.")
     country_code: String! @doc(description: "The country code and label for the billing or shipping address.")
     telephone: String @doc(description: "The telephone number for the billing or shipping address.")
+    vat_id: String @doc(description: "The VAT company number for billing or shipping address.")
     save_in_address_book: Boolean @doc(description: "Determines whether to save the address in the customer's address book. The default value is true.")
 }

@@ -229,6 +230,7 @@ interface CartAddressInterface @typeResolver(class: "\\Magento\\QuoteGraphQl\\Mo
     postcode: String @doc(description: "The ZIP or postal code of the billing or shipping address.")
     country: CartAddressCountry! @doc(description: "An object containing the country label and code.")
     telephone: String @doc(description: "The telephone number for the billing or shipping address.")
+    vat_id: String @doc(description: "The VAT company number for billing or shipping address.")
 }

 type ShippingCartAddress implements CartAddressInterface @doc(description: "Contains shipping addresses and methods.") {
