diff --git a/vendor/magento/module-directory/Model/Currency.php b/vendor/magento/module-directory/Model/Currency.php
index dd17b00e0e02e..fbc547f75a41c 100644
--- a/vendor/magento/module-directory/Model/Currency.php
+++ b/vendor/magento/module-directory/Model/Currency.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */

 namespace Magento\Directory\Model;
@@ -438,10 +438,6 @@ private function formatCurrency(string $price, array $options): string
             $this->getCode() ?? $this->numberFormatter->getTextAttribute(\NumberFormatter::CURRENCY_CODE)
         );

-        if ((array_key_exists(LocaleCurrency::CURRENCY_OPTION_DISPLAY, $options)
-            && $options[LocaleCurrency::CURRENCY_OPTION_DISPLAY] === \Magento\Framework\Currency::NO_SYMBOL)) {
-            $formattedCurrency = str_replace(' ', '', $formattedCurrency);
-        }
         if (preg_match('/^(\x{200F})/u', $formattedCurrency, $match)) {
             $formattedCurrency = preg_replace('/^' . $match[1] . '/u', '', $formattedCurrency);
         }
@@ -488,6 +484,8 @@ private function setOptions(array $options): void
         if (array_key_exists(LocaleCurrency::CURRENCY_OPTION_DISPLAY, $options)
             && $options[LocaleCurrency::CURRENCY_OPTION_DISPLAY] === \Magento\Framework\Currency::NO_SYMBOL) {
             $this->numberFormatter->setSymbol(\NumberFormatter::CURRENCY_SYMBOL, '');
+            $this->numberFormatter->setTextAttribute(\NumberFormatter::POSITIVE_PREFIX, '');
+            $this->numberFormatter->setTextAttribute(\NumberFormatter::POSITIVE_SUFFIX, '');
         }
         if (array_key_exists('precision', $options)) {
             $this->numberFormatter->setAttribute(\NumberFormatter::FRACTION_DIGITS, $options['precision']);

