diff --git a/vendor/magento/module-catalog-import-export/Model/Import/Product.php b/vendor/magento/module-catalog-import-export/Model/Import/Product.php
index bec6517..5afa370 100644
--- a/vendor/magento/module-catalog-import-export/Model/Import/Product.php
+++ b/vendor/magento/module-catalog-import-export/Model/Import/Product.php
@@ -2041,10 +2041,7 @@ class Product extends AbstractEntity
             }
             if (self::SCOPE_STORE == $rowScope) {
                 if (self::SCOPE_WEBSITE == $attribute->getIsGlobal()) {
-                    // check website defaults already set
-                    if (!isset($attributes[$attrTable][$rowSku][$attrId][$rowStore])) {
-                        $storeIds = $this->storeResolver->getStoreIdToWebsiteStoreIds($rowStore);
-                    }
+                    $storeIds = $this->storeResolver->getStoreIdToWebsiteStoreIds($rowStore);
                 } elseif (self::SCOPE_STORE == $attribute->getIsGlobal()) {
                     $storeIds = [$rowStore];
                 }
