diff --git a/vendor/magento/module-sales-rule/Plugin/CouponUsagesDecrement.php b/vendor/magento/module-sales-rule/Plugin/CouponUsagesDecrement.php
index 3be801a288479..845f3b0ea5a3d 100644
--- a/vendor/magento/module-sales-rule/Plugin/CouponUsagesDecrement.php
+++ b/vendor/magento/module-sales-rule/Plugin/CouponUsagesDecrement.php
@@ -54,6 +54,9 @@ public function afterCancel(OrderService $subject, bool $result, $orderId): bool
         }
 
         $order = $this->orderRepository->get($orderId);
+        if (!empty($order->getDiscountInvoiced())) {
+            return $result;
+        }
         $this->updateCouponUsages->execute($order, false);
 
         return $result;
