diff --git a/vendor/magento/module-email/Block/Adminhtml/Template/Grid/Renderer/Action.php b/vendor/magento/module-email/Block/Adminhtml/Template/Grid/Renderer/Action.php
index 65f9e41b074..7b4964f1782 100644
--- a/vendor/magento/module-email/Block/Adminhtml/Template/Grid/Renderer/Action.php
+++ b/vendor/magento/module-email/Block/Adminhtml/Template/Grid/Renderer/Action.php
@@ -24,8 +24,8 @@ class Action extends \Magento\Backend\Block\Widget\Grid\Column\Renderer\Action
 
         $actions[] = [
             'url' => $this->getUrl('adminhtml/*/preview', ['id' => $row->getId()]),
-            'popup' => true,
             'caption' => __('Preview'),
+	     'target' => '_blank'
         ];
 
         $this->getColumn()->setActions($actions);
diff --git a/vendor/magento/module-email/view/adminhtml/templates/preview/iframeswitcher.phtml b/vendor/magento/module-email/view/adminhtml/templates/preview/iframeswitcher.phtml
index fd66acb7aca..ce097c9aa1b 100644
--- a/vendor/magento/module-email/view/adminhtml/templates/preview/iframeswitcher.phtml
+++ b/vendor/magento/module-email/view/adminhtml/templates/preview/iframeswitcher.phtml
@@ -34,7 +34,7 @@ require([
         $('#preview_form').submit();
     });
 
-    $('#preview_iframe').load(function() {
+    $('#preview_iframe').on('load', function() {
         $(this).height($(this).contents().height());
     });
 });
