diff --git a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
index 4491ad1ee99..24baf22287c 100644
--- a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
+++ b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
@@ -122,24 +122,6 @@ class Helper
      */
     private $categoryLinkFactory;
 
-    /**
-     * @var array
-     */
-    private $productDataKeys = [
-        'weight',
-        'special_price',
-        'cost',
-        'country_of_manufacture',
-        'description',
-        'short_description',
-        'meta_description',
-        'meta_keyword',
-        'meta_title',
-        'page_layout',
-        'custom_design',
-        'gift_wrapping_price'
-    ];
-
     /**
      * Constructor
      *
@@ -222,12 +204,6 @@ class Helper
             $productData['product_has_weight'] = 0;
         }
 
-        foreach ($productData as $key => $value) {
-            if (in_array($key, $this->productDataKeys) && $value === '') {
-                $productData[$key] = null;
-            }
-        }
-
         foreach (['category_ids', 'website_ids'] as $field) {
             if (!isset($productData[$field])) {
                 $productData[$field] = [];
