diff --git a/vendor/magento/module-inventory-configurable-product/Plugin/Model/ResourceModel/Attribute/IsEnabledOptionSelectBuilder.php b/vendor/magento/module-inventory-configurable-product/Plugin/Model/ResourceModel/Attribute/IsEnabledOptionSelectBuilder.php
index dfe682f50048..edddcc150021 100644
--- a/vendor/magento/module-inventory-configurable-product/Plugin/Model/ResourceModel/Attribute/IsEnabledOptionSelectBuilder.php
+++ b/vendor/magento/module-inventory-configurable-product/Plugin/Model/ResourceModel/Attribute/IsEnabledOptionSelectBuilder.php
@@ -82,7 +82,7 @@ public function afterGetSelect(
             . " AND entity_status_store.store_id = {$storeId}",
             []
         )->where(
-            $select->getConnection()->getIfNullSql('entity_status_global.value', 'entity_status_store.value') . ' = ?',
+            $select->getConnection()->getIfNullSql('entity_status_store.value', 'entity_status_global.value') . ' = ?',
             ProductStatus::STATUS_ENABLED
         );
 
