diff --git a/vendor/magento/framework/Session/SaveHandler.php b/vendor/magento/framework/Session/SaveHandler.php
index 1e96dac40f3..1bdc4cedff9 100644
--- a/vendor/magento/framework/Session/SaveHandler.php
+++ b/vendor/magento/framework/Session/SaveHandler.php
@@ -126,7 +126,7 @@ class SaveHandler implements SaveHandlerInterface
         $sessionMaxSize = $this->sessionMaxSizeConfig->getSessionMaxSize();
         $sessionSize = strlen($sessionData);

-        if ($sessionSize !== null && $sessionMaxSize < $sessionSize) {
+        if ($sessionMaxSize !== null && $sessionMaxSize < $sessionSize) {
             $sessionData = '';
             if ($this->appState->getAreaCode() === Area::AREA_FRONTEND) {
                 $this->messageManager->addErrorMessage(

