diff --git a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
index b16cb3e801e..8ee11e881b2 100644
--- a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
+++ b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
@@ -40,7 +40,7 @@ class PhpCookieManager implements CookieManagerInterface
      * RFC 2109 - Page 15
      * http://www.ietf.org/rfc/rfc6265.txt
      */
-    public const MAX_NUM_COOKIES = 50;
+    public const MAX_NUM_COOKIES = 200;
     public const MAX_COOKIE_SIZE = 4096;
     public const EXPIRE_NOW_TIME = 1;
     public const EXPIRE_AT_END_OF_SESSION_TIME = 0;
