diff --git a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php
index 81ab67bdf26..dfcab034200 100644
--- a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php
+++ b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php
@@ -80,7 +80,7 @@ class AttributeFilter
             // For non-numeric types set the attributeValue to 'false' to trigger their removal from the db
             if ($attributeType === 'varchar' || $attributeType === 'text' || $attributeType === 'datetime') {
                 $attribute->setIsRequired(false);
-                $productData[$attributeCode] = $attribute->getDefaultValue() ?: false;
+                $productData[$attributeCode] = false;
             } else {
                 $productData[$attributeCode] = null;
             }
