diff --git a/vendor/magento/module-quote-graph-ql/etc/schema.graphqls b/vendor/magento/module-quote-graph-ql/etc/schema.graphqls
index 2d8849a79dad..6149be1eff63 100644
--- a/vendor/magento/module-quote-graph-ql/etc/schema.graphqls
+++ b/vendor/magento/module-quote-graph-ql/etc/schema.graphqls
@@ -119,6 +119,7 @@ input CartAddressInput {
     postcode: String
     country_code: String!
     telephone: String
+    vat_id: String @doc(description: "The VAT company number for billing or shipping address.")
     save_in_address_book: Boolean @doc(description: "Determines whether to save the address in the customer's address book. The default value is true")
 }
 
@@ -225,6 +226,7 @@ interface CartAddressInterface @typeResolver(class: "\\Magento\\QuoteGraphQl\\Mo
     postcode: String
     country: CartAddressCountry!
     telephone: String
+    vat_id: String @doc(description: "The VAT company number for billing or shipping address.")
 }
 
 type ShippingCartAddress implements CartAddressInterface {
