diff --git a/vendor/magento/module-graph-ql/Model/Query/Logger/NewRelic.php b/vendor/magento/module-graph-ql/Model/Query/Logger/NewRelic.php
index 55f25c176ed..ca20a6970eb 100644
--- a/vendor/magento/module-graph-ql/Model/Query/Logger/NewRelic.php
+++ b/vendor/magento/module-graph-ql/Model/Query/Logger/NewRelic.php
@@ -49,7 +49,7 @@ class NewRelic implements LoggerInterface
             $this->newRelicWrapper->addCustomParameter($key, $value);
         }
 
-        $transactionName = $queryDetails[LoggerInterface::OPERATION_NAMES] ?: '';
+        $transactionName = $queryDetails[LoggerInterface::OPERATION_NAMES] ??= '';
 
         $this->newRelicWrapper->setTransactionName('GraphQL-' . $transactionName);
     }
