diff --git a/vendor/magento/module-inventory-catalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/AllowNegativeMinQtyPlugin.php b/vendor/magento/module-inventory-catalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/AllowNegativeMinQtyPlugin.php
deleted file mode 100644
index 509fc7f4841..00000000000
--- a/vendor/magento/module-inventory-catalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/AllowNegativeMinQtyPlugin.php
+++ /dev/null
@@ -1,32 +0,0 @@
-<?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
- */
-declare(strict_types=1);
-
-namespace Magento\InventoryCatalog\Plugin\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter;
-
-use Magento\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter;
-
-class AllowNegativeMinQtyPlugin
-{
-    /**
-     * Allow min_qty to be assigned a value below 0.
-     * @param StockDataFilter $subject
-     * @param callable $proceed
-     * @param array $stockData
-     * @return mixed
-     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     */
-    public function afterFilter(
-        StockDataFilter $subject,
-        array $result,
-        array $stockData
-    ) {
-        if (isset($stockData['min_qty'])) {
-            $result['min_qty'] = $stockData['min_qty'];
-        }
-        return $result;
-    }
-}
diff --git a/vendor/magento/module-inventory-catalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/StockDataFilterPlugin.php b/vendor/magento/module-inventory-catalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/StockDataFilterPlugin.php
new file mode 100644
index 00000000000..ad28bfb623c
--- /dev/null
+++ b/vendor/magento/module-inventory-catalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/StockDataFilterPlugin.php
@@ -0,0 +1,37 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter;
+
+use Magento\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter;
+
+class StockDataFilterPlugin
+{
+    /**
+     * Allow min_qty to be assigned a value below 0.
+     *
+     * @param StockDataFilter $subject
+     * @param array $result
+     * @param array $stockData
+     * @return mixed
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterFilter(
+        StockDataFilter $subject,
+        array $result,
+        array $stockData
+    ) {
+        if (!isset($stockData['qty'])) {
+            $result['qty'] = 0;
+        }
+
+        if (isset($stockData['min_qty'])) {
+            $result['min_qty'] = $stockData['min_qty'];
+        }
+        return $result;
+    }
+}
diff --git a/vendor/magento/module-inventory-catalog/etc/di.xml b/vendor/magento/module-inventory-catalog/etc/di.xml
index 29bb39faa6f..8ef18d01ecf 100644
--- a/vendor/magento/module-inventory-catalog/etc/di.xml
+++ b/vendor/magento/module-inventory-catalog/etc/di.xml
@@ -153,7 +153,7 @@
     <!-- Negative Min Quantity Threshold for Backorders -->
     <type name="Magento\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter">
         <plugin name="allow_negative_min_qty"
-                type="Magento\InventoryCatalog\Plugin\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter\AllowNegativeMinQtyPlugin"/>
+                type="Magento\InventoryCatalog\Plugin\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter\StockDataFilterPlugin"/>
     </type>
     <type name="Magento\CatalogInventory\Api\Data\StockItemInterface">
         <plugin name="adapt_min_qty_to_backorders"
