diff --git a/vendor/magento/module-ui/view/base/web/js/grid/provider.js b/vendor/magento/module-ui/view/base/web/js/grid/provider.js
index 021ae83eb1c7..01dcd2680799 100644
--- a/vendor/magento/module-ui/view/base/web/js/grid/provider.js
+++ b/vendor/magento/module-ui/view/base/web/js/grid/provider.js
@@ -15,8 +15,9 @@ define([
     'Magento_Ui/js/modal/alert',
     'mage/translate',
     'uiElement',
+    'uiRegistry',
     'Magento_Ui/js/grid/data-storage'
-], function ($, _, utils, resolver, layout, alert, $t, Element) {
+], function ($, _, utils, resolver, layout, alert, $t, Element, registry) {
     'use strict';
 
     return Element.extend({
@@ -178,9 +179,9 @@ define([
          * @param {Object} requestConfig
          */
         updateRequestConfig: function (requestConfig) {
-            if (this.storage()) {
-                _.extend(this.storage().requestConfig, requestConfig);
-            }
+            registry.get(this.storageConfig.provider, function (storage) {
+                _.extend(storage.requestConfig, requestConfig);
+            });
         }
     });
 });
