diff --git a/vendor/magento/module-customer/Model/Plugin/CustomerNotification.php b/vendor/magento/module-customer/Model/Plugin/CustomerNotification.php
index ab3cf8cb7d85..c3fb90f1375e 100644
--- a/vendor/magento/module-customer/Model/Plugin/CustomerNotification.php
+++ b/vendor/magento/module-customer/Model/Plugin/CustomerNotification.php
@@ -110,6 +110,7 @@ public function beforeExecute(ActionInterface $subject)
 
         if (!$this->isFrontendRequest()
             || !$this->isPostRequest()
+            || $this->isLogoutRequest()
             || !$this->isSessionUpdateRegisteredFor($customerId)) {
             return;
         }
@@ -146,6 +147,18 @@ private function isPostRequest(): bool
         return $this->request instanceof HttpRequestInterface && $this->request->isPost();
     }
 
+    /**
+     * Checks if the current request is a logout request.
+     *
+     * @return bool
+     */
+    private function isLogoutRequest(): bool
+    {
+        return $this->request->getRouteName() === 'customer'
+            && $this->request->getControllerName() === 'account'
+            && $this->request->getActionName() === 'logout';
+    }
+
     /**
      * Check if the current application area is frontend.
      *
