diff --git a/vendor/magento/module-import-export/Model/Export/Entity/AbstractEav.php b/vendor/magento/module-import-export/Model/Export/Entity/AbstractEav.php
index d9dd98bc54c..3e0b403089a 100644
--- a/vendor/magento/module-import-export/Model/Export/Entity/AbstractEav.php
+++ b/vendor/magento/module-import-export/Model/Export/Entity/AbstractEav.php
@@ -286,7 +286,8 @@ abstract class AbstractEav extends \Magento\ImportExport\Model\Export\AbstractEn
 
             if ($this->isMultiselect($attributeCode)) {
                 $values = [];
-                $attributeValue = explode(Import::DEFAULT_GLOBAL_MULTI_VALUE_SEPARATOR, $attributeValue);
+                $attributeValue =
+                    $attributeValue ? explode(Import::DEFAULT_GLOBAL_MULTI_VALUE_SEPARATOR, $attributeValue) : [];
                 foreach ($attributeValue as $value) {
                     $values[] = $this->getAttributeValueById($attributeCode, $value);
                 }
