diff --git a/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php b/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php
index e0122a00a656f..8c60feefb6431 100644
--- a/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php
+++ b/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php
@@ -22,8 +22,6 @@ class GetDiscounts
      */
     public function execute(Quote $quote, array $discounts): ?array
     {
-        $discounts = $discounts ?: $quote->getBillingAddress()->getExtensionAttributes()->getDiscounts() ?? [];
-
         if (empty($discounts)) {
             return null;
         }
diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/Discounts.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/Discounts.php
index 83c64258226cf..4095d6b483279 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/Discounts.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/Discounts.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2019 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);
 
@@ -12,6 +12,7 @@
 use Magento\Framework\GraphQl\Query\ResolverInterface;
 use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
 use Magento\QuoteGraphQl\Model\GetDiscounts;
+use Magento\Quote\Model\Quote;
 
 /**
  * @inheritdoc
@@ -37,11 +38,13 @@ public function resolve(Field $field, $context, ResolveInfo $info, ?array $value
         if (!isset($value['model'])) {
             throw new LocalizedException(__('"model" value should be specified'));
         }
+        /** @var Quote $quote */
         $quote = $value['model'];
-
+        $address = $quote->getIsVirtual() ? $quote->getBillingAddress() : $quote->getShippingAddress();
+        $discounts = $address->getExtensionAttributes()?->getDiscounts() ?? [];
         return $this->getDiscounts->execute(
             $quote,
-            $quote->getShippingAddress()->getExtensionAttributes()->getDiscounts() ?? []
+            $discounts
         );
     }
 }
