diff --git a/vendor/magento/module-eav/Model/Entity/Attribute/UniqueValidator.php b/vendor/magento/module-eav/Model/Entity/Attribute/UniqueValidator.php
index b1888b42bef92..bb5e589e86b7a 100644
--- a/vendor/magento/module-eav/Model/Entity/Attribute/UniqueValidator.php
+++ b/vendor/magento/module-eav/Model/Entity/Attribute/UniqueValidator.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2018 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\Eav\Model\Entity\Attribute;

@@ -23,8 +23,9 @@ public function validate(
         $entityLinkField,
         array $entityIds
     ) {
-        if (isset($entityIds[0])) {
-            return $entityIds[0] == $object->getData($entityLinkField);
+        if ($entityIds) {
+            // check for current and future updates
+            return in_array($object->getData($entityLinkField), $entityIds);
         }
         return true;
     }
