diff --git a/vendor/magento/module-wishlist/Model/ResourceModel/Item/Collection.php b/vendor/magento/module-wishlist/Model/ResourceModel/Item/Collection.php
index ce1a1cdc2942..77202f8a64f6 100644
--- a/vendor/magento/module-wishlist/Model/ResourceModel/Item/Collection.php
+++ b/vendor/magento/module-wishlist/Model/ResourceModel/Item/Collection.php
@@ -80,8 +80,6 @@ class Collection extends \Magento\Framework\Model\ResourceModel\Db\Collection\Ab
     protected $_isProductNameJoined = false;
 
     /**
-     * Adminhtml sales
-     *
      * @var \Magento\Sales\Helper\Admin
      */
     protected $_adminhtmlSales = null;
@@ -351,7 +349,7 @@ protected function _assignProducts()
         /** @var \Magento\Wishlist\Model\Item $item */
         foreach ($this as $item) {
             $product = $productCollection->getItemById($item->getProductId());
-            if ($product) {
+            if ($product && !$product->getIsHidden()) {
                 if ($checkInStock && !$product->isInStock()) {
                     $this->removeItemByKey($item->getId());
                 } else {
