diff --git a/vendor/magento/module-theme/view/base/requirejs-config.js b/vendor/magento/module-theme/view/base/requirejs-config.js
index 06e17f76d3c7..911ff329d205 100644
--- a/vendor/magento/module-theme/view/base/requirejs-config.js
+++ b/vendor/magento/module-theme/view/base/requirejs-config.js
@@ -68,7 +68,6 @@ var config = {
         'jquery/validate': 'jquery/jquery.validate',
         'jquery/file-uploader': 'jquery/fileUploader/jquery.fileuploader',
         'prototype': 'legacy-build.min',
-        'jquery/jquery.cookie': 'js-cookie/cookie-wrapper',
         'jquery/jquery-storageapi': 'js-storage/storage-wrapper',
         'text': 'mage/requirejs/text',
         'domReady': 'requirejs/domReady',
diff --git a/lib/web/jquery/jquery.cookie.js b/lib/web/jquery/jquery.cookie.js
new file mode 100644
index 000000000000..5268ae4ea75f
--- /dev/null
+++ b/lib/web/jquery/jquery.cookie.js
@@ -0,0 +1,11 @@
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+define([
+    'jquery',
+    'js-cookie/cookie-wrapper'
+], function () {
+
+});
diff --git a/lib/web/js-storage/js.storage.js b/lib/web/js-storage/js.storage.js
index b016fc0aa636..0ac802854a43 100644
--- a/lib/web/js-storage/js.storage.js
+++ b/lib/web/js-storage/js.storage.js
@@ -14,7 +14,7 @@
 (function (factory) {
     var registeredInModuleLoader = false;
     if (typeof define === 'function' && define.amd) {
-        define(['jquery', 'js-cookie/cookie-wrapper'], factory);
+        define(['jquery', 'jquery/jquery.cookie'], factory);
         registeredInModuleLoader = true;
     }
     if (typeof exports === 'object') {
