diff --git a/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Eav/Source.php b/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Eav/Source.php
index ce097dd95d77..4b0868fadba9 100644
--- a/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Eav/Source.php
+++ b/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Eav/Source.php
@@ -343,7 +343,7 @@ protected function _prepareMultiselectIndex($entityIds = null, $attributeId = nu
         $this->_addAttributeToSelect($select, 'status', "pvd.{$productIdField}", 'cs.store_id', $statusCond);
 
         if ($entityIds !== null) {
-            $select->where('cpe.entity_id IN(?)', $entityIds);
+            $select->where('cpe.entity_id IN(?)', $entityIds, \Zend_Db::INT_TYPE);
         }
         /**
          * Add additional external limitation
