diff --git a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
index 94dc4c27d9e6a..cf03bee75573e 100644
--- a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
+++ b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
@@ -31,7 +31,7 @@ class PhpCookieManager implements CookieManagerInterface
      * RFC 2109 - Page 26
      * http://www.ietf.org/rfc/rfc6265.txt
      */
-    public const MAX_NUM_COOKIES = 50;
+    public const MAX_NUM_COOKIES = 200;
     public const MAX_COOKIE_SIZE = 4096;
     public const EXPIRE_NOW_TIME = 1;
     public const EXPIRE_AT_END_OF_SESSION_TIME = 0;

