diff --git a/vendor/magento/module-aws-s3/Driver/AwsS3.php b/vendor/magento/module-aws-s3/Driver/AwsS3.php
index 76e3e2fc333..02c7e94de44 100644
--- a/vendor/magento/module-aws-s3/Driver/AwsS3.php
+++ b/vendor/magento/module-aws-s3/Driver/AwsS3.php
@@ -527,6 +527,10 @@ class AwsS3 implements RemoteDriverInterface
      */
     public function rename($oldPath, $newPath, DriverInterface $targetDriver = null): bool
     {
+        if ($oldPath === $newPath) {
+            return true;
+        }
+
         try {
             $this->adapter->move(
                 $this->normalizeRelativePath($oldPath, true),
