diff --git a/vendor/magento/module-catalog-import-export/Model/Import/Product/Validator/Backorders.php b/vendor/magento/module-catalog-import-export/Model/Import/Product/Validator/Backorders.php
new file mode 100644
index 0000000000000..23db673117b9d
--- /dev/null
+++ b/vendor/magento/module-catalog-import-export/Model/Import/Product/Validator/Backorders.php
@@ -0,0 +1,55 @@
+<?php
+/**
+ * Copyright 2025 Adobe
+ * All Rights Reserved.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogImportExport\Model\Import\Product\Validator;
+
+use Magento\CatalogImportExport\Model\Import\Product\RowValidatorInterface;
+use Magento\CatalogInventory\Model\Source\Backorders as BackordersSource;
+
+class Backorders extends AbstractImportValidator implements RowValidatorInterface
+{
+    /**
+     * @param BackordersSource $backordersSource
+     */
+    public function __construct(
+        private readonly BackordersSource $backordersSource
+    ) {
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function isValid($value)
+    {
+        $this->_clearMessages();
+        if (!isset($value['allow_backorders'])) {
+            return true;
+        }
+
+        $backorders = $value['allow_backorders'];
+        if (in_array($backorders, [$this->context->getEmptyAttributeValueConstant(), ''], true)) {
+            return true;
+        }
+
+        if (!is_numeric($backorders)) {
+            $messageTemplate = $this->context->retrieveMessageTemplate(self::ERROR_INVALID_ATTRIBUTE_TYPE);
+            $message = sprintf($messageTemplate, 'allow_backorders');
+            $this->_addMessages([$message]);
+            return false;
+        }
+
+        $allowedValues = array_column($this->backordersSource->toOptionArray(), 'label', 'value');
+        if (!isset($allowedValues[$backorders])) {
+            $messageTemplate = $this->context->retrieveMessageTemplate(self::ERROR_INVALID_ATTRIBUTE_OPTION);
+            $message = sprintf($messageTemplate, 'allow_backorders');
+            $this->_addMessages([$message]);
+            return false;
+        }
+
+        return true;
+    }
+}
diff --git a/vendor/magento/module-catalog-import-export/etc/di.xml b/vendor/magento/module-catalog-import-export/etc/di.xml
index 43fdda6227ac7..7447d73af539d 100644
--- a/vendor/magento/module-catalog-import-export/etc/di.xml
+++ b/vendor/magento/module-catalog-import-export/etc/di.xml
@@ -36,6 +36,7 @@
                 <item name="quantity" xsi:type="object">Magento\CatalogImportExport\Model\Import\Product\Validator\Quantity</item>
                 <item name="layout_update" xsi:type="object">Magento\CatalogImportExport\Model\Import\Product\Validator\LayoutUpdate</item>
                 <item name="layout_update_permissions" xsi:type="object">Magento\CatalogImportExport\Model\Import\Product\Validator\LayoutUpdatePermissions</item>
+                <item name="backorders" xsi:type="object">Magento\CatalogImportExport\Model\Import\Product\Validator\Backorders</item>
             </argument>
         </arguments>
     </type>
