diff --git a/vendor/magento/module-catalog/Model/ResourceModel/MediaImageDeleteProcessor.php b/vendor/magento/module-catalog/Model/ResourceModel/MediaImageDeleteProcessor.php
index f49ddef01ca7..cfe9fb2cceb0 100644
--- a/vendor/magento/module-catalog/Model/ResourceModel/MediaImageDeleteProcessor.php
+++ b/vendor/magento/module-catalog/Model/ResourceModel/MediaImageDeleteProcessor.php
@@ -101,7 +101,7 @@ public function execute(DataObject $product): void
      */
     private function canDeleteImage(string $file): bool
     {
-        return $this->productGallery->countImageUses($file) <= 1;
+        return $this->productGallery->countImageUses($file) < 1;
     }
 
     /**
