diff --git a/vendor/magento/module-inventory-cache/Model/FlushCacheByCacheTag.php b/vendor/magento/module-inventory-cache/Model/FlushCacheByCacheTag.php
index e742f05203c9..40c73b9f4f16 100644
--- a/vendor/magento/module-inventory-cache/Model/FlushCacheByCacheTag.php
+++ b/vendor/magento/module-inventory-cache/Model/FlushCacheByCacheTag.php
@@ -59,7 +59,10 @@ public function execute(string $cacheTag, array $entityIds): void
             $cacheContext = $this->cacheContextFactory->create();
             $cacheContext->registerEntities($cacheTag, $entityIds);
             $this->eventManager->dispatch('clean_cache_by_tags', ['object' => $cacheContext]);
-            $this->appCache->clean($cacheContext->getIdentities());
+            $tags = $cacheContext->getIdentities();
+            if ($tags) {
+                $this->appCache->clean($tags);
+            }
         }
     }
 }
