diff --git a/lib/web/fotorama/fotorama.js b/lib/web/fotorama/fotorama.js
index a60999ef7647..5abcc4c85164 100644
--- a/lib/web/fotorama/fotorama.js
+++ b/lib/web/fotorama/fotorama.js
@@ -1219,7 +1219,9 @@ fotoramaVersion = '4.6.4';
     }
 
     function stopEvent(e, stopPropagation) {
-        e.preventDefault ? e.preventDefault() : (e.returnValue = false);
+        if (!Modernizr.touch) {
+            e.preventDefault ? e.preventDefault() : (e.returnValue = false);
+        }
         stopPropagation && e.stopPropagation && e.stopPropagation();
     }
 
