diff --git a/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/Initializer/StockItem.php b/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/Initializer/StockItem.php
index f104552b4e0f..ea4c35de053b 100644
--- a/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/Initializer/StockItem.php
+++ b/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator/Initializer/StockItem.php
@@ -35,6 +35,8 @@ class StockItem
 
     /**
      * @var StockStateProviderInterface
+     * @deprecated
+     * @see was overriding ItemBackorders value with the Default Scope value; caused discrepancy in multistock config
      */
     private $stockStateProvider;
 
@@ -122,11 +124,6 @@ public function initialize(
             $quoteItem->setHasError(true);
         }
 
-        /* We need to ensure that any possible plugin will not erase the data */
-        $backOrdersQty = $this->stockStateProvider->checkQuoteItemQty($stockItem, $rowQty, $qtyForCheck, $qty)
-            ->getItemBackorders();
-        $result->setItemBackorders($backOrdersQty);
-
         if ($stockItem->hasIsChildItem()) {
             $stockItem->unsIsChildItem();
         }
