diff --git a/vendor/magento/framework/Filesystem/Driver/File.php b/vendor/magento/framework/Filesystem/Driver/File.php
index 1affad552137..4edb095f6c48 100644
--- a/vendor/magento/framework/Filesystem/Driver/File.php
+++ b/vendor/magento/framework/Filesystem/Driver/File.php
@@ -391,8 +391,8 @@ public function symlink($source, $destination, DriverInterface $targetDriver = n
      */
     public function deleteFile($path)
     {
-        $result = @unlink($this->getScheme() . $path);
-        if (!$result) {
+        @unlink($this->getScheme() . $path);
+        if ($this->isFile($path)) {
             throw new FileSystemException(
                 new Phrase(
                     'The "%1" file can\'t be deleted. %2',
@@ -400,7 +400,7 @@ public function deleteFile($path)
                 )
             );
         }
-        return $result;
+        return true;
     }
 
     /**
