diff --git a/vendor/magento/module-inventory-configurable-product-admin-ui/view/adminhtml/templates/catalog/product/edit/attribute/steps/bulk.phtml b/vendor/magento/module-inventory-configurable-product-admin-ui/view/adminhtml/templates/catalog/product/edit/attribute/steps/bulk.phtml
index 6c0b4a1df043..67f19c0de5b5 100644
--- a/vendor/magento/module-inventory-configurable-product-admin-ui/view/adminhtml/templates/catalog/product/edit/attribute/steps/bulk.phtml
+++ b/vendor/magento/module-inventory-configurable-product-admin-ui/view/adminhtml/templates/catalog/product/edit/attribute/steps/bulk.phtml
@@ -79,12 +79,7 @@
                         <div data-role="uploader" class="uploader">
                             <div class="image-browse">
                                 <span><?= $block->escapeHtml(__('Browse Files...')) ?></span>
-                                <input type="file"
-                                       id=""
-                                       name="image"
-                                       class="admin__control-file"
-                                       multiple="multiple"
-                                       data-url="<?= /* @noEscape */ $block->getUrl('catalog/product_gallery/upload') ?>" />
+                                <span class="browse-file" data-url="<?= /* @noEscape */ $block->getUrl('catalog/product_gallery/upload') ?>"></span>
                             </div>
                         </div>
                         <div class="product-image-wrapper">
@@ -351,8 +346,7 @@
                                     <span><?= $block->escapeHtml(
                                             __('Browse Files...')
                                         ); ?></span>
-                                        <input type="file" name="image" multiple="multiple"
-                                               data-url="<?= /* @noEscape */ $block->getUrl('catalog/product_gallery/upload') ?>" />
+                                        <span class="browse-file" data-url="<?= /* @noEscape */ $block->getUrl('catalog/product_gallery/upload') ?>"></span>
                                     </div>
                                 </div>
                                 <div class="product-image-wrapper">
