diff --git a/vendor/magento/module-customer/Model/Plugin/UpdateCustomer.php b/vendor/magento/module-customer/Model/Plugin/UpdateCustomer.php
index e8d4bdaec2d22..28b9bd5cc6516 100644
--- a/vendor/magento/module-customer/Model/Plugin/UpdateCustomer.php
+++ b/vendor/magento/module-customer/Model/Plugin/UpdateCustomer.php
@@ -63,7 +63,9 @@ public function beforeSave(
             $customerId === $customerSessionId
         ) {
             $customer = $this->getUpdatedCustomer($customerRepository->getById($customerId), $customer);
-        } elseif ($userType === UserContextInterface::USER_TYPE_ADMIN && $customerId) {
+        } elseif ($customerId && in_array($userType, [UserContextInterface::USER_TYPE_ADMIN,
+                UserContextInterface::USER_TYPE_INTEGRATION], true)
+        ) {
             $customer = $this->getUpdatedCustomer($customerRepository->getById($customerId), $customer);
         }
 

