diff --git a/vendor/magento/module-customer/Model/Address/AbstractAddress.php b/vendor/magento/module-customer/Model/Address/AbstractAddress.php
index f710ef6846fd..cbce80138b12 100644
--- a/vendor/magento/module-customer/Model/Address/AbstractAddress.php
+++ b/vendor/magento/module-customer/Model/Address/AbstractAddress.php
@@ -451,6 +451,9 @@ public function getRegionId()
                 (string)$this->getRegionCode(),
                 (string)$this->getCountryId()
             );
+            if (empty($regionId)) {
+                $regionId = $this->getData('region_id');
+            }
             $this->setData('region_id', $regionId);
         }
 
