diff --git a/vendor/magento/module-review-graph-ql/Model/DataProvider/CustomerReviewsDataProvider.php b/vendor/magento/module-review-graph-ql/Model/DataProvider/CustomerReviewsDataProvider.php
index 42adc8009c0..f25e32575c7 100644
--- a/vendor/magento/module-review-graph-ql/Model/DataProvider/CustomerReviewsDataProvider.php
+++ b/vendor/magento/module-review-graph-ql/Model/DataProvider/CustomerReviewsDataProvider.php
@@ -36,10 +36,10 @@ class CustomerReviewsDataProvider
      * @param int $customerId
      * @param int $currentPage
      * @param int $pageSize
-     *
+     * @param int $storeId
      * @return ReviewsCollection
      */
-    public function getData(int $customerId, int $currentPage, int $pageSize): ReviewsCollection
+    public function getData(int $customerId, int $currentPage, int $pageSize, int $storeId): ReviewsCollection
     {
         /** @var ReviewsCollection $reviewsCollection */
         $reviewsCollection = $this->collectionFactory->create();
@@ -47,6 +47,7 @@ class CustomerReviewsDataProvider
             ->addCustomerFilter($customerId)
             ->setPageSize($pageSize)
             ->setCurPage($currentPage)
+            ->addStoreFilter($storeId)
             ->setDateOrder();
         $reviewsCollection->getSelect()->join(
             ['cpe' => $reviewsCollection->getTable('catalog_product_entity')],
diff --git a/vendor/magento/module-review-graph-ql/Model/Resolver/Customer/Reviews.php b/vendor/magento/module-review-graph-ql/Model/Resolver/Customer/Reviews.php
index 8c0bca63f8e..b177c915275 100644
--- a/vendor/magento/module-review-graph-ql/Model/Resolver/Customer/Reviews.php
+++ b/vendor/magento/module-review-graph-ql/Model/Resolver/Customer/Reviews.php
@@ -82,7 +82,8 @@ class Reviews implements ResolverInterface
         $reviewsCollection = $this->customerReviewsDataProvider->getData(
             (int) $context->getUserId(),
             $args['currentPage'],
-            $args['pageSize']
+            $args['pageSize'],
+            (int) $context->getExtensionAttributes()->getStore()->getId()
         );

         return $this->aggregatedReviewsDataProvider->getData($reviewsCollection);
