diff --git a/vendor/magento/module-sales/Model/ResourceModel/Collection/ExpiredQuotesCollection.php b/vendor/magento/module-sales/Model/ResourceModel/Collection/ExpiredQuotesCollection.php
index 895d73cc4cfff..05d7005f48d31 100644
--- a/vendor/magento/module-sales/Model/ResourceModel/Collection/ExpiredQuotesCollection.php
+++ b/vendor/magento/module-sales/Model/ResourceModel/Collection/ExpiredQuotesCollection.php
@@ -14,9 +14,6 @@
 use Magento\Store\Api\Data\StoreInterface;
 use Magento\Store\Model\ScopeInterface;
 
-/**
- * Class ExpiredQuotesCollection
- */
 class ExpiredQuotesCollection
 {
     /**
@@ -71,8 +68,8 @@ public function getExpiredQuotes(StoreInterface $store): AbstractCollection
 
         /** @var $quotes Collection */
         $quotes = $this->quoteCollectionFactory->create();
-        $quotes->addFieldToFilter('store_id', $store->getId());
-        $quotes->addFieldToFilter('updated_at', ['to' => date("Y-m-d", time() - $lifetime)]);
+        $quotes->addFieldToFilter('main_table.store_id', $store->getId());
+        $quotes->addFieldToFilter('main_table.updated_at', ['to' => date("Y-m-d", time() - $lifetime)]);
 
         return $quotes;
     }
