diff --git a/vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php b/vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php
index 404fd27232b..6f8564d17a9 100644
--- a/vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php
+++ b/vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php
@@ -166,6 +166,7 @@ abstract class AbstractAction
      */
     abstract public function execute($ids);
 
+    // phpcs:disable
     /**
      * Synchronize data between index storage and original storage
      *
@@ -197,6 +198,7 @@ abstract class AbstractAction
         return $this;
     }
 
+    // phpcs:enable
     /**
      * Prepare website current dates table
      *
@@ -432,6 +434,7 @@ abstract class AbstractAction
      */
     private function deleteIndexData(array $entityIds)
     {
+        $entityIds = array_unique(array_map('intval', $entityIds));
         foreach ($this->dimensionCollectionFactory->create() as $dimensions) {
             $select = $this->getConnection()->select()->from(
                 ['index_price' => $this->tableMaintainer->getMainTableByDimensions($dimensions)],
@@ -442,6 +445,7 @@ abstract class AbstractAction
         }
     }
 
+    // phpcs:disable
     /**
      * Copy relations product index from primary index to temporary index table by parent entity
      *
@@ -491,6 +495,7 @@ abstract class AbstractAction
         return $this;
     }
 
+    // phpcs:enable
     /**
      * Retrieve index table by dimension that will be used for write operations.
      *
