diff --git a/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php b/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php
index 8c60feefb6431..e0122a00a656f 100644
--- a/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php
+++ b/vendor/magento/module-quote-graph-ql/Model/GetDiscounts.php
@@ -22,6 +22,8 @@ class GetDiscounts
      */
     public function execute(Quote $quote, array $discounts): ?array
     {
+        $discounts = $discounts ?: $quote->getBillingAddress()->getExtensionAttributes()->getDiscounts() ?? [];
+
         if (empty($discounts)) {
             return null;
         }
