diff --git a/vendor/magento/module-elasticsearch/Model/Adapter/FieldMapper/Product/FieldProvider/StaticField.php b/vendor/magento/module-elasticsearch/Model/Adapter/FieldMapper/Product/FieldProvider/StaticField.php
index db03abac553..b9d4c339364 100644
--- a/vendor/magento/module-elasticsearch/Model/Adapter/FieldMapper/Product/FieldProvider/StaticField.php
+++ b/vendor/magento/module-elasticsearch/Model/Adapter/FieldMapper/Product/FieldProvider/StaticField.php
@@ -164,7 +164,8 @@ class StaticField implements FieldProviderInterface
                 ),
                 'index' => $this->indexTypeConverter->convert(
                     IndexTypeConverterInterface::INTERNAL_NO_ANALYZE_VALUE
-                )
+                ),
+                'normalizer' => 'folding',
             ];
         }

@@ -202,7 +203,8 @@ class StaticField implements FieldProviderInterface
                     ),
                     'index' => $this->indexTypeConverter->convert(
                         IndexTypeConverterInterface::INTERNAL_NO_ANALYZE_VALUE
-                    )
+                    ),
+                    'normalizer' => 'folding',
                 ];
             }
         }
diff --git a/vendor/magento/module-elasticsearch/Model/Adapter/Index/Builder.php b/vendor/magento/module-elasticsearch/Model/Adapter/Index/Builder.php
index df0aab23914..2b4e5c3ef27 100644
--- a/vendor/magento/module-elasticsearch/Model/Adapter/Index/Builder.php
+++ b/vendor/magento/module-elasticsearch/Model/Adapter/Index/Builder.php
@@ -105,6 +105,12 @@ class Builder implements BuilderInterface
                 'tokenizer' => $tokenizer,
                 'filter' => array_merge($filter, $synonymFilter),
                 'char_filter' => $charFilter,
+                'normalizer' => [
+                    'folding' => [
+                        'type' => 'custom',
+                        'filter' => ['asciifolding', 'lowercase'],
+                    ],
+                ],
             ],
         ];


