diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemErrors.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemErrors.php
index 13636e46651..006c105661f 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemErrors.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemErrors.php
@@ -70,7 +70,7 @@ class CartItemErrors implements ResolverInterface
         $errors = [];
         foreach ($cartItem->getErrorInfos() as $error) {
             $errorType = $error['code'] ?? self::ERROR_UNDEFINED;
-            $message = $error['message'] ?? $cartItem->getMessage();
+            $message = (string) ($error['message'] ?? $cartItem->getMessage());
             $errorEnumCode = $this->enumLookup->getEnumValueFromField(
                 'CartItemErrorType',
                 (string)$errorType
