diff --git a/vendor/magento/module-sales-rule/Controller/Adminhtml/Promo/Quote/Save.php b/vendor/magento/module-sales-rule/Controller/Adminhtml/Promo/Quote/Save.php
index e407142e689..aefadd0c029 100644
--- a/vendor/magento/module-sales-rule/Controller/Adminhtml/Promo/Quote/Save.php
+++ b/vendor/magento/module-sales-rule/Controller/Adminhtml/Promo/Quote/Save.php
@@ -64,7 +64,7 @@ class Save extends \Magento\SalesRule\Controller\Adminhtml\Promo\Quote implement
     {
         $data = $this->getRequest()->getPostValue();
         if ($data) {
-            $data['simple_free_shipping'] = ($data['simple_free_shipping'] === '')
+            $data['simple_free_shipping'] = (($data['simple_free_shipping'] ?? '') === '')
                     ? null : $data['simple_free_shipping'];
 
             try {
