diff --git a/vendor/magento/module-quote/Model/Quote/Item/CartItemPersister.php b/vendor/magento/module-quote/Model/Quote/Item/CartItemPersister.php
index 41651d3c788..5481f57695d 100644
--- a/vendor/magento/module-quote/Model/Quote/Item/CartItemPersister.php
+++ b/vendor/magento/module-quote/Model/Quote/Item/CartItemPersister.php
@@ -75,7 +75,7 @@ class CartItemPersister
                 $buyRequestData = $this->cartItemOptionProcessor->getBuyRequest($productType, $item);
                 if (is_object($buyRequestData)) {
                     /** Update item product options */
-                    if ($currentItem->getQty() !== $buyRequestData->getQty()) {
+                    if ($quote->getIsActive()) {
                         $item = $quote->updateItem($itemId, $buyRequestData);
                     }
                 } else {
