diff --git a/vendor/magento/module-paypal/Controller/Express/AbstractExpress/PlaceOrder.php b/vendor/magento/module-paypal/Controller/Express/AbstractExpress/PlaceOrder.php
index 95dc8ee487e..54bc4381cd2 100644
--- a/vendor/magento/module-paypal/Controller/Express/AbstractExpress/PlaceOrder.php
+++ b/vendor/magento/module-paypal/Controller/Express/AbstractExpress/PlaceOrder.php
@@ -112,6 +112,14 @@ class PlaceOrder extends \Magento\Paypal\Controller\Express\AbstractExpress
                     ->setLastOrderStatus($order->getStatus());
             }
 
+            $this->_eventManager->dispatch(
+                'checkout_submit_all_after',
+                [
+                    'order' => $order,
+                    'quote' => $this->_getQuote()
+                ]
+            );
+
             $this->_eventManager->dispatch(
                 'paypal_express_place_order_success',
                 [
diff --git a/vendor/magento/module-paypal/Controller/Express/OnAuthorization.php b/vendor/magento/module-paypal/Controller/Express/OnAuthorization.php
index 0d0716e87bf..4dc12d8878d 100644
--- a/vendor/magento/module-paypal/Controller/Express/OnAuthorization.php
+++ b/vendor/magento/module-paypal/Controller/Express/OnAuthorization.php
@@ -139,6 +139,14 @@ class OnAuthorization extends AbstractExpress implements HttpPostActionInterface
                     ->setLastRealOrderId($order->getIncrementId())
                     ->setLastOrderStatus($order->getStatus());
 
+                $this->_eventManager->dispatch(
+                    'checkout_submit_all_after',
+                    [
+                        'order' => $order,
+                        'quote' => $quote
+                    ]
+                );
+
                 $this->_eventManager->dispatch(
                     'paypal_express_place_order_success',
                     [
