diff --git a/vendor/magento/module-open-search/Model/Adapter/DynamicTemplates/PriceMapper.php b/vendor/magento/module-open-search/Model/Adapter/DynamicTemplates/PriceMapper.php
index 03ced99cc632..545d8b95c006 100644
--- a/vendor/magento/module-open-search/Model/Adapter/DynamicTemplates/PriceMapper.php
+++ b/vendor/magento/module-open-search/Model/Adapter/DynamicTemplates/PriceMapper.php
@@ -19,7 +19,8 @@ public function processTemplates(array $templates): array
     {
         $templates[] = [
             'price_mapping' => [
-                'match' => 'price_*',
+                "match_pattern" => "regex",
+                'match' => 'price_\\d+_\\d+',
                 'match_mapping_type' => 'string',
                 'mapping' => [
                     'type' => 'double',

