diff --git a/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Cancel.php b/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Cancel.php
index 14c52ba7f66e6..2aa454ca89e59 100644
--- a/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Cancel.php
+++ b/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Cancel.php
@@ -1,15 +1,15 @@
 <?php
 /**
- *
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\Sales\Controller\Adminhtml\Order\Invoice;
 
 use Magento\Framework\App\Action\HttpPostActionInterface;
+use Magento\Framework\App\Action\HttpGetActionInterface;
 use Magento\Sales\Controller\Adminhtml\Invoice\AbstractInvoice\View;
 
-class Cancel extends View implements HttpPostActionInterface
+class Cancel extends View implements HttpPostActionInterface, HttpGetActionInterface
 {
     /**
      * Authorization level of a basic admin session
diff --git a/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Capture.php b/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Capture.php
index ea53a8b9bc84a..2d4f2191f0c24 100644
--- a/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Capture.php
+++ b/vendor/magento/module-sales/Controller/Adminhtml/Order/Invoice/Capture.php
@@ -1,15 +1,14 @@
 <?php
 /**
- *
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\Sales\Controller\Adminhtml\Order\Invoice;
 
-use Magento\Framework\App\Action\HttpPostActionInterface;
+use Magento\Framework\App\Action\HttpGetActionInterface;
 use Magento\Sales\Controller\Adminhtml\Invoice\AbstractInvoice\View;
 
-class Capture extends View implements HttpPostActionInterface
+class Capture extends View implements HttpGetActionInterface
 {
     /**
      * Authorization level of a basic admin session
