diff --git a/vendor/magento/module-quote/Model/QuoteManagement.php b/vendor/magento/module-quote/Model/QuoteManagement.php
index 1d4b8feba07..13ad51697da 100644
--- a/vendor/magento/module-quote/Model/QuoteManagement.php
+++ b/vendor/magento/module-quote/Model/QuoteManagement.php
@@ -388,7 +388,10 @@ class QuoteManagement implements \Magento\Quote\Api\CartManagementInterface
 
         if ($quote->getCheckoutMethod() === self::METHOD_GUEST) {
             $quote->setCustomerId(null);
-            $quote->setCustomerEmail($quote->getBillingAddress()->getEmail());
+            $billingAddress = $quote->getBillingAddress();
+            if (!$quote->getCustomerEmail()) {
+                $quote->setCustomerEmail($billingAddress ? $billingAddress->getEmail() : null);
+            }
             if ($quote->getCustomerFirstname() === null && $quote->getCustomerLastname() === null) {
                 $quote->setCustomerFirstname($quote->getBillingAddress()->getFirstname());
                 $quote->setCustomerLastname($quote->getBillingAddress()->getLastname());
