diff --git a/vendor/magento/framework/View/Asset/File.php b/vendor/magento/framework/View/Asset/File.php
index 0f31ad9ef22f7..e359f1aa402a3 100644
--- a/vendor/magento/framework/View/Asset/File.php
+++ b/vendor/magento/framework/View/Asset/File.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
  */
 
 namespace Magento\Framework\View\Asset;
@@ -16,6 +16,8 @@
  */
 class File implements MergeableInterface
 {
+    private const MAX_READ_ATTEMPTS = 3;
+
     /**
      * @var string
      */
@@ -183,11 +185,27 @@ public function getSourceContentType()
      */
     public function getContent()
     {
-        $content = $this->source->getContent($this);
-        if (false === $content) {
-            throw new File\NotFoundException("Unable to get content for '{$this->getPath()}'");
+        $attempt    = 0;
+        while ($attempt < self::MAX_READ_ATTEMPTS) {
+            $attempt++;
+            $content = trim($this->source->getContent($this));
+
+            if ($content) {
+                return $content;
+            }
+
+            if ($attempt < self::MAX_READ_ATTEMPTS) {
+                usleep(random_int(10000, 1000000));
+            }
         }
-        return $content;
+
+        throw new File\NotFoundException(
+            sprintf(
+                "Unable to get content for '%s' after %d attempts.",
+                $this->getPath(),
+                self::MAX_READ_ATTEMPTS
+            )
+        );
     }
 
     /**
