diff --git a/vendor/magento/framework/Setup/Declaration/Schema/Db/SchemaBuilder.php b/vendor/magento/framework/Setup/Declaration/Schema/Db/SchemaBuilder.php
index 5d8f59cb4a949..de4f3dcc95109 100644
--- a/vendor/magento/framework/Setup/Declaration/Schema/Db/SchemaBuilder.php
+++ b/vendor/magento/framework/Setup/Declaration/Schema/Db/SchemaBuilder.php
@@ -86,6 +86,7 @@ public function build(Schema $schema)
         $tablesWithJsonTypeField = [];
         if (isset($data['table'])) {
             foreach ($data['table'] as $keyTable => $tableColumns) {
+                $tableColumns['column'] ??= [];
                 foreach ($tableColumns['column'] as $keyColumn => $columnData) {
                     if ($columnData['type'] == 'json') {
                         $tablesWithJsonTypeField[$keyTable] = $keyColumn;
