diff --git a/vendor/magento/module-catalog-widget/Block/Product/ProductsList.php b/vendor/magento/module-catalog-widget/Block/Product/ProductsList.php
index 7e6693ce68e..d35590efc93 100644
--- a/vendor/magento/module-catalog-widget/Block/Product/ProductsList.php
+++ b/vendor/magento/module-catalog-widget/Block/Product/ProductsList.php
@@ -43,28 +43,27 @@ class ProductsList extends AbstractProduct implements BlockInterface, IdentityIn
     /**
      * Default value for products count that will be shown
      */
-    const DEFAULT_PRODUCTS_COUNT = 10;
+    public const DEFAULT_PRODUCTS_COUNT = 10;
 
     /**
      * Name of request parameter for page number value
      *
-     * @deprecated @see $this->getData('page_var_name')
+     * @deprecated
+     * @see $this->getData('page_var_name')
      */
-    const PAGE_VAR_NAME = 'np';
+    public const PAGE_VAR_NAME = 'np';
 
     /**
      * Default value for products per page
      */
-    const DEFAULT_PRODUCTS_PER_PAGE = 5;
+    public const DEFAULT_PRODUCTS_PER_PAGE = 5;
 
     /**
      * Default value whether show pager or not
      */
-    const DEFAULT_SHOW_PAGER = false;
+    public const DEFAULT_SHOW_PAGER = false;
 
     /**
-     * Instance of pager block
-     *
      * @var Pager
      */
     protected $pager;
@@ -75,15 +74,11 @@ class ProductsList extends AbstractProduct implements BlockInterface, IdentityIn
     protected $httpContext;
 
     /**
-     * Catalog product visibility
-     *
      * @var Visibility
      */
     protected $catalogProductVisibility;
 
     /**
-     * Product collection factory
-     *
      * @var CollectionFactory
      */
     protected $productCollectionFactory;
@@ -223,6 +218,7 @@ class ProductsList extends AbstractProduct implements BlockInterface, IdentityIn
             $this->_storeManager->getStore()->getId(),
             $this->_design->getDesignTheme()->getId(),
             $this->httpContext->getValue(\Magento\Customer\Model\Context::CONTEXT_GROUP),
+            $this->json->serialize($this->httpContext->getValue('tax_rates')),
             (int)$this->getRequest()->getParam($this->getData('page_var_name'), 1),
             $this->getProductsPerPage(),
             $this->getProductsCount(),
@@ -537,7 +533,8 @@ class ProductsList extends AbstractProduct implements BlockInterface, IdentityIn
      * Get currency of product
      *
      * @return PriceCurrencyInterface
-     * @deprecated 100.2.0
+     * @deprecated
+     * @see Constructor injection
      */
     private function getPriceCurrency()
     {
