diff --git a/vendor/magento/module-sales-rule/Model/Rule/Action/Discount/ByFixed.php b/vendor/magento/module-sales-rule/Model/Rule/Action/Discount/ByFixed.php
index 2ad935247a8ab..b067e27838e8e 100644
--- a/vendor/magento/module-sales-rule/Model/Rule/Action/Discount/ByFixed.php
+++ b/vendor/magento/module-sales-rule/Model/Rule/Action/Discount/ByFixed.php
@@ -27,8 +27,9 @@ public function calculate($rule, $item, $qty)
         $itemPrice = $this->validator->getItemPrice($item);
         $baseItemPrice = $this->validator->getItemBasePrice($item);
 
-        $discountAmountMin = min(($itemPrice * $qty) - $itemDiscountAmount, $discountAmount * $qty);
-        $baseDiscountAmountMin = min(($baseItemPrice * $qty) - $itemBaseDiscountAmount, $baseDiscountAmount * $qty);
+        $discountAmountMin = min(($itemPrice * $item->getQty()) - $itemDiscountAmount, $discountAmount * $qty);
+        $baseDiscountAmountMin =
+            min(($baseItemPrice * $item->getQty()) - $itemBaseDiscountAmount, $baseDiscountAmount * $qty);
 
         $discountData->setAmount($discountAmountMin);
         $discountData->setBaseAmount($baseDiscountAmountMin);
diff --git a/vendor/magento/module-sales-rule/Model/Utility.php b/vendor/magento/module-sales-rule/Model/Utility.php
index e91bde82eac0e..9a1deeaf43901 100644
--- a/vendor/magento/module-sales-rule/Model/Utility.php
+++ b/vendor/magento/module-sales-rule/Model/Utility.php
@@ -8,6 +8,9 @@
 
 use Magento\Framework\Pricing\PriceCurrencyInterface;
 
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
 class Utility
 {
     /**
@@ -152,13 +155,11 @@ public function canProcessRule($rule, $address)
      *
      * @param \Magento\SalesRule\Model\Rule\Action\Discount\Data $discountData
      * @param \Magento\Quote\Model\Quote\Item\AbstractItem $item
-     * @param float $qty
      * @return void
      */
     public function minFix(
         \Magento\SalesRule\Model\Rule\Action\Discount\Data $discountData,
-        \Magento\Quote\Model\Quote\Item\AbstractItem $item,
-        $qty
+        \Magento\Quote\Model\Quote\Item\AbstractItem $item
     ) {
         $itemPrice = $this->getItemPrice($item);
         $baseItemPrice = $this->getItemBasePrice($item);
@@ -166,8 +167,9 @@ public function minFix(
         $itemDiscountAmount = $item->getDiscountAmount();
         $itemBaseDiscountAmount = $item->getBaseDiscountAmount();
 
-        $discountAmount = min($itemDiscountAmount + $discountData->getAmount(), $itemPrice * $qty);
-        $baseDiscountAmount = min($itemBaseDiscountAmount + $discountData->getBaseAmount(), $baseItemPrice * $qty);
+        $discountAmount = min($itemDiscountAmount + $discountData->getAmount(), $itemPrice * $item->getQty());
+        $baseDiscountAmount =
+            min($itemBaseDiscountAmount + $discountData->getBaseAmount(), $baseItemPrice * $item->getQty());
 
         $discountData->setAmount($discountAmount);
         $discountData->setBaseAmount($baseDiscountAmount);
