diff --git a/vendor/magento/module-quote/Plugin/SendOrderNotification.php b/vendor/magento/module-quote/Plugin/SendOrderNotification.php
new file mode 100644
index 000000000000..f7189ac0492d
--- /dev/null
+++ b/vendor/magento/module-quote/Plugin/SendOrderNotification.php
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\Quote\Plugin;
+
+use Magento\Framework\Event\Observer;
+use Magento\Framework\App\RequestInterface;
+use Magento\Quote\Observer\SubmitObserver;
+use Magento\Sales\Model\Order;
+
+/**
+ * Send admin order confirmation
+ */
+class SendOrderNotification
+{
+    /**
+     * @var RequestInterface $request
+     */
+    private RequestInterface $request;
+
+    /**
+     * @param RequestInterface $request
+     */
+    public function __construct(RequestInterface $request)
+    {
+        $this->request = $request;
+    }
+
+    /**
+     * Adjusts order flag for confirmation email delivery
+     *
+     * @param SubmitObserver $subject
+     * @param Observer $observer
+     * @return Observer[]
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeExecute(SubmitObserver $subject, Observer $observer): array
+    {
+        /** @var  Order $order */
+        $order = $observer->getEvent()->getOrder();
+        $requestInfo = $this->request->getParam('order');
+        if (!empty($requestInfo)) {
+            $order->setCanSendNewEmailFlag((bool)($requestInfo['send_confirmation'] ?? 0));
+        }
+
+        return [$observer];
+    }
+}
diff --git a/vendor/magento/module-quote/etc/adminhtml/di.xml b/vendor/magento/module-quote/etc/adminhtml/di.xml
index 08e87c7db9ec..58368a445393 100644
--- a/vendor/magento/module-quote/etc/adminhtml/di.xml
+++ b/vendor/magento/module-quote/etc/adminhtml/di.xml
@@ -17,4 +17,7 @@
             </argument>
         </arguments>
     </type>
-</config>
\ No newline at end of file
+    <type name="Magento\Quote\Observer\SubmitObserver">
+        <plugin name="beforeSendOrderNotification" type="Magento\Quote\Plugin\SendOrderNotification" sortOrder="1" disabled="false" />
+    </type>
+</config>
