diff --git a/vendor/magento/framework/Mview/View/ChangelogBatchWalker.php b/vendor/magento/framework/Mview/View/ChangelogBatchWalker.php
index 61956b0cd3ae..474c687a1381 100644
--- a/vendor/magento/framework/Mview/View/ChangelogBatchWalker.php
+++ b/vendor/magento/framework/Mview/View/ChangelogBatchWalker.php
@@ -24,30 +24,34 @@
 class ChangelogBatchWalker implements ChangelogBatchWalkerInterface
 {
     /**
-     * @var \Magento\Framework\App\ResourceConnection
+     * @var ResourceConnection
      */
     private ResourceConnection $resourceConnection;
+
     /**
-     * @var \Magento\Framework\DB\Query\Generator
+     * @var Generator
      */
     private Generator $generator;
+
     /**
-     * @var \Magento\Framework\Mview\View\ChangelogBatchWalker\IdsTableBuilderInterface
+     * @var IdsTableBuilderInterface
      */
     private IdsTableBuilderInterface $idsTableBuilder;
+
     /**
-     * @var \Magento\Framework\Mview\View\ChangelogBatchWalker\IdsSelectBuilderInterface
+     * @var IdsSelectBuilderInterface
      */
     private IdsSelectBuilderInterface $idsSelectBuilder;
+
     /**
-     * @var \Magento\Framework\Mview\View\ChangelogBatchWalker\IdsFetcherInterface
+     * @var IdsFetcherInterface
      */
     private IdsFetcherInterface $idsFetcher;
 
     /**
      * @param ResourceConnection $resourceConnection
-     * @param \Magento\Framework\DB\Query\Generator $generator
-     * @param \Magento\Framework\Mview\View\ChangelogBatchWalker\IdsContext $idsContext
+     * @param Generator $generator
+     * @param IdsContext $idsContext
      */
     public function __construct(
         ResourceConnection $resourceConnection,
@@ -80,7 +84,7 @@ public function walk(
         $idsTable = $this->idsTableBuilder->build($changelog);
 
         try {
-            $connection->createTemporaryTable($idsTable);
+            $connection->createTable($idsTable);
 
             $columns = $this->getIdsColumns($idsTable);
 
@@ -122,7 +126,7 @@ public function walk(
                 yield $ids;
             }
         } finally {
-            $connection->dropTemporaryTable($idsTable->getName());
+            $connection->dropTable($idsTable->getName());
         }
     }
 
diff --git a/vendor/magento/framework/Mview/View/ChangelogBatchWalker/IdsTableBuilder.php b/vendor/magento/framework/Mview/View/ChangelogBatchWalker/IdsTableBuilder.php
index 69b1527fe661..fe0753144dca 100644
--- a/vendor/magento/framework/Mview/View/ChangelogBatchWalker/IdsTableBuilder.php
+++ b/vendor/magento/framework/Mview/View/ChangelogBatchWalker/IdsTableBuilder.php
@@ -50,7 +50,6 @@ public function build(ChangelogInterface $changelog): Table
             ['unsigned' => true, 'nullable' => false],
             'Entity ID'
         );
-        $table->setOption('type', 'memory');
         $table->addIndex(
             self::INDEX_NAME_UNIQUE,
             [
