diff --git a/vendor/magento/module-bundle/Model/ResourceModel/Selection/Collection.php b/vendor/magento/module-bundle/Model/ResourceModel/Selection/Collection.php
index 303c33b571d3..ce32b2b54687 100644
--- a/vendor/magento/module-bundle/Model/ResourceModel/Selection/Collection.php
+++ b/vendor/magento/module-bundle/Model/ResourceModel/Selection/Collection.php
@@ -306,6 +306,9 @@ public function getNewEmptyItem()
     public function addPriceFilter($product, $searchMin, $useRegularPrice = false)
     {
         if ($product->getPriceType() == \Magento\Bundle\Model\Product\Price::PRICE_TYPE_DYNAMIC) {
+            if (!$this->getStoreId()) {
+                $this->setStoreId($this->_storeManager->getStore()->getId());
+            }
             $this->addPriceData();
             if ($useRegularPrice) {
                 $minimalPriceExpression = self::INDEX_TABLE_ALIAS . '.price';
