diff --git a/vendor/magento/module-rule/Model/Condition/AbstractCondition.php b/vendor/magento/module-rule/Model/Condition/AbstractCondition.php
index d5d3b80..7856629 100644
--- a/vendor/magento/module-rule/Model/Condition/AbstractCondition.php
+++ b/vendor/magento/module-rule/Model/Condition/AbstractCondition.php
@@ -885,8 +885,7 @@ abstract class AbstractCondition extends \Magento\Framework\DataObject implement
      */
     protected function _compareValues($validatedValue, $value, $strict = true)
     {
-        if (null === $value || null === $validatedValue ||
-            $strict && is_numeric($validatedValue) && is_numeric($value)) {
+        if ($strict && is_numeric($validatedValue) && is_numeric($value)) {
             return $validatedValue == $value;
         }
 
@@ -894,7 +893,7 @@ abstract class AbstractCondition extends \Magento\Framework\DataObject implement
         if ($strict) {
             $validatePattern = '^' . $validatePattern . '$';
         }
-        return (bool)preg_match('~' . $validatePattern . '~iu', $value);
+        return (bool)preg_match('~' . $validatePattern . '~iu', (string)$value);
     }
 
     /**
