diff --git a/vendor/magento/module-quote-graph-ql/Model/Cart/GetCartForCheckout.php b/vendor/magento/module-quote-graph-ql/Model/Cart/GetCartForCheckout.php
index 4b2d1afdea0..bd5f3d8cbaf 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Cart/GetCartForCheckout.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Cart/GetCartForCheckout.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2022 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);
 
@@ -57,7 +57,7 @@ class GetCartForCheckout
         try {
             $cart = $this->getCartForUser->execute($cartHash, $customerId, $storeId);
         } catch (NoSuchEntityException $e) {
-            throw new GraphQlNoSuchEntityException(__($e->getMessage()), $e);
+            throw new GraphQlNoSuchEntityException(__($e->getMessage()), $e, $e->getCode());
         }
         $this->checkoutAllowance->execute($cart);
 
