diff --git a/vendor/magento/module-reports/Block/Adminhtml/Grid/Column/Renderer/Currency.php b/vendor/magento/module-reports/Block/Adminhtml/Grid/Column/Renderer/Currency.php
index 1ebfd64b2f3..86746d74211 100644
--- a/vendor/magento/module-reports/Block/Adminhtml/Grid/Column/Renderer/Currency.php
+++ b/vendor/magento/module-reports/Block/Adminhtml/Grid/Column/Renderer/Currency.php
@@ -141,10 +141,10 @@ class Currency extends BackendCurrency
         $catalogPriceScope = $this->getCatalogPriceScope();
         $adminCurrencyCode = $this->getAdminCurrencyCode();

-        if (($catalogPriceScope != 0
+        if (((int)$catalogPriceScope !== 0
             && $adminCurrencyCode !== $currencyCode)) {
-            $storeCurrency = $this->currencyFactory->create()->load($adminCurrencyCode);
-            $currencyRate = $storeCurrency->getRate($currencyCode);
+            $currency = $this->currencyFactory->create()->load($adminCurrencyCode);
+            $currencyRate = $currency->getAnyRate($currencyCode);
         } else {
             $currencyRate = $this->_getRate($row);
         }
