diff --git a/vendor/magento/module-sales-rule/Model/Rule/Condition/Product/Subselect.php b/vendor/magento/module-sales-rule/Model/Rule/Condition/Product/Subselect.php
index ccec956d62060..1591287110e58 100644
--- a/vendor/magento/module-sales-rule/Model/Rule/Condition/Product/Subselect.php
+++ b/vendor/magento/module-sales-rule/Model/Rule/Condition/Product/Subselect.php
@@ -207,11 +207,11 @@ private function validateSubSelectConditions(mixed $item): bool
      *
      * @param mixed $item
      * @param mixed $attr
-     * @param int $total
-     * @return int|mixed
+     * @param float $total
+     * @return float
      * @SuppressWarnings(PHPMD.CyclomaticComplexity)
      */
-    private function getBaseRowTotalForChildrenProduct(mixed $item, mixed $attr, int $total): mixed
+    private function getBaseRowTotalForChildrenProduct(mixed $item, mixed $attr, float $total): float
     {
         $hasValidChild = false;
         $useChildrenTotal = ($item->getProductType() == Type::TYPE_BUNDLE);

