diff --git a/vendor/magento/module-inventory-in-store-pickup-frontend/view/frontend/web/js/model/pickup-address-converter.js b/vendor/magento/module-inventory-in-store-pickup-frontend/view/frontend/web/js/model/pickup-address-converter.js
index 28304a6b37a6..c7a3c1ea5de9 100644
--- a/vendor/magento/module-inventory-in-store-pickup-frontend/view/frontend/web/js/model/pickup-address-converter.js
+++ b/vendor/magento/module-inventory-in-store-pickup-frontend/view/frontend/web/js/model/pickup-address-converter.js
@@ -39,6 +39,15 @@ define(['underscore'], function (_) {
                         return false;
                     },
 
+                    /**
+                     * Returns if address editing is allowed.
+                     *
+                     * @return {Boolean}
+                     */
+                    isEditable: function () {
+                        return false;
+                    },
+
                     /**
                      * Returns address type
                      *
