diff --git a/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/MediaGalleryProcessor.php b/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/MediaGalleryProcessor.php
index e7de602292675..84cd3ae495747 100644
--- a/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/MediaGalleryProcessor.php
+++ b/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/MediaGalleryProcessor.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2019 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);
 
@@ -51,7 +51,9 @@ public function process(
         array $attributeNames,
         ?ContextInterface $context = null
     ): Collection {
-        if (in_array('media_gallery_entries', $attributeNames)) {
+        if (in_array('media_gallery_entries', $attributeNames) ||
+            in_array('media_gallery', $attributeNames)
+        ) {
             $mediaAttributes = $this->mediaConfig->getMediaAttributeCodes();
             foreach ($mediaAttributes as $mediaAttribute) {
                 if (!in_array($mediaAttribute, $attributeNames)) {
diff --git a/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls b/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls
index e5dc1d6d3894a..ce1db453253e4 100644
--- a/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls
+++ b/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls
@@ -229,6 +229,7 @@ interface MediaGalleryInterface @doc(description: "Contains basic information ab
     label: String @doc(description: "The label of the product image or video.")
     position: Int @doc(description: "The media item's position after it has been sorted.")
     disabled: Boolean @doc(description: "Indicates whether the image is hidden from view.")
+    types: [String] @doc(description: "Array of image types. It can have the following values: image, small_image, thumbnail.")
 }
 
 type ProductImage implements MediaGalleryInterface @doc(description: "Contains product image information, including the image URL and label.") {

