diff --git a/vendor/magento/module-elasticsearch/ElasticAdapter/SearchAdapter/Aggregation/Interval.php b/vendor/magento/module-elasticsearch/ElasticAdapter/SearchAdapter/Aggregation/Interval.php
index 6f18a2d45d83d..af6a9b6a13aa8 100644
--- a/vendor/magento/module-elasticsearch/ElasticAdapter/SearchAdapter/Aggregation/Interval.php
+++ b/vendor/magento/module-elasticsearch/ElasticAdapter/SearchAdapter/Aggregation/Interval.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2015 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);
 
@@ -146,7 +146,11 @@ public function loadPrevious($data, $index, $lower = null)
             $offset = $offset['value'];
         }
 
-        return $this->load($index - $offset + 1, $offset - 1, $lower);
+        if ($offset > 0) {
+            return $this->load($index - $offset + 1, $offset - 1, $lower);
+        }
+
+        return false;
     }
 
     /**
