diff --git a/vendor/magento/module-quote/Model/Quote/Item/Compare.php b/vendor/magento/module-quote/Model/Quote/Item/Compare.php
index f7fa741f0f1c3..ce957650feb33 100644
--- a/vendor/magento/module-quote/Model/Quote/Item/Compare.php
+++ b/vendor/magento/module-quote/Model/Quote/Item/Compare.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\Quote\Model\Quote\Item;

@@ -68,14 +68,9 @@ protected function getOptionValues($value)
      */
     public function compare(Item $target, Item $compared)
     {
-        if ($target->getSku() !== null && $target->getSku() === $compared->getSku()) {
-            return true;
-        }
-
         if ($target->getProductId() != $compared->getProductId()) {
             return false;
         }
-
         $targetOptionByCode = $target->getOptionsByCode();
         $comparedOptionsByCode = $compared->getOptionsByCode();
         if (!$target->compareOptions($targetOptionByCode, $comparedOptionsByCode)) {

