diff --git a/vendor/magento/module-quote/Model/Quote/Address.php b/vendor/magento/module-quote/Model/Quote/Address.php
index 6a06deb9cbb..52a98495649 100644
--- a/vendor/magento/module-quote/Model/Quote/Address.php
+++ b/vendor/magento/module-quote/Model/Quote/Address.php
@@ -1227,9 +1227,10 @@ class Address extends AbstractAddress implements
             ? $this->getBaseTaxAmount() + $this->getBaseDiscountTaxCompensationAmount()
             : 0;
 
+        // Note: ($x > $y - 0.0001) means ($x >= $y) for floats
         return $includeDiscount ?
-            ($this->getBaseSubtotalWithDiscount() + $taxes >= $amount) :
-            ($this->getBaseSubtotal() + $taxes >= $amount);
+            ($this->getBaseSubtotalWithDiscount() + $taxes > $amount - 0.0001) :
+            ($this->getBaseSubtotal() + $taxes > $amount - 0.0001);
     }
 
     /**
@@ -1385,7 +1386,7 @@ class Address extends AbstractAddress implements
      */
     public function getBaseSubtotalWithDiscount()
     {
-        return $this->getBaseSubtotal() + $this->getBaseDiscountAmount();
+        return $this->getBaseSubtotal() + $this->getBaseDiscountAmount() + $this->getBaseShippingDiscountAmount();
     }
 
     /**
