diff --git a/vendor/magento/module-catalog-import-export/Model/Import/Product.php b/vendor/magento/module-catalog-import-export/Model/Import/Product.php
index 25dc5360c8d..2f2f98a5e24 100644
--- a/vendor/magento/module-catalog-import-export/Model/Import/Product.php
+++ b/vendor/magento/module-catalog-import-export/Model/Import/Product.php
@@ -1908,10 +1908,7 @@ class Product extends AbstractEntity
                     }
                     if (self::SCOPE_STORE == $rowScope) {
                         if (self::SCOPE_WEBSITE == $attribute->getIsGlobal()) {
-                            // check website defaults already set
-                            if (!isset($attributes[$attrTable][$rowSku][$attrId][$rowStore])) {
-                                $storeIds = $this->storeResolver->getStoreIdToWebsiteStoreIds($rowStore);
-                            }
+                            $storeIds = $this->storeResolver->getStoreIdToWebsiteStoreIds($rowStore);
                         } elseif (self::SCOPE_STORE == $attribute->getIsGlobal()) {
                             $storeIds = [$rowStore];
                         }
