diff --git a/vendor/magento/module-inventory-in-store-pickup-sales-admin-ui/view/adminhtml/web/order/create/scripts-mixin.js b/vendor/magento/module-inventory-in-store-pickup-sales-admin-ui/view/adminhtml/web/order/create/scripts-mixin.js
index 600198b0404e..5eb2a0add041 100644
--- a/vendor/magento/module-inventory-in-store-pickup-sales-admin-ui/view/adminhtml/web/order/create/scripts-mixin.js
+++ b/vendor/magento/module-inventory-in-store-pickup-sales-admin-ui/view/adminhtml/web/order/create/scripts-mixin.js
@@ -74,6 +74,7 @@ define(
             window.AdminOrder.prototype.setShippingMethod = function (method) {
                 var data = {},
                     areas = [
+                        'items',
                         'shipping_method',
                         'totals',
                         'billing_method',
