diff --git a/vendor/magento/module-downloadable/Ui/DataProvider/Product/Form/Modifier/Data/Links.php b/vendor/magento/module-downloadable/Ui/DataProvider/Product/Form/Modifier/Data/Links.php
index 3be1094f7a4b..7c3c30482fd8 100644
--- a/vendor/magento/module-downloadable/Ui/DataProvider/Product/Form/Modifier/Data/Links.php
+++ b/vendor/magento/module-downloadable/Ui/DataProvider/Product/Form/Modifier/Data/Links.php
@@ -120,7 +120,7 @@ public function getLinksData()
             $linkData = [];
             $linkData['link_id'] = $link->getId();
             $linkData['title'] = $this->escaper->escapeHtml($link->getTitle());
-            $linkData['price'] = $this->getPriceValue($link->getPrice());
+            $linkData['price'] = $this->getPriceValue((float) $link->getPrice());
             $linkData['number_of_downloads'] = $link->getNumberOfDownloads();
             $linkData['is_shareable'] = $link->getIsShareable();
             $linkData['link_url'] = $link->getLinkUrl();
