diff --git a/vendor/magento/module-customer/Model/Metadata/Form/File.php b/vendor/magento/module-customer/Model/Metadata/Form/File.php
index 54b8b75c9ca..05788dcaf76 100644
--- a/vendor/magento/module-customer/Model/Metadata/Form/File.php
+++ b/vendor/magento/module-customer/Model/Metadata/Form/File.php
@@ -23,6 +23,8 @@ use Magento\Framework\Filesystem\Io\File as IoFile;
  */
 class File extends AbstractData
 {
+    public const UPLOADED_FILE_SUFFIX = '_uploaded';
+
     /**
      * Validator for check not protected extensions
      *
@@ -59,7 +61,8 @@ class File extends AbstractData
 
     /**
      * @var FileProcessorFactory
-     * @deprecated 101.0.0
+     * @deprecated 101.0.0 Call fileProcessor directly from code
+     * @see $this->fileProcessor
      */
     protected $fileProcessorFactory;
 
@@ -126,7 +129,7 @@ class File extends AbstractData
         $attrCode = $this->getAttribute()->getAttributeCode();
 
         // phpcs:disable Magento2.Security.Superglobal
-        $uploadedFile = $request->getParam($attrCode . '_uploaded');
+        $uploadedFile = $request->getParam($attrCode . static::UPLOADED_FILE_SUFFIX);
         if ($uploadedFile) {
             $value = $uploadedFile;
         } elseif ($this->_requestScope || !isset($_FILES[$attrCode])) {
@@ -424,7 +427,8 @@ class File extends AbstractData
      * Get file processor
      *
      * @return FileProcessor
-     * @deprecated 100.1.3
+     * @deprecated 100.1.3 we don’t use such approach anymore. Call fileProcessor directly
+     * @see $this->fileProcessor
      */
     protected function getFileProcessor()
     {
