diff --git a/vendor/magento/module-configurable-product/Model/Product/Type/Configurable.php b/vendor/magento/module-configurable-product/Model/Product/Type/Configurable.php
index 7f228caeb3e46..0558d68bd613f 100644
--- a/vendor/magento/module-configurable-product/Model/Product/Type/Configurable.php
+++ b/vendor/magento/module-configurable-product/Model/Product/Type/Configurable.php
@@ -1,9 +1,8 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
-
 namespace Magento\ConfigurableProduct\Model\Product\Type;
 
 use Magento\Catalog\Api\Data\ProductAttributeInterface;
@@ -1482,7 +1481,8 @@ private function getAttributesForCollection(\Magento\Catalog\Model\Product $prod
             'thumbnail',
             'status',
             'visibility',
-            'media_gallery'
+            'media_gallery',
+            'special_price',
         ];
 
         $usedAttributes = array_map(

