diff --git a/vendor/magento/module-ui/view/base/web/js/lib/view/utils/dom-observer.js b/vendor/magento/module-ui/view/base/web/js/lib/view/utils/dom-observer.js
index bcb8e97f98669..ea13611dfe92a 100644
--- a/vendor/magento/module-ui/view/base/web/js/lib/view/utils/dom-observer.js
+++ b/vendor/magento/module-ui/view/base/web/js/lib/view/utils/dom-observer.js
@@ -149,7 +149,7 @@ define([
         _.each(watchers.selectors, function (listeners, selector) {
             for (let data of listeners) {
                 if (!data.ctx.contains(node) || !$(node, data.ctx).is(selector)) {
-                    break;
+                    continue;
                 }
 
                 if (data.type === 'add') {
