diff --git a/vendor/magento/module-reports/Model/ResourceModel/Quote/Item/Collection.php b/vendor/magento/module-reports/Model/ResourceModel/Quote/Item/Collection.php
index e7dc28eb74a4..43ebd2137fc9 100644
--- a/vendor/magento/module-reports/Model/ResourceModel/Quote/Item/Collection.php
+++ b/vendor/magento/module-reports/Model/ResourceModel/Quote/Item/Collection.php
@@ -121,7 +121,7 @@ public function prepareActiveCartItems()

         $quoteItemsSelect->reset()
             ->from(['main_table' => $this->getTable('quote_item')], '')
-            ->columns('main_table.product_id')
+            ->columns(['main_table.product_id', 'main_table.name', 'main_table.price'])
             ->columns(['carts' => new \Zend_Db_Expr('COUNT(main_table.item_id)')])
             ->columns('quote.base_to_global_rate')
             ->joinInner(
