diff --git a/vendor/magento/module-inventory-in-store-pickup-sales/Model/Order/IsFulfillable.php b/vendor/magento/module-inventory-in-store-pickup-sales/Model/Order/IsFulfillable.php
index 0c51389acb9f..c128dccea722 100644
--- a/vendor/magento/module-inventory-in-store-pickup-sales/Model/Order/IsFulfillable.php
+++ b/vendor/magento/module-inventory-in-store-pickup-sales/Model/Order/IsFulfillable.php
@@ -74,7 +74,12 @@ public function execute(OrderInterface $order): bool
                 if ($item->getHasChildren()) {
                     continue;
                 }
-                if (!$this->isItemFulfillable($item->getSku(), $sourceCode, (float)$item->getQtyOrdered())) {
+
+                $stockItem =  $item->getProduct()->getExtensionAttributes()->getStockItem();
+                if (!$stockItem->getManageStock()) {
+                    return $stockItem->getIsInStock() === (bool)SourceItemInterface::STATUS_IN_STOCK &&
+                        $this->sourceRepository->get($sourceCode)->isEnabled();
+                } elseif (!$this->isItemFulfillable($item->getSku(), $sourceCode, (float)$item->getQtyOrdered())) {
                     return false;
                 }
             }
