diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php
index 3a9d430a8018..0c76010e4bfe 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php
@@ -98,12 +98,15 @@ private function getAppliedTaxes(Total $total, string $currency): array
         $rates = [];
 
         foreach ($appliedTaxes as $appliedTax) {
+            $totalPercentage =  $appliedTax['percent'];
             foreach ($appliedTax['rates'] as $appliedTaxRate) {
                 $rateTitle = $appliedTaxRate['title'];
                 if (!array_key_exists($rateTitle, $rates)) {
                     $rates[$rateTitle] = 0.0;
                 }
-                $rates[$rateTitle] += $appliedTax['amount'];
+                $percentage = $appliedTaxRate['percent'];
+                $taxValue = ($percentage / $totalPercentage) * $appliedTax['amount'];
+                $rates[$rateTitle] += round((float) $taxValue, 2);
             }
         }
 
