diff --git a/vendor/magento/module-re-captcha-frontend-ui/view/frontend/web/js/reCaptcha.js b/vendor/magento/module-re-captcha-frontend-ui/view/frontend/web/js/reCaptcha.js
index 14f2af32..20b1caf5 100644
--- a/vendor/magento/module-re-captcha-frontend-ui/view/frontend/web/js/reCaptcha.js
+++ b/vendor/magento/module-re-captcha-frontend-ui/view/frontend/web/js/reCaptcha.js
@@ -160,6 +160,13 @@ define(
                             grecaptcha.execute(widgetId);
                             event.preventDefault(event);
                             event.stopImmediatePropagation();
+                            if (this.$parentForm.valid()) {
+                                let formSubmitButton = this.$parentForm.find('button:not([type]), [type=submit]');
+
+                                if (formSubmitButton.length) { //eslint-disable-line max-depth
+                                    formSubmitButton.attr('disabled', true);
+                                }
+                            }
                         }
                     }.bind(this));
 
