diff --git a/lib/web/mage/validation/validation.js b/lib/web/mage/validation/validation.js
index 578d7968c4732..12c3d09d86e11 100644
--- a/lib/web/mage/validation/validation.js
+++ b/lib/web/mage/validation/validation.js
@@ -13,24 +13,19 @@ define([
     $.each({
         'validate-grouped-qty': [
             function (value, element, params) {
-                var result = false,
-                    total = 0;
+                var result = true,
+                    total = 0,
+                    inputVal = parseFloat($(element).val()) || 0;
 
-                $(params).find('input[data-validate*="validate-grouped-qty"]').each(function (i, e) {
-                    var val = $(e).val(),
-                        valInt;
-
-                    if (val && val.length > 0) {
-                        result = true;
-                        valInt = parseFloat(val) || 0;
+                if (inputVal < 0) {
+                    result = false;
+                }
 
-                        if (valInt >= 0) {
-                            total += valInt;
-                        } else {
-                            result = false;
+                $(params).find('input[data-validate*="validate-grouped-qty"]').each(function (i, e) {
+                    var groupedVal = parseFloat($(e).val()) || 0;
 
-                            return result;
-                        }
+                    if (groupedVal >= 0) {
+                        total += groupedVal;
                     }
                 });
 
