diff --git a/vendor/magento/module-catalog/view/frontend/templates/product/list/toolbar/limiter.phtml b/vendor/magento/module-catalog/view/frontend/templates/product/list/toolbar/limiter.phtml
index fd20210af389..04b0fe696628 100644
--- a/vendor/magento/module-catalog/view/frontend/templates/product/list/toolbar/limiter.phtml
+++ b/vendor/magento/module-catalog/view/frontend/templates/product/list/toolbar/limiter.phtml
@@ -23,7 +23,9 @@
                     <?php if ($block->isLimitCurrent($_key)):?>
                         selected="selected"
                     <?php endif ?>>
-                    <?= $block->escapeHtml($localeFormatter->formatNumber((int) $_limit)) ?>
+                    <?= $block->escapeHtml(
+                        is_numeric($_limit) ? $localeFormatter->formatNumber((int) $_limit) : $_limit
+                    ) ?>
                 </option>
             <?php endforeach; ?>
         </select>
