diff --git a/vendor/magento/module-backend/Block/Widget/Grid/Extended.php b/vendor/magento/module-backend/Block/Widget/Grid/Extended.php
index a7bdd13bb00b..3048077fbec4 100644
--- a/vendor/magento/module-backend/Block/Widget/Grid/Extended.php
+++ b/vendor/magento/module-backend/Block/Widget/Grid/Extended.php
@@ -1019,6 +1019,7 @@ public function getCsvFile()
         $stream = $this->_directory->openFile($file, 'w+');

         $stream->lock();
+        $stream->write(pack('CCC', 0xef, 0xbb, 0xbf));
         $stream->writeCsv($this->_getExportHeaders());
         $this->_exportIterateCollection('_exportCsvItem', [$stream]);

