diff --git a/vendor/magento/framework/Api/DataObjectHelper.php b/vendor/magento/framework/Api/DataObjectHelper.php
index 6d3fbcaca9d..27c1e46e381 100644
--- a/vendor/magento/framework/Api/DataObjectHelper.php
+++ b/vendor/magento/framework/Api/DataObjectHelper.php
@@ -305,8 +305,8 @@ class DataObjectHelper
                         // (2) remove set_ in start of name
                         // (3) add name without is_ prefix
                         preg_replace(
-                            ['/(^|,)(?!set)[^,]*/S','/(.)([A-Z])/S', '/(^|,)set_/iS', '/(^|,)is_([^,]+)/is'],
-                            ['', '$1_$2', '$1', '$1$2,is_$2'],
+                            ['/(^|,)(?!set)[^,]*/S','/([A-Z])/S', '/(^|,)set_/iS', '/(^|,)is_([^,]+)/is'],
+                            ['', '_$1', '$1', '$1$2,is_$2'],
                             implode(',', $dataObjectMethods)
                         )
                     )
diff --git a/vendor/magento/framework/Api/SimpleDataObjectConverter.php b/vendor/magento/framework/Api/SimpleDataObjectConverter.php
index 5c95f83cb4a..e2384a1c560 100644
--- a/vendor/magento/framework/Api/SimpleDataObjectConverter.php
+++ b/vendor/magento/framework/Api/SimpleDataObjectConverter.php
@@ -172,6 +172,6 @@ class SimpleDataObjectConverter
      */
     public static function camelCaseToSnakeCase($name)
     {
-        return $name !== null ? strtolower(preg_replace('/(.)([A-Z])/', "$1_$2", $name)) : '';
+        return $name !== null ? strtolower(ltrim(preg_replace('/([A-Z])/m', "_$1", $name), '_')) : '';
     }
 }
