diff --git a/vendor/magento/module-catalog-graph-ql/Model/Config/FilterAttributeReader.php b/vendor/magento/module-catalog-graph-ql/Model/Config/FilterAttributeReader.php
index 6976086e748..fc46e5eeb21 100644
--- a/vendor/magento/module-catalog-graph-ql/Model/Config/FilterAttributeReader.php
+++ b/vendor/magento/module-catalog-graph-ql/Model/Config/FilterAttributeReader.php
@@ -108,6 +108,7 @@ class FilterAttributeReader implements ReaderInterface
         $filterTypeMap = [
             'price' => self::FILTER_RANGE_TYPE,
             'date' => self::FILTER_RANGE_TYPE,
+            'datetime' => self::FILTER_RANGE_TYPE,
             'select' => self::FILTER_EQUAL_TYPE,
             'multiselect' => self::FILTER_EQUAL_TYPE,
             'boolean' => self::FILTER_EQUAL_TYPE,
