diff --git a/vendor/magento/module-quote-graph-ql/Model/CartItem/GetPaginatedCartItems.php b/vendor/magento/module-quote-graph-ql/Model/CartItem/GetPaginatedCartItems.php
index 2ba7c19347f..86718fc9864 100644
--- a/vendor/magento/module-quote-graph-ql/Model/CartItem/GetPaginatedCartItems.php
+++ b/vendor/magento/module-quote-graph-ql/Model/CartItem/GetPaginatedCartItems.php
@@ -28,12 +28,12 @@ class GetPaginatedCartItems
      *
      * @param Quote $cart
      * @param int $pageSize
-     * @param int $offset
+     * @param int $currentPage
      * @param string $orderBy
      * @param string $order
      * @return array
      */
-    public function execute(Quote $cart, int $pageSize, int $offset, string $orderBy, string $order): array
+    public function execute(Quote $cart, int $pageSize, int $currentPage, string $orderBy, string $order): array
     {
         if (!$cart->getId()) {
             return [
@@ -46,8 +46,9 @@ class GetPaginatedCartItems
             ->addFieldToFilter('parent_item_id', ['null' => true])
             ->addFieldToFilter('quote_id', $cart->getId())
             ->setOrder($orderBy, $order)
-            ->setCurPage($offset)
-            ->setPageSize($pageSize);
+            ->setCurPage($currentPage)
+            ->setPageSize($pageSize)
+            ->setQuote($cart);
 
         $items = [];
         $itemDeletedCount = 0;
diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemsPaginated.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemsPaginated.php
index a32f94a9a80..f81d55ac9ca 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemsPaginated.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItemsPaginated.php
@@ -49,7 +49,6 @@ class CartItemsPaginated implements ResolverInterface
 
         $pageSize = $args['pageSize'];
         $currentPage = $args['currentPage'];
-        $offset = ($currentPage - 1) * $pageSize;
         $order = CartItemsPaginated::SORT_ORDER;
         $orderBy = CartItemsPaginated::SORT_ORDER_BY;
 
@@ -59,7 +58,7 @@ class CartItemsPaginated implements ResolverInterface
         }
 
         $allVisibleItems = $cart->getAllVisibleItems();
-        $paginatedCartItems = $this->pagination->execute($cart, $pageSize, (int) $offset, $orderBy, $order);
+        $paginatedCartItems = $this->pagination->execute($cart, $pageSize, (int) $currentPage, $orderBy, $order);
 
         $cartItems = [];
         /** @var CartItemInterface $cartItem */
