diff --git a/vendor/magento/module-paypal/Model/Payflow/Transparent.php b/vendor/magento/module-paypal/Model/Payflow/Transparent.php
index f9233c15a4a7..ff715c271509 100644
--- a/vendor/magento/module-paypal/Model/Payflow/Transparent.php
+++ b/vendor/magento/module-paypal/Model/Payflow/Transparent.php
@@ -389,20 +389,6 @@ public function denyPayment(InfoInterface $payment)
         return true;
     }
 
-    /**
-     * @inheritDoc
-     */
-    public function fetchTransactionInfo(InfoInterface $payment, $transactionId)
-    {
-        $result = parent::fetchTransactionInfo($payment, $transactionId);
-        $this->_canFetchTransactionInfo = false;
-        if ($payment->getIsTransactionApproved()) {
-            $this->acceptPayment($payment);
-        }
-
-        return $result;
-    }
-
     /**
      * Marks payment as fraudulent.
      *
