diff --git a/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldCsv.php b/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldCsv.php
index 33ab4427254b..a9e109a55d7f 100644
--- a/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldCsv.php
+++ b/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldCsv.php
@@ -1,20 +1,19 @@
 <?php
 /**
- *
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);

 namespace Magento\Reports\Controller\Adminhtml\Report\Product;

 use Magento\Backend\Block\Widget\Grid\ExportInterface;
-use Magento\Framework\App\Action\HttpPostActionInterface;
+use Magento\Framework\App\Action\HttpGetActionInterface;
 use Magento\Framework\App\ResponseInterface;
 use Magento\Framework\App\Filesystem\DirectoryList;
 use Magento\Reports\Controller\Adminhtml\Report\Product;

-class ExportSoldCsv extends Product implements HttpPostActionInterface
+class ExportSoldCsv extends Product implements HttpGetActionInterface
 {
     /**
      * Authorization level of a basic admin session
diff --git a/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldExcel.php b/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldExcel.php
index 1851e2ce2a26..41737d045590 100644
--- a/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldExcel.php
+++ b/vendor/magento/module-reports/Controller/Adminhtml/Report/Product/ExportSoldExcel.php
@@ -1,20 +1,19 @@
 <?php
 /**
- *
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);

 namespace Magento\Reports\Controller\Adminhtml\Report\Product;

 use Magento\Backend\Block\Widget\Grid\ExportInterface;
-use Magento\Framework\App\Action\HttpPostActionInterface;
+use Magento\Framework\App\Action\HttpGetActionInterface;
 use Magento\Framework\App\ResponseInterface;
 use Magento\Framework\App\Filesystem\DirectoryList;
 use Magento\Reports\Controller\Adminhtml\Report\Product;

-class ExportSoldExcel extends Product implements HttpPostActionInterface
+class ExportSoldExcel extends Product implements HttpGetActionInterface
 {
     /**
      * Authorization level of a basic admin session

