diff --git a/vendor/magento/module-configurable-product/view/frontend/web/js/options-updater.js b/vendor/magento/module-configurable-product/view/frontend/web/js/options-updater.js
index 558a1fdf310..74e1514fad6 100644
--- a/vendor/magento/module-configurable-product/view/frontend/web/js/options-updater.js
+++ b/vendor/magento/module-configurable-product/view/frontend/web/js/options-updater.js
@@ -1,7 +1,8 @@
 define([
     'jquery',
     'underscore',
-    'Magento_Customer/js/customer-data'
+    'Magento_Customer/js/customer-data',
+    'domReady!'
 ], function ($, _, customerData) {
     'use strict';

diff --git a/vendor/magento/module-swatches/view/frontend/web/js/configurable-customer-data.js b/vendor/magento/module-swatches/view/frontend/web/js/configurable-customer-data.js
index 6467c6276da..fac7540ad28 100644
--- a/vendor/magento/module-swatches/view/frontend/web/js/configurable-customer-data.js
+++ b/vendor/magento/module-swatches/view/frontend/web/js/configurable-customer-data.js
@@ -8,7 +8,7 @@ require([
             formSelector: '#product_addtocart_form',
             swatchSelector: '.swatch-opt'
         },
-        swatchWidgetName = 'mageSwatchRenderer',
+        swatchWidgetName = 'mage-SwatchRenderer',
         widgetInitEvent = 'swatch.initialized',

     /**

