diff --git a/vendor/magento/module-review/Controller/Adminhtml/Product/MassUpdateStatus.php b/vendor/magento/module-review/Controller/Adminhtml/Product/MassUpdateStatus.php
index ff4acfb9648..835c3c0cc3f 100644
--- a/vendor/magento/module-review/Controller/Adminhtml/Product/MassUpdateStatus.php
+++ b/vendor/magento/module-review/Controller/Adminhtml/Product/MassUpdateStatus.php
@@ -127,6 +127,7 @@ class MassUpdateStatus extends ProductController implements HttpPostActionInterf
                     ->getIdFieldName(),
                 $this->getRequest()->getParam('reviews')
             );
+            $collection->addStoreData();

             $this->collection = $collection;
         }
