diff --git a/vendor/magento/module-catalog/Helper/Product/View.php b/vendor/magento/module-catalog/Helper/Product/View.php
index ca9b550ca31d..abf151c713e9 100644
--- a/vendor/magento/module-catalog/Helper/Product/View.php
+++ b/vendor/magento/module-catalog/Helper/Product/View.php
@@ -129,14 +129,7 @@ private function preparePageMetadata(ResultPage $resultPage, $product)
             $pageConfig->setKeywords($product->getName());
         }
 
-        $description = $product->getMetaDescription();
-        if ($description) {
-            $pageConfig->setDescription($description);
-        } else {
-            $productDescription = is_string($product->getDescription()) ?
-                $this->string->substr(strip_tags($product->getDescription()), 0, 255) : '';
-            $pageConfig->setDescription($productDescription);
-        }
+        $pageConfig->setDescription($product->getMetaDescription());
 
         if ($this->_catalogProduct->canUseCanonicalTag()) {
             $pageConfig->addRemotePageAsset(
diff --git a/setup/src/Magento/Setup/Fixtures/SimpleProductsFixture.php b/setup/src/Magento/Setup/Fixtures/SimpleProductsFixture.php
index 84fb2e5beed4..8212a97c12b0 100644
--- a/setup/src/Magento/Setup/Fixtures/SimpleProductsFixture.php
+++ b/setup/src/Magento/Setup/Fixtures/SimpleProductsFixture.php
@@ -39,7 +39,7 @@ class SimpleProductsFixture extends Fixture
     /**
      * Simple product sku pattern
      */
-    const SKU_PATTERN = 'product_dynamic_%s';
+    public const SKU_PATTERN = 'product_dynamic_%s';
 
     /**
      * @var int
@@ -257,6 +257,12 @@ public function execute()
             'category_ids' => function ($index, $entityNumber) {
                 return $this->websiteCategoryProvider->getCategoryId($index);
             },
+            'meta_keyword' => function ($productId) {
+                return sprintf($this->getSkuPattern(), $productId);
+            },
+            'meta_title' => function ($productId) {
+                return sprintf($this->getSkuPattern(), $productId);
+            },
             'attribute_set_id' => $attributeSet,
             'additional_attributes' => $additionalAttributes,
             'status' => function () {
diff --git a/setup/src/Magento/Setup/Model/FixtureGenerator/SimpleProductTemplateGenerator.php b/setup/src/Magento/Setup/Model/FixtureGenerator/SimpleProductTemplateGenerator.php
index 662bd4c5ca3a..4af988a6c427 100644
--- a/setup/src/Magento/Setup/Model/FixtureGenerator/SimpleProductTemplateGenerator.php
+++ b/setup/src/Magento/Setup/Model/FixtureGenerator/SimpleProductTemplateGenerator.php
@@ -39,7 +39,7 @@ public function __construct(ProductFactory $productFactory, array $fixture)
     }
 
     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     public function generateEntity()
     {
@@ -70,6 +70,9 @@ private function getProductTemplate($attributeSet, $additionalAttributes = [])
                 'name' => 'template name' . $productRandomizerNumber,
                 'url_key' => 'template-url' . $productRandomizerNumber,
                 'sku' => 'template_sku_simple' . $productRandomizerNumber,
+                'meta_description' => 'Simple Product',
+                'meta_keyword' => $productRandomizerNumber,
+                'meta_title' => $productRandomizerNumber,
                 'price' => 10,
                 'visibility' => Visibility::VISIBILITY_BOTH,
                 'status' => Status::STATUS_ENABLED,
