diff --git a/vendor/magento/framework/App/DeploymentConfig.php b/vendor/magento/framework/App/DeploymentConfig.php
index 9045c3539e336..9c977754fd251 100644
--- a/vendor/magento/framework/App/DeploymentConfig.php
+++ b/vendor/magento/framework/App/DeploymentConfig.php
@@ -1,9 +1,8 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2014 Adobe
+ * All Rights Reserved.
  */
-
 namespace Magento\Framework\App;
 
 use Magento\Framework\Config\ConfigOptionsListConstants;
@@ -239,7 +238,11 @@ private function getAllEnvOverrides(): array
                 ) {
                     // convert MAGENTO_DC_DB__CONNECTION__DEFAULT__HOST into db/connection/default/host
                     $flatKey = strtolower(str_replace([self::MAGENTO_ENV_PREFIX, '__'], ['', '/'], $key));
-                    $this->envOverrides[$flatKey] = $value;
+                    $this->envOverrides[$flatKey] = match ($value) {
+                        'true', 'TRUE' => true,
+                        'false', 'FALSE' => false,
+                        default => $value,
+                    };
                 }
             }
         }
