diff --git a/vendor/magento/module-cms/Model/Wysiwyg/DefaultConfigProvider.php b/vendor/magento/module-cms/Model/Wysiwyg/DefaultConfigProvider.php
index 09820927e6036..aca6a109f4b42 100644
--- a/vendor/magento/module-cms/Model/Wysiwyg/DefaultConfigProvider.php
+++ b/vendor/magento/module-cms/Model/Wysiwyg/DefaultConfigProvider.php
@@ -1,27 +1,30 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2017 Adobe
+ * All Rights Reserved.
  */

 declare(strict_types=1);

 namespace Magento\Cms\Model\Wysiwyg;

+use Magento\Framework\DataObject;
+use Magento\Framework\View\Asset\Repository;
+
 /**
  * Class DefaultConfigProvider returns data required to render tinymce editor
  */
 class DefaultConfigProvider implements \Magento\Framework\Data\Wysiwyg\ConfigProviderInterface
 {
     /**
-     * @var \Magento\Framework\View\Asset\Repository
+     * @var Repository
      */
-    private $assetRepo;
+    private Repository $assetRepo;

     /**
-     * @param \Magento\Framework\View\Asset\Repository $assetRepo
+     * @param Repository $assetRepo
      */
-    public function __construct(\Magento\Framework\View\Asset\Repository $assetRepo)
+    public function __construct(Repository $assetRepo)
     {
         $this->assetRepo = $assetRepo;
     }
@@ -29,12 +32,12 @@ public function __construct(\Magento\Framework\View\Asset\Repository $assetRepo)
     /**
      * @inheritdoc
      */
-    public function getConfig(\Magento\Framework\DataObject $config) : \Magento\Framework\DataObject
+    public function getConfig(DataObject $config) : DataObject
     {
         $config->addData([
             'tinymce' => [
-                'toolbar' => ' blocks | formatselect | bold italic underline | alignleft aligncenter alignright | '
-                    . 'bullist numlist | link table charmap',
+                'toolbar' => ' blocks fontfamily fontsize| formatselect | bold italic underline ' .
+                    '| alignleft aligncenter alignright | bullist numlist | link table charmap',
                 'plugins' => implode(
                     ' ',
                     [
