diff --git a/vendor/magento/module-page-builder/Model/Wysiwyg/DefaultConfigProvider.php b/vendor/magento/module-page-builder/Model/Wysiwyg/DefaultConfigProvider.php
index a179e74e3..e499f6471 100644
--- a/vendor/magento/module-page-builder/Model/Wysiwyg/DefaultConfigProvider.php
+++ b/vendor/magento/module-page-builder/Model/Wysiwyg/DefaultConfigProvider.php
@@ -42,7 +42,7 @@ public function getConfig(\Magento\Framework\DataObject $config): \Magento\Frame
         $config->addData(
             [
                 'tinymce' => [
-                    'toolbar' => 'undo redo | styles | fontsizeselect | lineheight | forecolor backcolor ' .
+                    'toolbar' => 'undo redo | styles | fontfamily fontsizeinput | lineheight | forecolor backcolor ' .
                         '| bold italic underline | alignleft aligncenter alignright | numlist bullist ' .
                         '| link image table charmap',
 
