diff --git a/vendor/magento/module-sales/Model/Order/Webapi/ChangeOutputArray.php b/vendor/magento/module-sales/Model/Order/Webapi/ChangeOutputArray.php
index 9fd47a43d4dc4..eeff67c68345c 100644
--- a/vendor/magento/module-sales/Model/Order/Webapi/ChangeOutputArray.php
+++ b/vendor/magento/module-sales/Model/Order/Webapi/ChangeOutputArray.php
@@ -73,6 +73,6 @@ public function execute(
      */
     private function round(float $value): float
     {
-        return (float) max($value, 0);
+        return (float) max(round($value, 2), 0);
     }
 }
