diff --git a/vendor/magento/module-quote/Model/QuoteManagement.php b/vendor/magento/module-quote/Model/QuoteManagement.php
index 51b68411d408..0a25ee498eed 100644
--- a/vendor/magento/module-quote/Model/QuoteManagement.php
+++ b/vendor/magento/module-quote/Model/QuoteManagement.php
@@ -412,7 +412,9 @@ public function placeOrder($cartId, PaymentInterface $paymentMethod = null)
         if ($quote->getCheckoutMethod() === self::METHOD_GUEST || !$customerId) {
             $quote->setCustomerId(null);
             $billingAddress = $quote->getBillingAddress();
-            $quote->setCustomerEmail($billingAddress ? $billingAddress->getEmail() : null);
+            if (!$quote->getCustomerEmail()) {
+                $quote->setCustomerEmail($billingAddress ? $billingAddress->getEmail() : null);
+            }
             if ($quote->getCustomerFirstname() === null
                 && $quote->getCustomerLastname() === null
                 && $billingAddress
