diff --git a/vendor/magento/module-webapi/Controller/Rest/InputParamsResolver.php b/vendor/magento/module-webapi/Controller/Rest/InputParamsResolver.php
index 185255083214c..1283b21ba345a 100644
--- a/vendor/magento/module-webapi/Controller/Rest/InputParamsResolver.php
+++ b/vendor/magento/module-webapi/Controller/Rest/InputParamsResolver.php
@@ -138,11 +138,12 @@ public function getInputData()
                 $serviceMethodName
             );
             $inputData = array_merge($inputData, $this->request->getParams());
+
+            $inputData = $this->filterInputData($inputData);
         } else {
             $inputData = $this->request->getRequestData();
         }
 
-        $inputData = $this->filterInputData($inputData);
         $this->validateParameters($serviceClassName, $serviceMethodName, array_keys($route->getParameters()));
 
         return $this->paramsOverrider->override($inputData, $route->getParameters());

