diff --git a/vendor/magento/module-paypal/Model/Express/Checkout.php b/vendor/magento/module-paypal/Model/Express/Checkout.php
index ca3ed2977f38e..70edc51d364e8 100644
--- a/vendor/magento/module-paypal/Model/Express/Checkout.php
+++ b/vendor/magento/module-paypal/Model/Express/Checkout.php
@@ -481,6 +481,7 @@ public function setCustomerWithAddressChange(
      */
     public function start($returnUrl, $cancelUrl, $button = null)
     {
+        $this->_quote->setPayment($this->_quote->getPayment());
         $this->_quote->collectTotals();
 
         if (!$this->_quote->getGrandTotal()) {

