diff --git a/vendor/magento/module-quote/Model/Cart/AddProductsToCart.php b/vendor/magento/module-quote/Model/Cart/AddProductsToCart.php
index 9be1e9d32e37..3af58e3cd1be 100644
--- a/vendor/magento/module-quote/Model/Cart/AddProductsToCart.php
+++ b/vendor/magento/module-quote/Model/Cart/AddProductsToCart.php
@@ -7,6 +7,7 @@
 
 namespace Magento\Quote\Model\Cart;
 
+use Magento\Catalog\Api\Data\ProductInterface;
 use Magento\Framework\Exception\NoSuchEntityException;
 use Magento\Quote\Api\CartRepositoryInterface;
 use Magento\Quote\Model\Cart\BuyRequest\BuyRequestBuilder;
@@ -18,6 +19,8 @@
 /**
  * Unified approach to add products to the Shopping Cart.
  * Client code must validate, that customer is eligible to call service with provided {cartId} and {cartItems}
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  */
 class AddProductsToCart
 {
@@ -165,7 +168,9 @@ private function addItemToCart(Quote $cart, Data\CartItem $cartItem, int $cartIt
                 $cartItemPosition
             );
         } else {
-            $product = $this->productReader->getProductBySku($sku);
+            $productBySku = $this->productReader->getProductBySku($sku);
+            $product = isset($productBySku) ? clone $productBySku : null;
+            $product = $this->getProductOptions($product);
             if (!$product || !$product->isSaleable() || !$product->isAvailable()) {
                 $errors[] = $this->error->create(
                     __('Could not find a product with SKU "%sku"', ['sku' => $sku])->render(),
@@ -206,4 +211,16 @@ private function prepareErrorOutput(Quote $cart, array $errors = []): AddProduct
 
         return $output;
     }
+
+    /**
+     * Set options from product options collection
+     *
+     * @param ProductInterface|null $productItem
+     * @return ProductInterface|null
+     */
+    private function getProductOptions(ProductInterface|null $productItem): ?ProductInterface
+    {
+        $productItem?->setOptions($productItem->getProductOptionsCollection()->getItems());
+        return $productItem;
+    }
 }
