diff --git a/vendor/magento/module-page-cache/Model/App/Response/HttpPlugin.php b/vendor/magento/module-page-cache/Model/App/Response/HttpPlugin.php
index 7de1eb4597505..bd3d0cdd5359f 100644
--- a/vendor/magento/module-page-cache/Model/App/Response/HttpPlugin.php
+++ b/vendor/magento/module-page-cache/Model/App/Response/HttpPlugin.php
@@ -43,8 +43,7 @@ public function beforeSendResponse(HttpResponse $subject)
 
         $currentVary = $this->context->getVaryString();
         $varyCookie = $this->request->get(HttpResponse::COOKIE_VARY_STRING);
-        if ($currentVary !== $varyCookie) {
-            //prevent caching with the old vary cookie
+        if (isset($varyCookie) && ($currentVary !== $varyCookie)) {
             $subject->setNoCacheHeaders();
         }
         $subject->sendVary();
