diff --git a/vendor/magento/module-catalog/Model/Product/Option/Type/Select.php b/vendor/magento/module-catalog/Model/Product/Option/Type/Select.php
index b2c7fc79d84bd..b49d7843a7baf 100644
--- a/vendor/magento/module-catalog/Model/Product/Option/Type/Select.php
+++ b/vendor/magento/module-catalog/Model/Product/Option/Type/Select.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
 
 namespace Magento\Catalog\Model\Product\Option\Type;
@@ -70,6 +70,7 @@ public function __construct(
      * @param array $values All product option values, i.e. array (option_id => mixed, option_id => mixed...)
      * @return $this
      * @throws LocalizedException
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
      */
     public function validateUserValue($values)
     {
@@ -92,6 +93,8 @@ public function validateUserValue($values)
             if (!$this->_isSingleSelection()) {
                 if (is_string($value)) {
                     $value = explode(',', $value);
+                } elseif (!is_array($value)) {
+                    $value = [$value];
                 }
                 $valuesCollection = $option->getOptionValuesByOptionId($value, $this->getProduct()->getStoreId());
                 $valueCount = is_array($value) ? count($value) : 0;
