diff --git a/vendor/magento/module-customer/Observer/AfterAddressSaveObserver.php b/vendor/magento/module-customer/Observer/AfterAddressSaveObserver.php
index 33290306e48..0a98f6521de 100644
--- a/vendor/magento/module-customer/Observer/AfterAddressSaveObserver.php
+++ b/vendor/magento/module-customer/Observer/AfterAddressSaveObserver.php
@@ -34,7 +34,7 @@ class AfterAddressSaveObserver implements ObserverInterface
     /**
      * VAT ID validation processed flag code
      */
-    const VIV_PROCESSED_FLAG = 'viv_after_address_save_processed';
+    public const VIV_PROCESSED_FLAG = 'viv_after_address_save_processed';
 
     /**
      * @var HelperAddress
@@ -141,8 +141,7 @@ class AfterAddressSaveObserver implements ObserverInterface
             if ($customerAddress->getVatId() == ''
                 || !$this->_customerVat->isCountryInEU($customerAddress->getCountry())
             ) {
-                $defaultGroupId = $customer->getGroupId() ? $customer->getGroupId() :
-                    $this->_groupManagement->getDefaultGroup($customer->getStore())->getId();
+                $defaultGroupId = $this->_groupManagement->getDefaultGroup($customer->getStore())->getId();
                 if (!$customer->getDisableAutoGroupChange() && $customer->getGroupId() != $defaultGroupId) {
                     $customer->setGroupId($defaultGroupId);
                     $customer->save();
