diff --git a/vendor/magento/module-sales/Block/Order/Recent.php b/vendor/magento/module-sales/Block/Order/Recent.php
index 6aef07ce8eb14..cb5828a2ff63b 100644
--- a/vendor/magento/module-sales/Block/Order/Recent.php
+++ b/vendor/magento/module-sales/Block/Order/Recent.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\Sales\Block\Order;

@@ -23,7 +23,7 @@ class Recent extends \Magento\Framework\View\Element\Template
     /**
      * Limit of orders
      */
-    const ORDER_LIMIT = 5;
+    public const ORDER_LIMIT = 5;

     /**
      * @var CollectionFactoryInterface
@@ -90,9 +90,6 @@ private function getRecentOrders()
         )->addAttributeToFilter(
             'customer_id',
             $customerId
-        )->addAttributeToFilter(
-            'store_id',
-            $this->storeManager->getStore()->getId()
         )->addAttributeToFilter(
             'status',
             ['in' => $this->_orderConfig->getVisibleOnFrontStatuses()]
@@ -122,6 +119,8 @@ public function getViewUrl($order)
      * @param object $order
      * @return string
      * @deprecated 102.0.3 Action does not exist
+     * @see This method is not used anymore
+     *
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
     public function getTrackUrl($order)
