diff --git a/vendor/magento/module-page-builder/view/adminhtml/web/js/content-type/text/preview.js b/vendor/magento/module-page-builder/view/adminhtml/web/js/content-type/text/preview.js
index 298b689548..aafdb29431 100644
--- a/vendor/magento/module-page-builder/view/adminhtml/web/js/content-type/text/preview.js
+++ b/vendor/magento/module-page-builder/view/adminhtml/web/js/content-type/text/preview.js
@@ -135,7 +135,6 @@ define(["jquery", "Magento_PageBuilder/js/events", "underscore", "Magento_PageBu
       }

       var wysiwygConfig = this.config.additional_data.wysiwygConfig.wysiwygConfigData;
-      wysiwygConfig.adapter.settings.paste_as_text = true;

       if (focus) {
         wysiwygConfig.adapter.settings.auto_focus = this.element.id;
diff --git a/vendor/magento/module-page-builder/view/adminhtml/web/ts/js/content-type/text/preview.ts b/vendor/magento/module-page-builder/view/adminhtml/web/ts/js/content-type/text/preview.ts
index bff3546342..ec30828221 100644
--- a/vendor/magento/module-page-builder/view/adminhtml/web/ts/js/content-type/text/preview.ts
+++ b/vendor/magento/module-page-builder/view/adminhtml/web/ts/js/content-type/text/preview.ts
@@ -155,8 +155,6 @@ export default class Preview extends BasePreview {

         const wysiwygConfig = this.config.additional_data.wysiwygConfig.wysiwygConfigData;

-        wysiwygConfig.adapter.settings.paste_as_text = true;
-
         if (focus) {
             wysiwygConfig.adapter.settings.auto_focus = this.element.id;
             wysiwygConfig.adapter.settings.init_instance_callback = () => {
