diff --git a/vendor/magento/module-bundle/Pricing/Adjustment/DefaultSelectionPriceListProvider.php b/vendor/magento/module-bundle/Pricing/Adjustment/DefaultSelectionPriceListProvider.php
index 297c4659cb87..b2a3f0c6f3c9 100644
--- a/vendor/magento/module-bundle/Pricing/Adjustment/DefaultSelectionPriceListProvider.php
+++ b/vendor/magento/module-bundle/Pricing/Adjustment/DefaultSelectionPriceListProvider.php
@@ -56,8 +56,8 @@ public function getPriceList(Product $bundleProduct, $searchMin, $useRegularPric
                 [(int)$option->getOptionId()],
                 $bundleProduct
             );
+            $selectionsCollection->setFlag('has_stock_status_filter', true);
             $selectionsCollection->removeAttributeToSelect();
-            $selectionsCollection->addQuantityFilter();

             if (!$useRegularPrice) {
                 $selectionsCollection->addAttributeToSelect('special_price');
@@ -112,6 +112,9 @@ private function isShouldFindMinOption(Product $bundleProduct, $searchMin)
     private function addMiniMaxPriceList(Product $bundleProduct, $selectionsCollection, $searchMin, $useRegularPrice)
     {
         $selectionsCollection->addPriceFilter($bundleProduct, $searchMin, $useRegularPrice);
+        if ($bundleProduct->isSalable()) {
+            $selectionsCollection->addQuantityFilter();
+        }
         $selectionsCollection->setPage(0, 1);

         $selection = $selectionsCollection->getFirstItem();
