diff --git a/vendor/magento/module-theme/Block/Html/Header.php b/vendor/magento/module-theme/Block/Html/Header.php
index 9f155268fe38..1550ebaa367d 100644
--- a/vendor/magento/module-theme/Block/Html/Header.php
+++ b/vendor/magento/module-theme/Block/Html/Header.php
@@ -59,7 +59,6 @@ public function getWelcome()
                 ScopeInterface::SCOPE_STORE
             );
         }
-        $this->_data['welcome'] =  $this->escaper->escapeQuote($this->_data['welcome'], true);
-        return __($this->_data['welcome']);
+        return $this->escaper->escapeQuote(__($this->_data['welcome'])->render(), true);
     }
 }
