diff --git a/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Formatter/LayerFormatter.php b/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Formatter/LayerFormatter.php
index 6df29fa..d3b4e31 100644
--- a/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Formatter/LayerFormatter.php
+++ b/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Formatter/LayerFormatter.php
@@ -24,7 +24,7 @@ class LayerFormatter
     public function buildLayer($layerName, $itemsCount, $requestName, $position = null): array
     {
         return [
-            'label' => $layerName,
+            'label' => __($layerName),
             'count' => $itemsCount,
             'attribute_code' => $requestName,
             'position' => isset($position) ? (int)$position : null
