diff --git a/vendor/magento/module-user/Block/Role/Grid/User.php b/vendor/magento/module-user/Block/Role/Grid/User.php
index 963eed151cc..65bd5cfc64a 100644
--- a/vendor/magento/module-user/Block/Role/Grid/User.php
+++ b/vendor/magento/module-user/Block/Role/Grid/User.php
@@ -16,7 +16,7 @@ use Magento\Backend\Block\Widget\Grid\Column;
 class User extends \Magento\Backend\Block\Widget\Grid\Extended
 {
     /**
-     * Core registry
+     * Framework class for Core Registry
      *
      * @var \Magento\Framework\Registry
      */
@@ -206,7 +206,16 @@ class User extends \Magento\Backend\Block\Widget\Grid\Extended
             if ($json) {
                 return $this->getJSONString($inRoleUser);
             }
-            return $this->escapeJs($this->escapeHtml($inRoleUser));
+            $escapedInRoleUser = $this->escapeHtml($inRoleUser);
+            if (is_array($escapedInRoleUser)) {
+                return array_map(
+                    function ($value) {
+                        return $this->escapeJs($value);
+                    },
+                    $escapedInRoleUser
+                );
+            }
+            return $this->escapeJs($escapedInRoleUser);
         }
         $roleId = $this->getRoleId();
         $users = $this->getUsersFormData();
