diff --git a/vendor/magento/module-eav/Model/Entity/Attribute/Backend/ArrayBackend.php b/vendor/magento/module-eav/Model/Entity/Attribute/Backend/ArrayBackend.php
index 7094503544ce..098c9c9040cf 100644
--- a/vendor/magento/module-eav/Model/Entity/Attribute/Backend/ArrayBackend.php
+++ b/vendor/magento/module-eav/Model/Entity/Attribute/Backend/ArrayBackend.php
@@ -64,6 +64,12 @@ public function validate($object)
      */
     private function prepare(array $data): string
     {
-        return implode(',', array_filter(array_unique($data), 'is_numeric'));
+        return implode(
+            ',',
+            array_filter(
+                array_unique($data),
+                fn($value) => is_numeric($value) || !empty($value)
+            )
+        );
     }
 }
