diff --git a/vendor/magento/module-product-video/view/frontend/web/js/fotorama-add-video-events.js b/vendor/magento/module-product-video/view/frontend/web/js/fotorama-add-video-events.js
index 670d91febe9f..a8a168c03aa9 100644
--- a/vendor/magento/module-product-video/view/frontend/web/js/fotorama-add-video-events.js
+++ b/vendor/magento/module-product-video/view/frontend/web/js/fotorama-add-video-events.js
@@ -558,7 +558,7 @@ define([
             }

             if (this.isFullscreen && this.fotoramaItem.data('fotorama').activeFrame.i === number) {
-                this.fotoramaItem.data('fotorama').activeFrame.$stageFrame[0].trigger('click');
+                this.fotoramaItem.data('fotorama').activeFrame.$stageFrame.trigger('click');
             }
         },

@@ -700,7 +700,7 @@ define([
             if (activeIndexIsBase && number === 1 && $(window).width() > this.MobileMaxWidth) {
                 setTimeout($.proxy(function () {
                     fotorama.requestFullScreen();
-                    this.fotoramaItem.data('fotorama').activeFrame.$stageFrame[0].trigger('click');
+                    this.fotoramaItem.data('fotorama').activeFrame.$stageFrame.trigger('click');
                     this.Base = false;
                 }, this), 50);
             }
diff --git a/lib/web/fotorama/fotorama.js b/lib/web/fotorama/fotorama.js
index a60999ef7647..17736feef448 100644
--- a/lib/web/fotorama/fotorama.js
+++ b/lib/web/fotorama/fotorama.js
@@ -1140,7 +1140,7 @@ fotoramaVersion = '4.6.4';

     function addEvent(el, e, fn, bool) {
         if (!e) return;
-        el.addEventListener ? el.addEventListener(e, fn, !!bool) : el.attachEvent('on' + e, fn);
+        el.addEventListener ? el.addEventListener(e, fn, {passive: !!bool}) : el.attachEvent('on' + e, fn);
     }

     /**
diff --git a/lib/web/magnifier/magnify.js b/lib/web/magnifier/magnify.js
index 7d193fc1cd97..f5cc2114a3f1 100644
--- a/lib/web/magnifier/magnify.js
+++ b/lib/web/magnifier/magnify.js
@@ -505,7 +505,7 @@ define([
             if (!$fotoramaStage.hasClass('magnify-wheel-loaded')) {
                 if (fotoramaStage && fotoramaStage.addEventListener) {
                     if ('onwheel' in document) {
-                        fotoramaStage.addEventListener('wheel', onWheel, { passive: true });
+                        fotoramaStage.addEventListener('wheel', onWheel, { passive: false });
                     } else if ('onmousewheel' in document) {
                         fotoramaStage.addEventListener('mousewheel', onWheel);
                     } else {
