diff --git a/vendor/magento/framework/Data/Form/Element/Date.php b/vendor/magento/framework/Data/Form/Element/Date.php
index 222f9588a1ccc..d85a245709975 100644
--- a/vendor/magento/framework/Data/Form/Element/Date.php
+++ b/vendor/magento/framework/Data/Form/Element/Date.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
 
 /**
@@ -123,7 +123,7 @@ public function getValue($format = null)
             return '';
         }
         if (null === $format) {
-            $format = $this->getDateFormat();
+            $format = $this->getDateFormat() ?: $this->getFormat();
             $format .= ($format && $this->getTimeFormat()) ? ' ' : '';
             $format .= $this->getTimeFormat() ? $this->getTimeFormat() : '';
         }
