diff --git a/vendor/magento/module-inventory-source-selection-api/Model/Algorithms/Result/GetDefaultSortedSourcesResult.php b/vendor/magento/module-inventory-source-selection-api/Model/Algorithms/Result/GetDefaultSortedSourcesResult.php
index c7f69e05512d..7d4640ad1195 100644
--- a/vendor/magento/module-inventory-source-selection-api/Model/Algorithms/Result/GetDefaultSortedSourcesResult.php
+++ b/vendor/magento/module-inventory-source-selection-api/Model/Algorithms/Result/GetDefaultSortedSourcesResult.php
@@ -95,7 +95,7 @@ public function execute(
 
         $itemsTdDeliver = [];
         foreach ($inventoryRequest->getItems() as $item) {
-            $normalizedSku = $this->normalizeSku($item->getSku());
+            $normalizedSku = $this->normalizeSku(trim($item->getSku()));
             $itemsTdDeliver[$normalizedSku] = $item->getQty();
         }
 
@@ -111,14 +111,14 @@ public function execute(
             );
 
         foreach ($sourceItems as $sourceItem) {
-            $normalizedSku = $this->normalizeSku($sourceItem->getSku());
+            $normalizedSku = $this->normalizeSku(trim($sourceItem->getSku()));
             $sourceItemQtyAvailable = $this->getSourceItemQtyAvailable->execute($sourceItem);
             $qtyToDeduct = min(max($sourceItemQtyAvailable, 0.0), $itemsTdDeliver[$normalizedSku] ?? 0.0);
 
             $sourceItemSelections[] = $this->sourceSelectionItemFactory->create(
                 [
                     'sourceCode' => $sourceItem->getSourceCode(),
-                    'sku' => $sourceItem->getSku(),
+                    'sku' => trim($sourceItem->getSku()),
                     'qtyToDeduct' => $qtyToDeduct,
                     'qtyAvailable' => $sourceItemQtyAvailable
                 ]
