diff --git a/vendor/magento/module-elasticsearch/SearchAdapter/Filter/Builder/Range.php b/vendor/magento/module-elasticsearch/SearchAdapter/Filter/Builder/Range.php
index a2cab32ea4e6..5cad2317ce9b 100644
--- a/vendor/magento/module-elasticsearch/SearchAdapter/Filter/Builder/Range.php
+++ b/vendor/magento/module-elasticsearch/SearchAdapter/Filter/Builder/Range.php
@@ -26,6 +26,8 @@ public function __construct(
     }
 
     /**
+     * Add the range filters
+     *
      * @param RequestFilterInterface|RangeFilterRequest $filter
      * @return array
      */
@@ -33,10 +35,10 @@ public function buildFilter(RequestFilterInterface $filter)
     {
         $filterQuery = [];
         $fieldName = $this->fieldMapper->getFieldName($filter->getField());
-        if ($filter->getFrom()) {
+        if ($filter->getFrom() !== null) {
             $filterQuery['range'][$fieldName]['gte'] = $filter->getFrom();
         }
-        if ($filter->getTo()) {
+        if ($filter->getTo() !== null) {
             $filterQuery['range'][$fieldName]['lte'] = $filter->getTo();
         }
         return [$filterQuery];
