diff --git a/vendor/magento/module-eav/Model/Entity/Attribute/Backend/Datetime.php b/vendor/magento/module-eav/Model/Entity/Attribute/Backend/Datetime.php
index 6099d329571e8..14a40fb558b41 100644
--- a/vendor/magento/module-eav/Model/Entity/Attribute/Backend/Datetime.php
+++ b/vendor/magento/module-eav/Model/Entity/Attribute/Backend/Datetime.php
@@ -1,9 +1,8 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2011 Adobe
+ * All Rights Reserved.
  */
-
 namespace Magento\Eav\Model\Entity\Attribute\Backend;
 
 /**
@@ -57,6 +56,14 @@ public function beforeSave($object)
             $object->setData($attributeName . '_is_formated', true);
         }
 
+        $defaultValue = $this->getDefaultValue();
+        if ($object->getData($attributeName) === null
+            && $defaultValue !== null
+            && !$object->hasData($attributeName)) {
+            $object->setData($attributeName, $defaultValue);
+            $object->setData($attributeName . '_is_formated', true);
+        }
+
         return $this;
     }
 

