diff --git a/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator.php b/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator.php
index 59c4722c3aad..75ef96c6c7fb 100644
--- a/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator.php
+++ b/vendor/magento/module-catalog-inventory/Model/Quote/Item/QuantityValidator.php
@@ -1,13 +1,12 @@
 <?php
 /**
- * Product inventory data validator
- *
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
  */
 
 namespace Magento\CatalogInventory\Model\Quote\Item;
 
+use Magento\Catalog\Model\Product\Attribute\Source\Status;
 use Magento\CatalogInventory\Api\Data\StockItemInterface;
 use Magento\CatalogInventory\Api\StockRegistryInterface;
 use Magento\CatalogInventory\Api\StockStateInterface;
@@ -27,6 +26,7 @@
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  *
  * @deprecated 100.3.0 Replaced with Multi Source Inventory
+ * @see Multi Source Inventory
  * @link https://devdocs.magento.com/guides/v2.4/inventory/index.html
  * @link https://devdocs.magento.com/guides/v2.4/inventory/inventory-api-reference.html
  */
@@ -156,6 +156,7 @@ public function validate(Observer $observer)
         if ($stockStatus) {
             if ($stockStatus->getStockStatus() === Stock::STOCK_OUT_OF_STOCK
                     || $parentStockStatus && $parentStockStatus->getStockStatus() == Stock::STOCK_OUT_OF_STOCK
+                || (int) $quoteItem->getProduct()->getStatus() !== Status::STATUS_ENABLED
             ) {
                 $hasError = $quoteItem->getStockStateResult()
                     ? $quoteItem->getStockStateResult()->getHasError() : false;
