diff --git a/vendor/magento/module-product-alert/Model/Mailing/AlertProcessor.php b/vendor/magento/module-product-alert/Model/Mailing/AlertProcessor.php
index 83ac4abd896c..a77ca851ac74 100644
--- a/vendor/magento/module-product-alert/Model/Mailing/AlertProcessor.php
+++ b/vendor/magento/module-product-alert/Model/Mailing/AlertProcessor.php
@@ -27,6 +27,8 @@
 
 /**
  * Class for mailing Product Alerts
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  */
 class AlertProcessor
 {
@@ -139,6 +141,7 @@ public function process(string $alertType, array $customerIds, int $websiteId):
      * @param int $websiteId
      * @return array
      * @throws \Exception
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
      */
     private function processAlerts(string $alertType, array $customerIds, int $websiteId): array
     {
@@ -160,6 +163,7 @@ private function processAlerts(string $alertType, array $customerIds, int $websi
         /** @var Website $website */
         $website = $this->storeManager->getWebsite($websiteId);
         $defaultStoreId = $website->getDefaultStore()->getId();
+        $products = [];
 
         /** @var Price|Stock $alert */
         foreach ($collection as $alert) {
@@ -174,7 +178,12 @@ private function processAlerts(string $alertType, array $customerIds, int $websi
                     $customer = $this->customerRepository->getById($alert->getCustomerId());
                 }
 
-                $product = $this->productRepository->getById($alert->getProductId(), false, $defaultStoreId);
+                if (!isset($products[$alert->getProductId()])) {
+                    $product = $this->productRepository->getById($alert->getProductId(), false, $defaultStoreId, true);
+                    $products[$alert->getProductId()] = $product;
+                } else {
+                    $product = $products[$alert->getProductId()];
+                }
 
                 switch ($alertType) {
                     case self::ALERT_TYPE_STOCK:
