diff --git a/vendor/magento/module-customer-graph-ql/Model/Resolver/Cache/Customer/ModelHydrator.php b/vendor/magento/module-customer-graph-ql/Model/Resolver/Cache/Customer/ModelHydrator.php
index 4b4c187bbd949..dba82f13cb3b9 100644
--- a/vendor/magento/module-customer-graph-ql/Model/Resolver/Cache/Customer/ModelHydrator.php
+++ b/vendor/magento/module-customer-graph-ql/Model/Resolver/Cache/Customer/ModelHydrator.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2023 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);
 
@@ -10,12 +10,13 @@
 use Magento\Customer\Model\Data\Customer;
 use Magento\Customer\Model\Data\CustomerFactory;
 use Magento\Framework\EntityManager\HydratorPool;
+use Magento\Framework\ObjectManager\ResetAfterRequestInterface;
 use Magento\GraphQlResolverCache\Model\Resolver\Result\HydratorInterface;
 
 /**
  * Customer resolver data hydrator to rehydrate propagated model.
  */
-class ModelHydrator implements HydratorInterface
+class ModelHydrator implements HydratorInterface, ResetAfterRequestInterface
 {
     /**
      * @var CustomerFactory
@@ -59,4 +60,14 @@ public function hydrate(array &$resolverData): void
             $resolverData['model'] = $this->customerModels[$resolverData['model_id']];
         }
     }
+
+    /**
+     * Reset customerModels
+     *
+     * @return void
+     */
+    public function _resetState(): void
+    {
+        $this->customerModels = [];
+    }
 }

