diff --git a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
index 115e26df31b..f0e0ff73b83 100644
--- a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
+++ b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
@@ -121,24 +121,6 @@ class Helper
      */
     private $categoryLinkFactory;
 
-    /**
-     * @var array
-     */
-    private $productDataKeys = [
-        'weight',
-        'special_price',
-        'cost',
-        'country_of_manufacture',
-        'description',
-        'short_description',
-        'meta_description',
-        'meta_keyword',
-        'meta_title',
-        'page_layout',
-        'custom_design',
-        'gift_wrapping_price'
-    ];
-
     /**
      * Constructor
      *
@@ -221,12 +203,6 @@ class Helper
             $productData['product_has_weight'] = 0;
         }
 
-        foreach ($productData as $key => $value) {
-            if (in_array($key, $this->productDataKeys) && $value === '') {
-                $productData[$key] = null;
-            }
-        }
-
         foreach (['category_ids', 'website_ids'] as $field) {
             if (!isset($productData[$field])) {
                 $productData[$field] = [];
diff --git a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php
index 81ab67bdf26..bd5d03f212c 100644
--- a/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php
+++ b/vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper/AttributeFilter.php
@@ -102,6 +102,7 @@ class AttributeFilter
     {
         $considerUseDefaultsAttribute = !isset($useDefaults[$attribute]) || $useDefaults[$attribute] === '1';
 
-        return ($value === '' && $considerUseDefaultsAttribute && ($product->getData($attribute) === null));
+        return ($value === '' && $considerUseDefaultsAttribute &&
+            ($product->hasData($attribute) && $product->getData($attribute) === null));
     }
 }
