diff --git a/vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Full.php b/vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Full.php
index d36c7507afa8..4d106bab92cb 100644
--- a/vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Full.php
+++ b/vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Full.php
@@ -311,8 +311,6 @@ private function reindexByBatchWithDimensions(
         if (!empty($entityIds)) {
             $this->dimensionTableMaintainer->createMainTmpTable($dimensions);
             $temporaryTable = $this->dimensionTableMaintainer->getMainTmpTable($dimensions);
-            $this->_emptyTable($temporaryTable);
-
             $priceIndexer->executeByDimensions($dimensions, \SplFixedArray::fromArray($entityIds, false));
 
             // Sync data from temp table to index table
@@ -320,6 +318,7 @@ private function reindexByBatchWithDimensions(
                 $temporaryTable,
                 $this->dimensionTableMaintainer->getMainReplicaTable($dimensions)
             );
+            $this->_defaultIndexerResource->getConnection()->dropTable($temporaryTable);
         }
     }
 
@@ -354,7 +353,6 @@ private function reindexBatch(PriceInterface $priceIndexer, Select $batch): void
         if (!empty($entityIds)) {
             // Temporary table will created if not exists
             $idxTableName = $this->_defaultIndexerResource->getIdxTable();
-            $this->_emptyTable($idxTableName);
 
             if ($priceIndexer->getIsComposite()) {
                 $this->_copyRelationIndexData($entityIds);
@@ -489,6 +487,7 @@ private function moveDataFromReplicaTableToReplicaTables(array $dimensions): voi
      * Retrieves the index table that should be used
      *
      * @deprecated 102.0.6
+     * @see only used in another deprecated method: _copyRelationIndexData
      */
     protected function getIndexTargetTable(): string
     {
diff --git a/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Price/CustomOptionPriceModifier.php b/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Price/CustomOptionPriceModifier.php
index 463da8762b7c..9d366584bd5f 100644
--- a/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Price/CustomOptionPriceModifier.php
+++ b/vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Price/CustomOptionPriceModifier.php
@@ -122,8 +122,8 @@ public function modifyPrice(IndexTableStructure $priceTable, array $entityIds =
         $query = $select->crossUpdateFromSelect(['i' => $finalPriceTable]);
         $connection->query($query);
 
-        $connection->delete($coaTable);
-        $connection->delete($copTable);
+        $connection->dropTemporaryTable($coaTable);
+        $connection->dropTemporaryTable($copTable);
     }
 
     /**
