diff --git a/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product.php b/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product.php
index 30be41072242..3e955ae30345 100644
--- a/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product.php
+++ b/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/Product.php
@@ -89,7 +89,7 @@ public function getList(
 
         $this->collectionPreProcessor->process($collection, $searchCriteria, $attributes, $context);
 
-        if ($isChildSearch) {
+        if (!$isChildSearch) {
             $visibilityIds = $isSearch
                 ? $this->visibility->getVisibleInSearchIds()
                 : $this->visibility->getVisibleInCatalogIds();
diff --git a/vendor/magento/module-related-product-graph-ql/Model/Resolver/Batch/AbstractLikedProducts.php b/vendor/magento/module-related-product-graph-ql/Model/Resolver/Batch/AbstractLikedProducts.php
index fac7b23d408e..f35af6f4885d 100644
--- a/vendor/magento/module-related-product-graph-ql/Model/Resolver/Batch/AbstractLikedProducts.php
+++ b/vendor/magento/module-related-product-graph-ql/Model/Resolver/Batch/AbstractLikedProducts.php
@@ -94,9 +94,7 @@ private function findRelations(array $products, array $loadAttributes, int $link
         $this->searchCriteriaBuilder->addFilter('entity_id', $relatedIds, 'in');
         $relatedSearchResult = $this->productDataProvider->getList(
             $this->searchCriteriaBuilder->create(),
-            $loadAttributes,
-            false,
-            true
+            $loadAttributes
         );
         //Filling related products map.
         /** @var \Magento\Catalog\Api\Data\ProductInterface[] $relatedProducts */
