diff --git a/vendor/magento/module-elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php b/vendor/magento/module-elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php
index 0edc63b10f9a..c905cfbdf4ce 100644
--- a/vendor/magento/module-elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php
+++ b/vendor/magento/module-elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php
@@ -250,6 +250,7 @@ private function convertAttribute(Attribute $attribute, array $attributeValues,
      * - "Visible in Advanced Search" (is_visible_in_advanced_search)
      * - "Use in Layered Navigation" (is_filterable)
      * - "Use in Search Results Layered Navigation" (is_filterable_in_search)
+     * - "Use in Sorting in Product Listing" (used_for_sort_by)
      *
      * @param Attribute $attribute
      * @return bool
@@ -261,6 +262,7 @@ private function isAttributeLabelsShouldBeMapped(Attribute $attribute): bool
             || $attribute->getIsVisibleInAdvancedSearch()
             || $attribute->getIsFilterable()
             || $attribute->getIsFilterableInSearch()
+            || $attribute->getUsedForSortBy()
         );
     }
 
