diff --git a/vendor/magento/module-import-export/Model/Import/Entity/AbstractEntity.php b/vendor/magento/module-import-export/Model/Import/Entity/AbstractEntity.php
index 9188f38390a2a..2f2f4a2612890 100644
--- a/vendor/magento/module-import-export/Model/Import/Entity/AbstractEntity.php
+++ b/vendor/magento/module-import-export/Model/Import/Entity/AbstractEntity.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2013 Adobe
+ * All Rights Reserved.
  */
 namespace Magento\ImportExport\Model\Import\Entity;
 
@@ -35,7 +35,7 @@ abstract class AbstractEntity
 
     public const DB_MAX_VARCHAR_LENGTH = 256;
 
-    public const DB_MAX_TEXT_LENGTH = 65536;
+    public const DB_MAX_TEXT_LENGTH = 16_777_215;
 
     public const ERROR_CODE_SYSTEM_EXCEPTION = 'systemException';
     public const ERROR_CODE_COLUMN_NOT_FOUND = 'columnNotFound';

