diff --git a/vendor/magento/module-catalog/Model/Category/Attribute/Backend/DefaultSortby.php b/vendor/magento/module-catalog/Model/Category/Attribute/Backend/DefaultSortby.php
new file mode 100644
index 000000000000..7a10b7f876c8
--- /dev/null
+++ b/vendor/magento/module-catalog/Model/Category/Attribute/Backend/DefaultSortby.php
@@ -0,0 +1,21 @@
+<?php
+/**
+ * Copyright 2024 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ */
+declare(strict_types=1);
+
+namespace Magento\Catalog\Model\Category\Attribute\Backend;
+
+class DefaultSortby extends Sortby
+{
+}
diff --git a/vendor/magento/module-catalog/Setup/Patch/Data/UpdateDefaultSortyByBackendType.php b/vendor/magento/module-catalog/Setup/Patch/Data/UpdateDefaultSortyByBackendType.php
new file mode 100644
index 000000000000..25a09b6e0f3f
--- /dev/null
+++ b/vendor/magento/module-catalog/Setup/Patch/Data/UpdateDefaultSortyByBackendType.php
@@ -0,0 +1,83 @@
+<?php
+/**
+ * Copyright 2024 Adobe
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Adobe.
+ */
+declare(strict_types=1);
+
+namespace Magento\Catalog\Setup\Patch\Data;
+
+use Magento\Catalog\Setup\CategorySetup;
+use Magento\Catalog\Setup\CategorySetupFactory;
+use Magento\Framework\Setup\ModuleDataSetupInterface;
+use Magento\Framework\Setup\Patch\DataPatchInterface;
+use Magento\Framework\Setup\Patch\PatchInterface;
+
+class UpdateDefaultSortyByBackendType implements DataPatchInterface
+{
+    /**
+     * @var ModuleDataSetupInterface
+     */
+    private $moduleDataSetup;
+
+    /**
+     * @var CategorySetupFactory
+     */
+    private $categorySetupFactory;
+
+    /**
+     * PatchInitial constructor.
+     * @param ModuleDataSetupInterface $moduleDataSetup
+     * @param CategorySetupFactory $categorySetupFactory
+     */
+    public function __construct(
+        ModuleDataSetupInterface $moduleDataSetup,
+        CategorySetupFactory $categorySetupFactory
+    ) {
+        $this->moduleDataSetup = $moduleDataSetup;
+        $this->categorySetupFactory = $categorySetupFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function apply(): UpdateDefaultSortyByBackendType
+    {
+        $mediaBackendModel = \Magento\Catalog\Model\Category\Attribute\Backend\DefaultSortby::class;
+        /** @var CategorySetup $categorySetup */
+        $categorySetup = $this->categorySetupFactory->create(['setup' => $this->moduleDataSetup]);
+        $categorySetup->updateAttribute(
+            'catalog_category',
+            'default_sort_by',
+            'backend_model',
+            $mediaBackendModel
+        );
+
+        return $this;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public static function getDependencies()
+    {
+        return [];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getAliases()
+    {
+        return [];
+    }
+}
diff --git a/vendor/magento/module-catalog/etc/di.xml b/vendor/magento/module-catalog/etc/di.xml
index e817bcbb42d2..faf698c45174 100644
--- a/vendor/magento/module-catalog/etc/di.xml
+++ b/vendor/magento/module-catalog/etc/di.xml
@@ -739,6 +739,7 @@
                 <item name="Magento\Catalog\Model\Product\Attribute\Backend\Category" xsi:type="string">int[]</item>
                 <item name="Magento\Catalog\Model\Product\Attribute\Backend\Stock" xsi:type="string">Magento\CatalogInventory\Api\Data\StockItemInterface[]</item>
                 <item name="Magento\Catalog\Model\Category\Attribute\Backend\Sortby" xsi:type="string">string[]</item>
+                <item name="Magento\Catalog\Model\Category\Attribute\Backend\DefaultSortby" xsi:type="string">string</item>
             </argument>
         </arguments>
     </type>
