diff --git a/vendor/magento/module-inventory-in-store-pickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php b/vendor/magento/module-inventory-in-store-pickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php
index b7bb733eb70e..1c69fe702360 100644
--- a/vendor/magento/module-inventory-in-store-pickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php
+++ b/vendor/magento/module-inventory-in-store-pickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php
@@ -1,7 +1,7 @@
 <?php
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2019 Adobe
+ * All Rights Reserved.
  */
 declare(strict_types=1);
 
@@ -45,7 +45,12 @@ public function __construct(ValidatorInterface $validator, DelimiterConfig $deli
     public function execute(string $searchTerm, DataObject $dataObject): void
     {
         $searchQuery = $this->getSearchQuery($searchTerm);
-        if ($this->validator->validate($searchQuery, $dataObject->getData(self::COUNTRY))) {
+        try {
+            $isValid = $this->validator->validate($searchQuery, $dataObject->getData(self::COUNTRY));
+        } catch (\InvalidArgumentException) {
+            $isValid = false;
+        }
+        if ($isValid) {
             $dataObject->setData(self::POSTCODE, $searchQuery);
         } else {
             $dataObject->setData(self::POSTCODE, '');
