diff --git a/vendor/magento/module-store/App/FrontController/Plugin/DefaultStore.php b/vendor/magento/module-store/App/FrontController/Plugin/DefaultStore.php
deleted file mode 100644
index 58340c6cc35..00000000000
--- a/vendor/magento/module-store/App/FrontController/Plugin/DefaultStore.php
+++ /dev/null
@@ -1,73 +0,0 @@
-<?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
- */
-namespace Magento\Store\App\FrontController\Plugin;
-
-use \Magento\Store\Model\StoreResolver\ReaderList;
-use \Magento\Store\Model\ScopeInterface;
-
-/**
- * Plugin to set default store for admin area.
- */
-class DefaultStore
-{
-    /**
-     * @var \Magento\Store\Model\StoreManagerInterface
-     */
-    protected $storeManager;
-
-    /**
-     * @var ReaderList
-     */
-    protected $readerList;
-
-    /**
-     * @var string
-     */
-    protected $runMode;
-
-    /**
-     * @var string
-     */
-    protected $scopeCode;
-
-    /**
-     * Initialize dependencies.
-     *
-     * @param \Magento\Store\Model\StoreManagerInterface $storeManager
-     * @param ReaderList $readerList
-     * @param string $runMode
-     * @param null $scopeCode
-     */
-    public function __construct(
-        \Magento\Store\Model\StoreManagerInterface $storeManager,
-        ReaderList $readerList,
-        $runMode = ScopeInterface::SCOPE_STORE,
-        $scopeCode = null
-    ) {
-        $this->runMode = $scopeCode ? $runMode : ScopeInterface::SCOPE_WEBSITE;
-        $this->scopeCode = $scopeCode;
-        $this->readerList = $readerList;
-        $this->storeManager = $storeManager;
-    }
-
-    /**
-     * Set current store for admin area
-     *
-     * @param \Magento\Framework\App\FrontController $subject
-     * @param \Magento\Framework\App\RequestInterface $request
-     * @return void
-     *
-     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     */
-    public function beforeDispatch(
-        \Magento\Framework\App\FrontController $subject,
-        \Magento\Framework\App\RequestInterface $request
-    ) {
-        $reader = $this->readerList->getReader($this->runMode);
-        $defaultStoreId = $reader->getDefaultStoreId($this->scopeCode);
-        $this->storeManager->setCurrentStore($defaultStoreId);
-    }
-}
diff --git a/vendor/magento/module-store/etc/adminhtml/di.xml b/vendor/magento/module-store/etc/adminhtml/di.xml
index e6e21f6ec0a..26fcbad0ff1 100644
--- a/vendor/magento/module-store/etc/adminhtml/di.xml
+++ b/vendor/magento/module-store/etc/adminhtml/di.xml
@@ -6,9 +6,6 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
-    <type name="Magento\Framework\App\FrontControllerInterface">
-        <plugin name="default_store_setter" type="Magento\Store\App\FrontController\Plugin\DefaultStore" />
-    </type>
     <type name="Magento\Framework\Notification\MessageList">
         <arguments>
             <argument name="messages" xsi:type="array">
diff --git a/vendor/magento/module-store/etc/di.xml b/vendor/magento/module-store/etc/di.xml
index 984a16eb349..643550ac9f8 100644
--- a/vendor/magento/module-store/etc/di.xml
+++ b/vendor/magento/module-store/etc/di.xml
@@ -115,12 +115,6 @@
             <argument name="cache" xsi:type="object">Magento\Framework\App\Cache\Type\Config</argument>
         </arguments>
     </type>
-    <type name="Magento\Store\App\FrontController\Plugin\DefaultStore">
-        <arguments>
-            <argument name="runMode" xsi:type="init_parameter">Magento\Store\Model\StoreManager::PARAM_RUN_TYPE</argument>
-            <argument name="scopeCode" xsi:type="init_parameter">Magento\Store\Model\StoreManager::PARAM_RUN_CODE</argument>
-        </arguments>
-    </type>
     <virtualType name="Magento\Store\Model\ResourceModel\Group\Collection\FetchStrategy" type="Magento\Framework\Data\Collection\Db\FetchStrategy\Cache">
         <arguments>
             <argument name="cache" xsi:type="object">Magento\Framework\App\Cache\Type\Collection</argument>
