diff --git a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
index b8e4747f5fd5f..7b585e4bdc09d 100644
--- a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
+++ b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
@@ -31,7 +31,7 @@ class PhpCookieManager implements CookieManagerInterface
      * RFC 2109 - Page 15
      * http://www.ietf.org/rfc/rfc6265.txt
      */
-    private const MAX_NUM_COOKIES = 50;
+    private const MAX_NUM_COOKIES = 200;
     public const MAX_COOKIE_SIZE = 4096;
     private const EXPIRE_NOW_TIME = 1;
     private const EXPIRE_AT_END_OF_SESSION_TIME = 0;
