diff --git a/vendor/paypal/module-braintree-core/Plugin/SalesOrderGridPlugin.php b/vendor/paypal/module-braintree-core/Plugin/SalesOrderGridPlugin.php
index 46403d35..12fa8cdb 100644
--- a/vendor/paypal/module-braintree-core/Plugin/SalesOrderGridPlugin.php
+++ b/vendor/paypal/module-braintree-core/Plugin/SalesOrderGridPlugin.php
@@ -38,6 +38,17 @@ public function beforeLoad(Collection $subject, bool $printQuery = false, bool $
                 $salesOrderTable . '.entity_id = main_table.' . $primaryKey,
                 $salesOrderTable . '.dispute_status'
             );
+            $wherePart = $subject->getSelect()->getPart('where');
+            if (!empty($wherePart)) {
+                /** @var string $condition */
+                foreach ($wherePart as $key => $condition) {
+                    if (!str_contains($condition, "`created_at`")) {
+                        continue;
+                    }
+                    $wherePart[$key] = str_replace("`created_at`", "main_table.created_at", $condition);
+                }
+                $subject->getSelect()->setPart('where', $wherePart);
+            }
         }

         return [$printQuery, $logQuery];
