diff --git a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
index 8266d27d1a3..90396f44713 100644
--- a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
+++ b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
@@ -5,7 +5,7 @@
  */
 ?>
 <br/>
-<h1>Issue 1</h1>
+<h1><span style="text-decoration: line-through;">Issue 1</span> Patch MC-1</h1>
 
 
 
