diff --git a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
index 8266d27d1a3..fec8cdb91c6 100644
--- a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
+++ b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
@@ -9,7 +9,7 @@
 
 
 
-<h1>Issue 2</h1>
+<h1><span style="text-decoration: line-through;">Issue 2</span> Patch MC-2</h1>
 
 
 
