diff -Nuar a/vendor/magento/module-customer/view/frontend/web/js/block-submit-on-send.js b/vendor/magento/module-customer/view/frontend/web/js/block-submit-on-send.js
--- a/vendor/magento/module-customer/view/frontend/web/js/block-submit-on-send.js
+++ b/vendor/magento/module-customer/view/frontend/web/js/block-submit-on-send.js
@@ -14,9 +14,15 @@ define([
 
         dataForm.submit(function () {
             $(this).find(':submit').attr('disabled', 'disabled');
+
+            if (this.isValid === false) {
+                $(this).find(':submit').prop('disabled', false);
+            }
+            this.isValid = true;
         });
         dataForm.bind('invalid-form.validate', function () {
             $(this).find(':submit').prop('disabled', false);
+            this.isValid = false;
         });
     };
 });
