diff --git a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
index d9459229807..caf6212cd80 100644
--- a/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
+++ b/vendor/magento/module-cms/view/frontend/templates/default/home.phtml
@@ -9,8 +9,8 @@
 
 
 
-<h1><span style="text-decoration: line-through;">Issue 2</span> Patch MC-2</h1>
+<h1><span style="text-decoration: line-through;">Issue 2</span> Patch MC-2, Patch MC-3</h1>
 
 
 
-<h1>Issue 3</h1>
+<h1><span style="text-decoration: line-through;">Issue 3</span> Patch MC-3</h1>
