diff -Naur a/vendor/magento/module-admin-analytics/Controller/Adminhtml/Config/DisableAdminUsage.php b/vendor/magento/module-admin-analytics/Controller/Adminhtml/Config/DisableAdminUsage.php
--- a/vendor/magento/module-admin-analytics/Controller/Adminhtml/Config/DisableAdminUsage.php
+++ b/vendor/magento/module-admin-analytics/Controller/Adminhtml/Config/DisableAdminUsage.php
@@ -23,7 +23,7 @@
     /**
      * @var Factory
      */
-    // MCTEST-1002: dependent patch
+    // MCTEST-1003: dependent patch
     private $configFactory;
 
     /**
