diff --git a/vendor/magento/module-configurable-product-graph-ql/Model/Variant/Collection.php b/vendor/magento/module-configurable-product-graph-ql/Model/Variant/Collection.php
index cd6d78e5c3f..d93541b9e44 100644
--- a/vendor/magento/module-configurable-product-graph-ql/Model/Variant/Collection.php
+++ b/vendor/magento/module-configurable-product-graph-ql/Model/Variant/Collection.php
@@ -17,6 +17,7 @@ use Magento\Framework\Api\SearchCriteriaBuilder;
 use Magento\GraphQl\Model\Query\ContextInterface;
 use Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Product\CollectionProcessorInterface;
 use Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Product\CollectionPostProcessor;
+use Magento\Catalog\Model\Product\Attribute\Source\Status;

 /**
  * Collection for fetching configurable child product data.
@@ -162,6 +163,9 @@ class Collection

             /** @var Product $childProduct */
             foreach ($childCollection as $childProduct) {
+                if ((int)$childProduct->getStatus() !== Status::STATUS_ENABLED) {
+                    continue;
+                }
                 $formattedChild = ['model' => $childProduct, 'sku' => $childProduct->getSku()];
                 $parentId = (int)$childProduct->getParentId();
                 if (!isset($this->childrenMap[$parentId])) {
