diff -Nuar a/vendor/magento/module-catalog-import-export/Model/Import/Product/Type/AbstractType.php b/vendor/magento/module-catalog-import-export/Model/Import/Product/Type/AbstractType.php
index 6571b16c875..bd17cfd2cd7 100644
--- a/vendor/magento/module-catalog-import-export/Model/Import/Product/Type/AbstractType.php
+++ b/vendor/magento/module-catalog-import-export/Model/Import/Product/Type/AbstractType.php
@@ -533,6 +533,7 @@ abstract class AbstractType
             if ($attrParams['is_static']) {
                 continue;
             }
+            $attrCode = mb_strtolower($attrCode);
             if (isset($rowData[$attrCode]) && strlen(trim($rowData[$attrCode]))) {
                 if (in_array($attrParams['type'], ['select', 'boolean'])) {
                     $resultAttrs[$attrCode] = $attrParams['options'][strtolower($rowData[$attrCode])];
