diff -Nuar a/vendor/magento/module-sales-rule/Model/Rule/Condition/Address.php b/vendor/magento/module-sales-rule/Model/Rule/Condition/Address.php
index cf6301cb31a..62c1cc08604 100644
--- a/vendor/magento/module-sales-rule/Model/Rule/Condition/Address.php
+++ b/vendor/magento/module-sales-rule/Model/Rule/Condition/Address.php
@@ -62,6 +62,7 @@ class Address extends \Magento\Rule\Model\Condition\AbstractCondition
     {
         $attributes = [
             'base_subtotal_with_discount' => __('Subtotal (Excl. Tax)'),
+            'base_subtotal_total_incl_tax' => __('Subtotal (Incl. Tax)'),
             'base_subtotal' => __('Subtotal'),
             'total_qty' => __('Total Items Quantity'),
             'weight' => __('Total Weight'),
@@ -99,6 +100,7 @@ class Address extends \Magento\Rule\Model\Condition\AbstractCondition
     {
         switch ($this->getAttribute()) {
             case 'base_subtotal':
+            case 'base_subtotal_total_incl_tax':
             case 'weight':
             case 'total_qty':
                 return 'numeric';
