diff --git a/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php b/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php
index 064bd99b9b6..549f25df707 100644
--- a/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php
+++ b/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php
@@ -96,7 +96,8 @@ class InputParamsResolver
         }
         $this->requestValidator->validate();
         $webapiResolvedParams = [];
-        foreach ($this->getInputData() as $key => $singleEntityParams) {
+        $inputData = $this->getInputData();
+        foreach ($inputData as $key => $singleEntityParams) {
             $webapiResolvedParams[$key] = $this->resolveBulkItemParams($singleEntityParams);
         }
         return $webapiResolvedParams;

