diff --git a/vendor/magento/module-configurable-product-graph-ql/Model/Options/Collection.php b/vendor/magento/module-configurable-product-graph-ql/Model/Options/Collection.php
index 19706e114d1d..fa8b669a1bdd 100644
--- a/vendor/magento/module-configurable-product-graph-ql/Model/Options/Collection.php
+++ b/vendor/magento/module-configurable-product-graph-ql/Model/Options/Collection.php
@@ -111,7 +111,7 @@ public function getAttributesByProductId(int $productId): array
      */
     private function fetch(): array
     {
-        if (empty($this->productIds) || !empty($this->attributeMap)) {
+        if (empty($this->productIds) || array_key_exists(end($this->productIds), $this->attributeMap)) {
             return $this->attributeMap;
         }

