diff --git a/vendor/magento/module-quote-graph-ql/Model/Cart/AddProductsToCart.php b/vendor/magento/module-quote-graph-ql/Model/Cart/AddProductsToCart.php
index 91c77a1a3ec..0360d9ccf54 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Cart/AddProductsToCart.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Cart/AddProductsToCart.php
@@ -13,7 +13,7 @@ use Magento\Quote\Api\CartRepositoryInterface;
 use Magento\Quote\Model\Quote;
 
 /**
- * Add products to cart
+ * Adding products to cart using GraphQL
  */
 class AddProductsToCart
 {
@@ -54,16 +54,6 @@ class AddProductsToCart
             $this->addProductToCart->execute($cart, $cartItemData);
         }
 
-        if ($cart->getData('has_error')) {
-            $e = new GraphQlInputException(__('Shopping cart errors'));
-            $errors = $cart->getErrors();
-            foreach ($errors as $error) {
-                /** @var MessageInterface $error */
-                $e->addError(new GraphQlInputException(__($error->getText())));
-            }
-            throw $e;
-        }
-
         $this->cartRepository->save($cart);
     }
 }
diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItems.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItems.php
index 2674b372861..8017a91b5cf 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItems.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartItems.php
@@ -9,6 +9,7 @@ namespace Magento\QuoteGraphQl\Model\Resolver;
 
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\GraphQl\Config\Element\Field;
+use Magento\Framework\GraphQl\Exception\GraphQlInputException;
 use Magento\Framework\GraphQl\Query\ResolverInterface;
 use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
 use Magento\Quote\Model\Quote\Item as QuoteItem;
@@ -29,6 +30,12 @@ class CartItems implements ResolverInterface
         $cart = $value['model'];
 
         $itemsData = [];
+        if ($cart->getData('has_error')) {
+            $errors = $cart->getErrors();
+            foreach ($errors as $error) {
+                $itemsData[] = new GraphQlInputException(__($error->getText()));
+            }
+        }
         foreach ($cart->getAllVisibleItems() as $cartItem) {
             /**
              * @var QuoteItem $cartItem
