diff --git a/vendor/magento/module-sales-graph-ql/etc/schema.graphqls b/vendor/magento/module-sales-graph-ql/etc/schema.graphqls
index 8a76f51d78e..1e9bbf9dac6 100644
--- a/vendor/magento/module-sales-graph-ql/etc/schema.graphqls
+++ b/vendor/magento/module-sales-graph-ql/etc/schema.graphqls
@@ -69,7 +69,7 @@ type OrderAddress @doc(description: "OrderAddress contains detailed information
     country_code: CountryCodeEnum @doc(description: "The customer's country")
     street: [String!]! @doc(description: "An array of strings that define the street number and name")
     company: String @doc(description: "The customer's company")
-    telephone: String! @doc(description: "The telephone number")
+    telephone: String @doc(description: "The telephone number")
     fax: String @doc(description: "The fax number")
     postcode: String @doc(description: "The customer's order ZIP or postal code")
     city: String! @doc(description: "The city or town")
