diff --git a/vendor/magento/module-inventory-indexer/Model/Queue/UpdateIndexSalabilityStatus/IndexProcessor.php b/vendor/magento/module-inventory-indexer/Model/Queue/UpdateIndexSalabilityStatus/IndexProcessor.php
index cebf1e8cdac..aac82dbd99e 100644
--- a/vendor/magento/module-inventory-indexer/Model/Queue/UpdateIndexSalabilityStatus/IndexProcessor.php
+++ b/vendor/magento/module-inventory-indexer/Model/Queue/UpdateIndexSalabilityStatus/IndexProcessor.php
@@ -116,7 +116,7 @@ class IndexProcessor
     {
         $data = [];
         foreach ($salabilityData as $sku => $isSalable) {
-            $currentStatus = $this->getIndexSalabilityStatus($sku, $stockId);
+            $currentStatus = $this->getIndexSalabilityStatus((string)$sku, $stockId);
             if ($isSalable != $currentStatus && $currentStatus !== null) {
                 $data[$sku] = $isSalable;
             }

