diff --git a/vendor/magento/module-catalog-inventory/Model/StockStateProvider.php b/vendor/magento/module-catalog-inventory/Model/StockStateProvider.php
index bfa854edeaa..3626882a0ff 100644
--- a/vendor/magento/module-catalog-inventory/Model/StockStateProvider.php
+++ b/vendor/magento/module-catalog-inventory/Model/StockStateProvider.php
@@ -135,7 +135,7 @@ class StockStateProvider implements StockStateProviderInterface
         $result->addData($this->checkQtyIncrements($stockItem, $qty)->getData());

         $result->setItemIsQtyDecimal($stockItem->getIsQtyDecimal());
-        if (!$stockItem->getIsQtyDecimal() && (floor($qty) !== $qty)) {
+        if (!$stockItem->getIsQtyDecimal() && (floor($qty) !== (float) $qty)) {
             $result->setHasError(true)
                 ->setMessage(__('You cannot use decimal quantity for this product.'))
                 ->setErrorCode('qty_decimal')
