diff --git a/vendor/magento/module-catalog/Model/ProductRepository.php b/vendor/magento/module-catalog/Model/ProductRepository.php
index fdee74c9559..5e61ea8d4a3 100644
--- a/vendor/magento/module-catalog/Model/ProductRepository.php
+++ b/vendor/magento/module-catalog/Model/ProductRepository.php
@@ -605,7 +605,6 @@ class ProductRepository implements \Magento\Catalog\Api\ProductRepositoryInterfa
                     if ($existingProduct->getData($attributeCode) === $value
                         && $attribute->getScope() !== EavAttributeInterface::SCOPE_GLOBAL_TEXT
                         && !is_array($value)
-                        && $attribute->getData('frontend_input') !== 'media_image'
                         && !$attribute->isStatic()
                         && !array_key_exists($attributeCode, $productDataToChange)
                         && $value !== null
