diff --git a/vendor/magento/module-sales-graph-ql/Model/Resolver/OrderTotal.php b/vendor/magento/module-sales-graph-ql/Model/Resolver/OrderTotal.php
index ab3ace45f33..d767cfd32cd 100644
--- a/vendor/magento/module-sales-graph-ql/Model/Resolver/OrderTotal.php
+++ b/vendor/magento/module-sales-graph-ql/Model/Resolver/OrderTotal.php
@@ -60,7 +60,8 @@ class OrderTotal implements ResolverInterface
                 ],
                 'taxes' => $this->getAppliedShippingTaxesDetails($order),
                 'discounts' => $this->getShippingDiscountDetails($order),
-            ]
+            ],
+            'model' => $order
         ];
     }

