diff --git a/vendor/magento/module-inventory-configurable-product/Pricing/Price/LowestPriceOptionsProvider/StockStatusBaseSelectProcessor.php b/vendor/magento/module-inventory-configurable-product/Pricing/Price/LowestPriceOptionsProvider/StockStatusBaseSelectProcessor.php
index d4b0d4f6515..74cdb54a028 100644
--- a/vendor/magento/module-inventory-configurable-product/Pricing/Price/LowestPriceOptionsProvider/StockStatusBaseSelectProcessor.php
+++ b/vendor/magento/module-inventory-configurable-product/Pricing/Price/LowestPriceOptionsProvider/StockStatusBaseSelectProcessor.php
@@ -129,8 +129,9 @@ class StockStatusBaseSelectProcessor implements BaseSelectProcessorInterface
                 []
             );
         }
-
-        $select->where(sprintf('stock.%1s = ?', $isSalableColumnName), 1);
+        if (!$this->stockConfig->isShowOutOfStock()) {
+            $select->where(sprintf('stock.%1s = ?', $isSalableColumnName), 1);
+        }

         return $select;
     }
