diff --git a/vendor/magento/module-ui/view/base/web/js/grid/columns/date.js b/vendor/magento/module-ui/view/base/web/js/grid/columns/date.js
index 88959cda749..29a589b20f7 100644
--- a/vendor/magento/module-ui/view/base/web/js/grid/columns/date.js
+++ b/vendor/magento/module-ui/view/base/web/js/grid/columns/date.js
@@ -48,7 +48,7 @@ define([

             date = moment.utc(this._super());

-            if (!_.isUndefined(this.timezone)) {
+            if (!_.isUndefined(this.timezone) && moment.tz.zone(this.timezone) !== null) {
                 date = date.tz(this.timezone);
             }

