diff --git a/vendor/magento/module-login-as-customer-graph-ql/Model/LoginAsCustomer/CreateCustomerToken.php b/vendor/magento/module-login-as-customer-graph-ql/Model/LoginAsCustomer/CreateCustomerToken.php
index a10bc10ffb8..45046342fd6 100755
--- a/vendor/magento/module-login-as-customer-graph-ql/Model/LoginAsCustomer/CreateCustomerToken.php
+++ b/vendor/magento/module-login-as-customer-graph-ql/Model/LoginAsCustomer/CreateCustomerToken.php
@@ -52,7 +52,7 @@ class CreateCustomerToken
      */
     public function execute(string $email, StoreInterface $store): array
     {
-        $customer = $this->customerFactory->create()->setWebsiteId((int)$store->getId())->loadByEmail($email);
+        $customer = $this->customerFactory->create()->setWebsiteId((int)$store->getWebsiteId())->loadByEmail($email);
 
         /* Check if customer email exist */
         if (!$customer->getId()) {
