diff -Nuar a/vendor/magento/module-quote/Model/ResourceModel/Quote/Item/Collection.php b/vendor/magento/module-quote/Model/ResourceModel/Quote/Item/Collection.php
index 392a815ed96..79fd2b1495c 100644
--- a/vendor/magento/module-quote/Model/ResourceModel/Quote/Item/Collection.php
+++ b/vendor/magento/module-quote/Model/ResourceModel/Quote/Item/Collection.php
@@ -276,7 +276,7 @@ class Collection extends \Magento\Framework\Model\ResourceModel\Db\VersionContro
             }
         }
         if ($this->recollectQuote && $this->_quote) {
-            $this->_quote->collectTotals();
+            $this->_quote->setTotalsCollectedFlag(false);
         }
         \Magento\Framework\Profiler::stop('QUOTE:' . __METHOD__);
 
