diff --git a/vendor/magento/module-quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php b/vendor/magento/module-quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php
index ccac95b3c04..d7cab05c441 100644
--- a/vendor/magento/module-quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php
+++ b/vendor/magento/module-quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php
@@ -119,9 +119,7 @@ class CollectTotalsObserver implements ObserverInterface
 
         $groupId = null;
         if (empty($customerVatNumber) || false == $this->customerVat->isCountryInEU($customerCountryCode)) {
-            $groupId = $customer->getId() ? $this->groupManagement->getDefaultGroup(
-                $storeId
-            )->getId() : $this->groupManagement->getNotLoggedInGroup()->getId();
+            $groupId = $customer->getId() ? $quote->getCustomerGroupId() : $this->groupManagement->getNotLoggedInGroup()->getId();
         } else {
             // Magento always has to emulate group even if customer uses default billing/shipping address
             $groupId = $this->customerVat->getCustomerGroupIdBasedOnVatNumber(
