diff --git a/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php b/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php
index eeaf4bee1b1..19825c98585 100644
--- a/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php
+++ b/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php
@@ -119,7 +119,6 @@ class Reorder extends Create implements HttpGetActionInterface
                 $this->_getOrderCreateModel()->initFromOrder($order);
                 $resultRedirect->setPath('sales/*');
             } catch (\Magento\Framework\Exception\LocalizedException $e) {
-                $this->logger->critical($e);
                 $this->messageManager->addErrorMessage($e->getMessage());
                 return $resultRedirect->setPath('sales/*');
             } catch (\Exception $e) {
