diff --git a/vendor/magento/module-catalog/view/frontend/templates/product/compare/list.phtml b/vendor/magento/module-catalog/view/frontend/templates/product/compare/list.phtml
index 0bea3ca03de..07cca77178a 100644
--- a/vendor/magento/module-catalog/view/frontend/templates/product/compare/list.phtml
+++ b/vendor/magento/module-catalog/view/frontend/templates/product/compare/list.phtml
@@ -73,7 +73,7 @@
                                               action="<?= $block->escapeUrl($this->helper(Magento\Catalog\Helper\Product\Compare::class)->getAddToCartUrl($item)) ?>"
                                               method="post">
                                             <?= $block->getBlockHtml('formkey') ?>
-                                            <button type="submit" class="action tocart primary">
+                                            <button type="submit" class="action tocart primary" disabled>
                                                 <span><?= $block->escapeHtml(__('Add to Cart')) ?></span>
                                             </button>
                                         </form>
@@ -144,15 +144,13 @@
                 </tbody>
             </table>
         </div>
-        <?php if (!$block->isRedirectToCartEnabled()) :?>
-            <script type="text/x-magento-init">
-            {
-                "[data-role=tocart-form]": {
-                    "catalogAddToCart": {}
-                }
+        <script type="text/x-magento-init">
+        {
+            "[data-role=tocart-form]": {
+                "catalogAddToCart": {}
             }
-            </script>
-        <?php endif; ?>
+        }
+        </script>
 <?php else :?>
     <div class="message info empty"><div><?= $block->escapeHtml(__('You have no items to compare.')) ?></div></div>
 <?php endif; ?>
diff --git a/vendor/magento/module-catalog/view/frontend/templates/product/list.phtml b/vendor/magento/module-catalog/view/frontend/templates/product/list.phtml
index 6a47978f1e5..a831bd7be6f 100644
--- a/vendor/magento/module-catalog/view/frontend/templates/product/list.phtml
+++ b/vendor/magento/module-catalog/view/frontend/templates/product/list.phtml
@@ -154,15 +154,13 @@ $_helper = $block->getData('outputHelper');
         </ol>
     </div>
     <?= $block->getToolbarHtml() ?>
-    <?php if (!$block->isRedirectToCartEnabled()): ?>
-        <script type="text/x-magento-init">
-        {
-            "[data-role=tocart-form], .form.map.checkout": {
-                "catalogAddToCart": {
-                    "product_sku": "<?= $escaper->escapeJs($_product->getSku()) ?>"
-                }
+    <script type="text/x-magento-init">
+    {
+        "[data-role=tocart-form], .form.map.checkout": {
+            "catalogAddToCart": {
+                "product_sku": "<?= $escaper->escapeJs($_product->getSku()) ?>"
             }
         }
-        </script>
-    <?php endif; ?>
+    }
+    </script>
 <?php endif; ?>
