diff --git a/vendor/magento/framework/Stdlib/DateTime/Timezone.php b/vendor/magento/framework/Stdlib/DateTime/Timezone.php
index 118a3e053bd..5786f92a56b 100644
--- a/vendor/magento/framework/Stdlib/DateTime/Timezone.php
+++ b/vendor/magento/framework/Stdlib/DateTime/Timezone.php
@@ -206,25 +206,9 @@ class Timezone implements TimezoneInterface
             case ($date instanceof \DateTimeImmutable):
                 $date = $date->setTimezone($timezone);
                 break;
-            case (!is_numeric($date)):
-                $timeType = $includeTime ? \IntlDateFormatter::SHORT : \IntlDateFormatter::NONE;
-                $formatter = new \IntlDateFormatter(
-                    $this->_localeResolver->getLocale(),
-                    \IntlDateFormatter::SHORT,
-                    $timeType,
-                    $timezone
-                );
-                $timestamp = $formatter->parse($date);
-                $date = $timestamp
-                    ? (new \DateTime('@' . $timestamp))->setTimezone($timezone)
-                    : new \DateTime($date, $timezone);
-                break;
-            case (is_numeric($date)):
-                $date = new \DateTime('@' . $date);
-                $date = $date->setTimezone($timezone);
-                break;
             default:
-                $date = new \DateTime($date, $timezone);
+                $date = new \DateTime(is_numeric($date) ? '@' . $date : $date);
+                $date->setTimezone($timezone);
                 break;
         }

