diff -Nuar a/vendor/magento/module-inventory-source-selection/Model/Request/InventoryRequest.php b/vendor/magento/module-inventory-source-selection/Model/Request/InventoryRequest.php
index 88e4246ea55..d70a5a610c1 100644
--- a/vendor/magento/module-inventory-source-selection/Model/Request/InventoryRequest.php
+++ b/vendor/magento/module-inventory-source-selection/Model/Request/InventoryRequest.php
@@ -74,6 +74,7 @@ public function __construct(
         $this->stockId = $stockId;

         //TODO: Temporary fix for resolving issue with webApi (https://github.com/magento-engcom/msi/issues/1524)
+        $this->items = [];
         foreach ($items as $item) {
             if (false === $item instanceof ItemRequestInterface) {
                 $this->items[] = $this->itemRequestFactory->create([
