diff --git a/vendor/magento/module-customer/Model/Validator/Name.php b/vendor/magento/module-customer/Model/Validator/Name.php
index 9bab241465d..9b4d57f1835 100644
--- a/vendor/magento/module-customer/Model/Validator/Name.php
+++ b/vendor/magento/module-customer/Model/Validator/Name.php
@@ -15,7 +15,7 @@ use Magento\Framework\Validator\AbstractValidator;
  */
 class Name extends AbstractValidator
 {
-    private const PATTERN_NAME = '/(?:[\p{L}\p{M}\,\-\_\.\'\s\d]){1,255}+/u';
+    private const PATTERN_NAME = '/(?:[\p{L}\p{M}\,\-\_\.\'’`\s\d]){1,255}+/u';
 
     /**
      * Validate name fields.

