diff -Nuar a/vendor/magento/framework/Search/Adapter/Mysql/TemporaryStorage.php b/vendor/magento/framework/Search/Adapter/Mysql/TemporaryStorage.php
index dcb977bfa2a..a398c7c0c61 100644
--- a/vendor/magento/framework/Search/Adapter/Mysql/TemporaryStorage.php
+++ b/vendor/magento/framework/Search/Adapter/Mysql/TemporaryStorage.php
@@ -93,7 +93,8 @@ class TemporaryStorage
                     self::FIELD_ENTITY_ID,
                     self::FIELD_SCORE,
                 ],
-                $data
+                $data,
+                \Magento\Framework\DB\Adapter\AdapterInterface::REPLACE
             );
         }
         return $table;
