diff --git a/lib/web/mage/utils/misc.js b/lib/web/mage/utils/misc.js
index b1c0c33324c..e01f10b79cc 100644
--- a/lib/web/mage/utils/misc.js
+++ b/lib/web/mage/utils/misc.js
@@ -93,9 +93,12 @@ define([
             var result = mageFormat;
 
             _.each(map, function (moment, mage) {
-                result = result.replace(mage, moment);
+                result = result.replace(
+                    new RegExp(mage + '(?=([^\u0027]*\u0027[^\u0027]*\u0027)*[^\u0027]*$)'),
+                    moment
+                );
             });
-
+            result = result.replace(/'(.*?)'/g, '[$1]');
             return result;
         },
 
