diff --git a/vendor/magento/module-inventory-shipping/Model/GetSourceSelectionResultFromInvoice.php b/vendor/magento/module-inventory-shipping/Model/GetSourceSelectionResultFromInvoice.php
index 21b3b47a053..ac91197d9be 100644
--- a/vendor/magento/module-inventory-shipping/Model/GetSourceSelectionResultFromInvoice.php
+++ b/vendor/magento/module-inventory-shipping/Model/GetSourceSelectionResultFromInvoice.php
@@ -110,8 +110,7 @@ class GetSourceSelectionResultFromInvoice
         $selectionRequestItems = [];
         foreach ($invoiceItems as $invoiceItem) {
             $orderItem = $invoiceItem->getOrderItem();
-
-            if ($orderItem->isDummy() || !$orderItem->getIsVirtual()) {
+            if (!$orderItem->getQtyToInvoice() > 0 && $orderItem->getLockedDoInvoice() || !$orderItem->getIsVirtual()) {
                 continue;
             }
 
