diff -Nuar a/vendor/magento/module-customer/Model/CustomerExtractor.php b/vendor/magento/module-customer/Model/CustomerExtractor.php
index be30b979947..6bcc3b8b83c 100644
--- a/vendor/magento/module-customer/Model/CustomerExtractor.php
+++ b/vendor/magento/module-customer/Model/CustomerExtractor.php
@@ -88,15 +88,20 @@ class CustomerExtractor
             $customerData,
             \Magento\Customer\Api\Data\CustomerInterface::class
         );
+
         $store = $this->storeManager->getStore();
+        $storeId = $store->getId();
+
         if ($isGroupIdEmpty) {
+            $groupId = isset($customerData['group_id']) ? $customerData['group_id']
+                : $this->customerGroupManagement->getDefaultGroup($storeId)->getId();
             $customerDataObject->setGroupId(
-                $this->customerGroupManagement->getDefaultGroup($store->getId())->getId()
+                $groupId
             );
         }
 
         $customerDataObject->setWebsiteId($store->getWebsiteId());
-        $customerDataObject->setStoreId($store->getId());
+        $customerDataObject->setStoreId($storeId);
 
         return $customerDataObject;
     }
