diff --git a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
index 1f454518e43..b3dcc784917 100644
--- a/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
+++ b/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php
@@ -29,7 +29,7 @@ class PhpCookieManager implements CookieManagerInterface
      * RFC 2109 - Page 15
      * http://www.ietf.org/rfc/rfc6265.txt
      */
-    private const MAX_NUM_COOKIES = 50;
+    private const MAX_NUM_COOKIES = 200;
     public const MAX_COOKIE_SIZE = 4096;
     private const EXPIRE_NOW_TIME = 1;
     private const EXPIRE_AT_END_OF_SESSION_TIME = 0;
