diff --git a/vendor/magento/module-catalog/Controller/Product/View.php b/vendor/magento/module-catalog/Controller/Product/View.php
index 570b8f541b7..06dcc45095f 100644
--- a/vendor/magento/module-catalog/Controller/Product/View.php
+++ b/vendor/magento/module-catalog/Controller/Product/View.php
@@ -125,7 +125,7 @@ class View extends ProductAction implements HttpGetActionInterface, HttpPostActi
                 return;
             }
             $resultRedirect = $this->resultRedirectFactory->create();
-            $resultRedirect->setRefererOrBaseUrl();
+            $resultRedirect->setUrl($this->_url->getCurrentUrl());
             return $resultRedirect;
         }

