diff --git a/vendor/magento/module-inventory-in-store-pickup-multishipping/Plugin/Checkout/Controller/Index/Index/DisableMultishippingPlugin.php b/vendor/magento/module-inventory-in-store-pickup-multishipping/Plugin/Checkout/Controller/Index/Index/DisableMultishippingPlugin.php
deleted file mode 100644
index 52bc9bf9a53..00000000000
--- a/vendor/magento/module-inventory-in-store-pickup-multishipping/Plugin/Checkout/Controller/Index/Index/DisableMultishippingPlugin.php
+++ /dev/null
@@ -1,47 +0,0 @@
-<?php
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
- */
-declare(strict_types=1);
-
-namespace Magento\InventoryInStorePickupMultishipping\Plugin\Checkout\Controller\Index\Index;
-
-use Magento\Checkout\Model\Cart;
-
-/**
- * Turns Off multiple address checkout for Quote.
- *
- * @TODO remove when fix from core will be delivered. @see https://github.com/magento/magento2/pull/24072
- */
-class DisableMultishippingPlugin
-{
-    /**
-     * @var Cart
-     */
-    private $cart;
-
-    /**
-     * @param Cart $cart
-     */
-    public function __construct(Cart $cart)
-    {
-        $this->cart = $cart;
-    }
-
-    /**
-     * Disable multiple address checkout.
-     *
-     * @param \Magento\Framework\App\Action\Action $subject
-     * @return void
-     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     */
-    public function beforeExecute(\Magento\Framework\App\Action\Action $subject): void
-    {
-        $quote = $this->cart->getQuote();
-        if ($quote->getIsMultiShipping()) {
-            $quote->setIsMultiShipping(0);
-            $this->cart->saveQuote();
-        }
-    }
-}
diff --git a/vendor/magento/module-inventory-in-store-pickup-multishipping/etc/frontend/di.xml b/vendor/magento/module-inventory-in-store-pickup-multishipping/etc/frontend/di.xml
deleted file mode 100644
index 7509b1eff91..00000000000
--- a/vendor/magento/module-inventory-in-store-pickup-multishipping/etc/frontend/di.xml
+++ /dev/null
@@ -1,12 +0,0 @@
-<?xml version="1.0"?>
-<!--
-/**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
- */
--->
-<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
-    <type name="Magento\Checkout\Controller\Index\Index">
-        <plugin name="multishipping_disabler" type="Magento\InventoryInStorePickupMultishipping\Plugin\Checkout\Controller\Index\Index\DisableMultishippingPlugin" sortOrder="50" />
-    </type>
-</config>
