diff -Nuar a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php
index 6a57a7662af..66cc9ed11ed 100644
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/CartPrices.php
@@ -45,6 +45,12 @@ class CartPrices implements ResolverInterface
 
         /** @var Quote $quote */
         $quote = $value['model'];
+        /**
+         * To calculate a right discount value
+         * before calculate totals
+         * need to reset Cart Fixed Rules in the quote
+         */
+        $quote->setCartFixedRules([]);
         $cartTotals = $this->totalsCollector->collectQuoteTotals($quote);
         $currency = $quote->getQuoteCurrencyCode();
 
