diff --git a/vendor/magento/module-catalog-rule/Model/Indexer/ReindexRuleProduct.php b/vendor/magento/module-catalog-rule/Model/Indexer/ReindexRuleProduct.php
index 944710773123..f4d397a25fb7 100644
--- a/vendor/magento/module-catalog-rule/Model/Indexer/ReindexRuleProduct.php
+++ b/vendor/magento/module-catalog-rule/Model/Indexer/ReindexRuleProduct.php
@@ -109,7 +109,7 @@ public function execute(Rule $rule, $batchCount, $useAdditionalTable = false)
                 : 0;

             foreach ($productIds as $productId => $validationByWebsite) {
-                if (empty($validationByWebsite[$websiteId])) {
+                if (!isset($validationByWebsite[$websiteId]) || $validationByWebsite[$websiteId] === null) {
                     continue;
                 }


