diff --git a/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php b/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php
index e212f33a35d..a1d13e5d353 100644
--- a/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php
+++ b/vendor/magento/module-catalog-graph-ql/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php
@@ -64,7 +64,7 @@ class ExtractDataFromCategoryTree
                 if (empty($tree)) {
                     $tree = $currentLevelTree;
                 }
-                $tree = $this->mergeCategoriesTrees($currentLevelTree, $tree);
+                $tree = $this->mergeCategoriesTrees($tree, $currentLevelTree);
             }
         }

