diff --git a/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockDataToCollection.php b/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockDataToCollection.php
index cf2ca38b238..5ec2d12a2e6 100644
--- a/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockDataToCollection.php
+++ b/vendor/magento/module-inventory-catalog/Model/ResourceModel/AddStockDataToCollection.php
@@ -55,7 +55,7 @@ class AddStockDataToCollection
             $isSalableColumnName = 'stock_status';
             $resource = $collection->getResource();
             $collection->getSelect()
-                ->join(
+                ->{$isFilterInStock ? 'join' : 'joinLeft'}(
                     ['stock_status_index' => $resource->getTable('cataloginventory_stock_status')],
                     sprintf('%s.entity_id = stock_status_index.product_id', Collection::MAIN_TABLE_ALIAS),
                     [IndexStructure::IS_SALABLE => $isSalableColumnName]
