diff --git a/vendor/magento/module-email/Model/Plugin/GetUrl.php b/vendor/magento/module-email/Model/Plugin/GetUrl.php
new file mode 100644
index 00000000000..f64e07dee2a
--- /dev/null
+++ b/vendor/magento/module-email/Model/Plugin/GetUrl.php
@@ -0,0 +1,34 @@
+<?php
+/**
+ * Copyright Â© Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Email\Model\Plugin;
+
+class GetUrl
+{
+    /**
+     * Generate unique Urls/links separated by store in \Magento\Email\Model\AbstractTemplate `getUrl` function.
+     *
+     * @param \Magento\Email\Model\AbstractTemplate $subject
+     * @param \Magento\Store\Model\Store $store
+     * @param string $route
+     * @param array $params
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeGetUrl(
+        \Magento\Email\Model\AbstractTemplate $subject,
+        \Magento\Store\Model\Store $store,
+        $route = '',
+        $params = []
+    ) {
+        /**
+         * Pass extra parameter to distinguish stores urls for property \Magento\Email\Model\AbstractTemplate `getUrl`
+         * in multi-store environment
+         */
+        $params['_escape_params'] = $store->getCode();
+
+        return [$store, $route, $params];
+    }
+}
diff --git a/vendor/magento/module-email/Model/Template/Filter.php b/vendor/magento/module-email/Model/Template/Filter.php
index 648e4ab8fc3..d1d54fd0567 100644
--- a/vendor/magento/module-email/Model/Template/Filter.php
+++ b/vendor/magento/module-email/Model/Template/Filter.php
@@ -567,6 +567,13 @@ class Filter extends \Magento\Framework\Filter\Template
             unset($params['url']);
         }

+        /**
+         * Pass extra parameter to distinguish stores urls for property Magento\Framework\Url $cacheUrl
+         * in multi-store environment
+         */
+        $this->urlModel->setScope($this->_storeManager->getStore());
+        $params['_escape_params'] = $this->_storeManager->getStore()->getCode();
+
         return $this->urlModel->getUrl($path, $params);
     }

diff --git a/vendor/magento/module-email/etc/di.xml b/vendor/magento/module-email/etc/di.xml
index 5f6e30d591b..31b807ff53a 100644
--- a/vendor/magento/module-email/etc/di.xml
+++ b/vendor/magento/module-email/etc/di.xml
@@ -61,6 +61,9 @@
         <plugin name="WindowsSmtpConfig" type="Magento\Email\Model\Plugin\WindowsSmtpConfig" />
         <plugin name="EmailDisable" type="Magento\Email\Model\Mail\TransportInterfacePlugin" />
     </type>
+    <type name="Magento\Email\Model\AbstractTemplate">
+        <plugin name="EmailTemplateLinkUrl" type="Magento\Email\Model\Plugin\GetUrl"/>
+    </type>
     <type name="Magento\Config\Model\Config\TypePool">
         <arguments>
             <argument name="sensitive" xsi:type="array">
