diff --git a/vendor/magento/module-persistent/Model/QuoteManager.php b/vendor/magento/module-persistent/Model/QuoteManager.php
index 35b07ebdb7c4..c649d0252c29 100644
--- a/vendor/magento/module-persistent/Model/QuoteManager.php
+++ b/vendor/magento/module-persistent/Model/QuoteManager.php
@@ -182,6 +182,7 @@ public function convertCustomerCartToGuest()
             $quote->getAddressesCollection()->walk('setCustomerId', ['customerId' => null]);
             $quote->getAddressesCollection()->walk('setEmail', ['email' => null]);
             $quote->collectTotals();
+            $quote->getCustomer()->setId(null);
             $this->persistentSession->getSession()->removePersistentCookie();
             $this->persistentSession->setSession(null);
             $this->quoteRepository->save($quote);
