diff --git a/vendor/magento/module-inventory-configurable-product-frontend-ui/view/frontend/web/js/configurable-variation-qty.js b/vendor/magento/module-inventory-configurable-product-frontend-ui/view/frontend/web/js/configurable-variation-qty.js
index 7a1c034c8c4..74138cc0156 100644
--- a/vendor/magento/module-inventory-configurable-product-frontend-ui/view/frontend/web/js/configurable-variation-qty.js
+++ b/vendor/magento/module-inventory-configurable-product-frontend-ui/view/frontend/web/js/configurable-variation-qty.js
@@ -29,7 +29,7 @@ define([
                     'salesChannelCode': salesChannelCode
                 }
             }).done(function (response) {
-                if (response.qty !== null) {
+                if (response.qty !== null && response.qty > 0) {
                     productQtyInfo.text(response.qty);
                     productQtyInfoBlock.show();
                 } else {
