diff --git a/vendor/magento/module-sales-rule/view/frontend/web/js/action/select-payment-method-mixin.js b/vendor/magento/module-sales-rule/view/frontend/web/js/action/select-payment-method-mixin.js
index 50d54d4e597..7639f294b53 100644
--- a/vendor/magento/module-sales-rule/view/frontend/web/js/action/select-payment-method-mixin.js
+++ b/vendor/magento/module-sales-rule/view/frontend/web/js/action/select-payment-method-mixin.js
@@ -19,6 +19,10 @@ define([
 
             originalSelectPaymentMethodAction(paymentMethod);
 
+            if (paymentMethod === null) {
+                return;
+            }
+
             $.when(
                 setPaymentInformationAction(
                     messageContainer,
