diff -Nuar a/vendor/magento/module-ui/view/base/web/js/lib/validation/rules.js b/vendor/magento/module-ui/view/base/web/js/lib/validation/rules.js
index 08f67955976..88f2b2508e5 100644
--- a/vendor/magento/module-ui/view/base/web/js/lib/validation/rules.js
+++ b/vendor/magento/module-ui/view/base/web/js/lib/validation/rules.js
@@ -60,13 +60,13 @@ define([
     return _.mapObject({
         'min_text_length': [
             function (value, params) {
-                return _.isUndefined(value) || value.length === 0 || value.length >= +params;
+                return _.isUndefined(value) || _.isNull(value) || value.length === 0 || value.length >= +params;
             },
             $.mage.__('Please enter more or equal than {0} symbols.')
         ],
         'max_text_length': [
             function (value, params) {
-                return !_.isUndefined(value) && value.length <= +params;
+                return _.isUndefined(value) || _.isNull(value) || value.length <= +params;
             },
             $.mage.__('Please enter less or equal than {0} symbols.')
         ],
