diff --git a/vendor/dotmailer/dotmailer-magento2-extension/Api/Data/CouponAttributeInterface.php b/vendor/dotmailer/dotmailer-magento2-extension/Api/Data/CouponAttributeInterface.php
index 1d64a4fe..14efa0bf 100644
--- a/vendor/dotmailer/dotmailer-magento2-extension/Api/Data/CouponAttributeInterface.php
+++ b/vendor/dotmailer/dotmailer-magento2-extension/Api/Data/CouponAttributeInterface.php
@@ -25,7 +25,7 @@ public function getEmail();
     public function setEmail($email);

     /**
-     * @return \DateTime|null
+     * @return string|null
      */
     public function getExpiresAt();

diff --git a/vendor/dotmailer/dotmailer-magento2-extension/Model/Coupon/CouponAttribute.php b/vendor/dotmailer/dotmailer-magento2-extension/Model/Coupon/CouponAttribute.php
index e9d03337..431af981 100644
--- a/vendor/dotmailer/dotmailer-magento2-extension/Model/Coupon/CouponAttribute.php
+++ b/vendor/dotmailer/dotmailer-magento2-extension/Model/Coupon/CouponAttribute.php
@@ -2,6 +2,7 @@

 namespace Dotdigitalgroup\Email\Model\Coupon;

+use DateTime;
 use Dotdigitalgroup\Email\Api\Data\CouponAttributeInterface;
 use Dotdigitalgroup\Email\Model\DateTimeFactory;
 use Magento\Framework\Api\AttributeValueFactory;
@@ -93,11 +94,7 @@ public function setEmail($email)
      */
     public function getExpiresAt()
     {
-        if ($expiresAt = $this->getData(self::EXPIRES_AT)) {
-            return $this->dateTimeFactory->create(['time' => $expiresAt])
-                ->getUtcDate();
-        }
-        return null;
+        return $this->getData(self::EXPIRES_AT);
     }

     /**
@@ -105,8 +102,20 @@ public function getExpiresAt()
      */
     public function setExpiresAt(string $expiresAt)
     {
-        $this->setData(self::EXPIRES_AT, $expiresAt);
-        return $this;
+        return $this->setData(self::EXPIRES_AT, $expiresAt);
+    }
+
+    /**
+     * @return DateTime|null
+     */
+    public function getExpiresAtDate()
+    {
+        $expiresAt = $this->getExpiresAt();
+        if ($expiresAt) {
+            return $this->dateTimeFactory->create(['time' => $expiresAt])
+                ->getUtcDate();
+        }
+        return null;
     }

     /**
diff --git a/vendor/dotmailer/dotmailer-magento2-extension/Plugin/CouponExpiredPlugin.php b/vendor/dotmailer/dotmailer-magento2-extension/Plugin/CouponExpiredPlugin.php
index df97fffd..f847f5ad 100644
--- a/vendor/dotmailer/dotmailer-magento2-extension/Plugin/CouponExpiredPlugin.php
+++ b/vendor/dotmailer/dotmailer-magento2-extension/Plugin/CouponExpiredPlugin.php
@@ -76,7 +76,7 @@ public function afterCanProcessRule(Utility $utility, bool $result, Rule $rule,
             }

             if ($ddgExtensionAttributes = $coupon->getExtensionAttributes()->getDdgExtensionAttributes()) {
-                $expiresAt = $ddgExtensionAttributes->getExpiresAt();
+                $expiresAt = $ddgExtensionAttributes->getExpiresAtDate();
                 if ($expiresAt && $this->timezone->date($expiresAt) < $this->timezone->date()) {
                     // individual coupon has expired
                     $rule->setIsValidForAddress($address, false);
