diff --git a/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php b/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php
index 995a6c216d3..c68cf190593 100644
--- a/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php
+++ b/vendor/magento/module-sales/Controller/Adminhtml/Order/Create/Reorder.php
@@ -90,10 +90,18 @@ class Reorder extends \Magento\Sales\Controller\Adminhtml\Order\Create
             }
             $resultRedirect->setPath('sales/order/view', ['order_id' => $orderId]);
         } else {
-            $order->setReordered(true);
-            $this->_getSession()->setUseOldShippingMethod(true);
-            $this->_getOrderCreateModel()->initFromOrder($order);
-            $resultRedirect->setPath('sales/*');
+            try {
+                $order->setReordered(true);
+                $this->_getSession()->setUseOldShippingMethod(true);
+                $this->_getOrderCreateModel()->initFromOrder($order);
+                $resultRedirect->setPath('sales/*');
+            } catch (\Magento\Framework\Exception\LocalizedException $e) {
+                $this->messageManager->addErrorMessage($e->getMessage());
+                return $resultRedirect->setPath('sales/*');
+            } catch (\Exception $e) {
+                $this->messageManager->addException($e, __('Error while processing order.'));
+                return $resultRedirect->setPath('sales/*');
+            }
         }
 
         return $resultRedirect;
