diff --git a/vendor/magento/module-sales/Model/Order/Email/Sender/ShipmentCommentSender.php b/vendor/magento/module-sales/Model/Order/Email/Sender/ShipmentCommentSender.php
index ad305c8..3a4e818 100644
--- a/vendor/magento/module-sales/Model/Order/Email/Sender/ShipmentCommentSender.php
+++ b/vendor/magento/module-sales/Model/Order/Email/Sender/ShipmentCommentSender.php
@@ -75,6 +75,7 @@ class ShipmentCommentSender extends NotifySender
             'formattedBillingAddress' => $this->getFormattedBillingAddress($order),
             'order_data' => [
                 'customer_name' => $order->getCustomerName(),
+                'is_not_virtual' => $order->getIsNotVirtual(),
                 'frontend_status_label' => $order->getFrontendStatusLabel()
             ]
         ];
