diff --git a/vendor/magento/module-inventory-import-export/Model/Import/SourceItemConvert.php b/vendor/magento/module-inventory-import-export/Model/Import/SourceItemConvert.php
index 93132f07eca..d688d89063b 100755
--- a/vendor/magento/module-inventory-import-export/Model/Import/SourceItemConvert.php
+++ b/vendor/magento/module-inventory-import-export/Model/Import/SourceItemConvert.php
@@ -27,6 +27,7 @@ class SourceItemConvert

     /**
      * Converts a data in sourceItem list.
+     *
      * @param array $bunch
      * @return SourceItemInterface[]
      */
@@ -40,9 +41,10 @@ class SourceItemConvert
             $sourceItem->setSku($rowData[Sources::COL_SKU]);
             $sourceItem->setQuantity((float)$rowData[Sources::COL_QTY]);

-            $status = (int)($rowData[Sources::COL_QTY] > 0);
             if (isset($rowData[Sources::COL_STATUS])) {
                 $status = (int)$rowData[Sources::COL_STATUS];
+            } else {
+                $status = 1;
             }
             $sourceItem->setStatus($status);

