diff --git a/vendor/magento/module-reports/Block/Adminhtml/Sales/Coupons/Grid.php b/vendor/magento/module-reports/Block/Adminhtml/Sales/Coupons/Grid.php
index 69ce13c52a2..358a0e85281 100644
--- a/vendor/magento/module-reports/Block/Adminhtml/Sales/Coupons/Grid.php
+++ b/vendor/magento/module-reports/Block/Adminhtml/Sales/Coupons/Grid.php
@@ -21,7 +21,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     protected $_columnGroupBy = 'period';

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      * @codeCoverageIgnore
      */
     protected function _construct()
@@ -32,7 +32,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     public function getResourceCollectionName()
     {
@@ -44,7 +44,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
      */
     protected function _prepareColumns()
@@ -100,9 +100,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
             ]
         );

-        if ($this->getFilterData()->getStoreIds()) {
-            $this->setStoreIds(explode(',', $this->getFilterData()->getStoreIds()));
-        }
+        $this->setStoreIds($this->_getStoreIds());
         $currencyCode = $this->getCurrentCurrencyCode();
         $rate = $this->getRate($currencyCode);

diff --git a/vendor/magento/module-reports/Block/Adminhtml/Sales/Invoiced/Grid.php b/vendor/magento/module-reports/Block/Adminhtml/Sales/Invoiced/Grid.php
index 66ab7fd1bb7..c3e263475a5 100644
--- a/vendor/magento/module-reports/Block/Adminhtml/Sales/Invoiced/Grid.php
+++ b/vendor/magento/module-reports/Block/Adminhtml/Sales/Invoiced/Grid.php
@@ -22,7 +22,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     protected $_columnGroupBy = 'period';

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      * @codeCoverageIgnore
      */
     protected function _construct()
@@ -32,7 +32,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     public function getResourceCollectionName()
     {
@@ -42,7 +42,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     protected function _prepareColumns()
     {
@@ -87,9 +87,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
             ]
         );

-        if ($this->getFilterData()->getStoreIds()) {
-            $this->setStoreIds(explode(',', $this->getFilterData()->getStoreIds()));
-        }
+        $this->setStoreIds($this->_getStoreIds());
         $currencyCode = $this->getCurrentCurrencyCode();
         $rate = $this->getRate($currencyCode);

diff --git a/vendor/magento/module-reports/Block/Adminhtml/Sales/Refunded/Grid.php b/vendor/magento/module-reports/Block/Adminhtml/Sales/Refunded/Grid.php
index f696c85dd95..6d3f26b63ab 100644
--- a/vendor/magento/module-reports/Block/Adminhtml/Sales/Refunded/Grid.php
+++ b/vendor/magento/module-reports/Block/Adminhtml/Sales/Refunded/Grid.php
@@ -22,7 +22,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     protected $_columnGroupBy = 'period';

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      * @codeCoverageIgnore
      */
     protected function _construct()
@@ -32,7 +32,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     public function getResourceCollectionName()
     {
@@ -42,7 +42,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     protected function _prepareColumns()
     {
@@ -74,9 +74,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
             ]
         );

-        if ($this->getFilterData()->getStoreIds()) {
-            $this->setStoreIds(explode(',', $this->getFilterData()->getStoreIds()));
-        }
+        $this->setStoreIds($this->_getStoreIds());
         $currencyCode = $this->getCurrentCurrencyCode();
         $rate = $this->getRate($currencyCode);

diff --git a/vendor/magento/module-reports/Block/Adminhtml/Sales/Shipping/Grid.php b/vendor/magento/module-reports/Block/Adminhtml/Sales/Shipping/Grid.php
index 851e7cdfdda..b354f4944b1 100644
--- a/vendor/magento/module-reports/Block/Adminhtml/Sales/Shipping/Grid.php
+++ b/vendor/magento/module-reports/Block/Adminhtml/Sales/Shipping/Grid.php
@@ -22,7 +22,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     protected $_columnGroupBy = 'period';

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      * @codeCoverageIgnore
      */
     protected function _construct()
@@ -33,7 +33,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     public function getResourceCollectionName()
     {
@@ -43,7 +43,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     }

     /**
-     * {@inheritdoc}
+     * @inheritdoc
      */
     protected function _prepareColumns()
     {
@@ -87,10 +87,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
             ]
         );

-        if ($this->getFilterData()->getStoreIds()) {
-            $this->setStoreIds(explode(',', $this->getFilterData()->getStoreIds()));
-        }
-
+        $this->setStoreIds($this->_getStoreIds());
         $currencyCode = $this->getCurrentCurrencyCode();
         $rate = $this->getRate($currencyCode);

diff --git a/vendor/magento/module-reports/Block/Adminhtml/Sales/Tax/Grid.php b/vendor/magento/module-reports/Block/Adminhtml/Sales/Tax/Grid.php
index f65a2b964fb..e209ea04a60 100644
--- a/vendor/magento/module-reports/Block/Adminhtml/Sales/Tax/Grid.php
+++ b/vendor/magento/module-reports/Block/Adminhtml/Sales/Tax/Grid.php
@@ -22,8 +22,6 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
     protected $_columnGroupBy = 'period';

     /**
-     * Config factory
-     *
      * @var \Magento\Sales\Model\Order\ConfigFactory
      */
     protected $_configFactory;
@@ -131,9 +129,7 @@ class Grid extends \Magento\Reports\Block\Adminhtml\Grid\AbstractGrid
             ]
         );

-        if ($this->getFilterData()->getStoreIds()) {
-            $this->setStoreIds(explode(',', $this->getFilterData()->getStoreIds()));
-        }
+        $this->setStoreIds($this->_getStoreIds());
         $currencyCode = $this->getCurrentCurrencyCode();

         $this->addColumn(

