diff --git a/vendor/magento/module-catalog-url-rewrite/Model/Category/Plugin/Category/Move.php b/vendor/magento/module-catalog-url-rewrite/Model/Category/Plugin/Category/Move.php
index f3984bf7d62..5a5b805ba18 100644
--- a/vendor/magento/module-catalog-url-rewrite/Model/Category/Plugin/Category/Move.php
+++ b/vendor/magento/module-catalog-url-rewrite/Model/Category/Plugin/Category/Move.php
@@ -67,13 +67,11 @@ class Move
         $categoryStoreId = $category->getStoreId();
         foreach ($category->getStoreIds() as $storeId) {
             $category->setStoreId($storeId);
-            if (!$this->isGlobalScope($storeId)) {
-                $this->updateCategoryUrlKeyForStore($category);
-                $category->unsUrlPath();
-                $category->setUrlPath($this->categoryUrlPathGenerator->getUrlPath($category));
-                $category->getResource()->saveAttribute($category, 'url_path');
-                $this->updateUrlPathForChildren($category);
-            }
+            $this->updateCategoryUrlKeyForStore($category);
+            $category->unsUrlPath();
+            $category->setUrlPath($this->categoryUrlPathGenerator->getUrlPath($category));
+            $category->getResource()->saveAttribute($category, 'url_path');
+            $this->updateUrlPathForChildren($category);
         }
         $category->setStoreId($categoryStoreId);
 
