diff -Nuar a/vendor/magento/module-downloadable-import-export/Model/Export/RowCustomizer.php b/vendor/magento/module-downloadable-import-export/Model/Export/RowCustomizer.php
index daa874e829e..5dc98f2d150 100644
--- a/vendor/magento/module-downloadable-import-export/Model/Export/RowCustomizer.php
+++ b/vendor/magento/module-downloadable-import-export/Model/Export/RowCustomizer.php
@@ -82,7 +82,9 @@ class RowCustomizer implements RowCustomizerInterface
             ->addAttributeToSelect('samples_title');
         // set global scope during export
         $this->storeManager->setCurrentStore(Store::DEFAULT_STORE_ID);
-        foreach ($collection as $product) {
+
+        while ($product = $productCollection->fetchItem()) {
+            /** @var $product \Magento\Catalog\Api\Data\ProductInterface */
             $productLinks = $this->linkRepository->getLinksByProduct($product);
             $productSamples = $this->sampleRepository->getSamplesByProduct($product);
             $this->downloadableData[$product->getId()] = [];
