diff --git a/vendor/magento/module-sales-rule/Model/RulesApplier.php b/vendor/magento/module-sales-rule/Model/RulesApplier.php
index 270732c8e02..5b674514f92 100644
--- a/vendor/magento/module-sales-rule/Model/RulesApplier.php
+++ b/vendor/magento/module-sales-rule/Model/RulesApplier.php
@@ -222,7 +222,7 @@ class RulesApplier
          * Discount can be applied on price included tax
          */
 
-        $this->validatorUtility->minFix($discountData, $item, $qty);
+        $this->validatorUtility->minFix($discountData, $item, $item->getTotalQty());
 
         return $discountData;
     }
