diff --git a/vendor/magento/module-configurable-product-graph-ql/Model/Cart/BuyRequest/SuperAttributeDataProvider.php b/vendor/magento/module-configurable-product-graph-ql/Model/Cart/BuyRequest/SuperAttributeDataProvider.php
index 0fa4b8da508..4877450b171 100644
--- a/vendor/magento/module-configurable-product-graph-ql/Model/Cart/BuyRequest/SuperAttributeDataProvider.php
+++ b/vendor/magento/module-configurable-product-graph-ql/Model/Cart/BuyRequest/SuperAttributeDataProvider.php
@@ -93,7 +93,7 @@ class SuperAttributeDataProvider implements BuyRequestDataProviderInterface
             throw new GraphQlNoSuchEntityException(__('Could not find specified product.'));
         }

-        $this->checkProductStock($sku, (float) $qty, (int) $cart->getStoreId());
+        $this->checkProductStock($sku, (float) $qty, (int) $cart->getStore()->getWebsiteId());

         $configurableProductLinks = $parentProduct->getExtensionAttributes()->getConfigurableProductLinks();
         if (!in_array($product->getId(), $configurableProductLinks)) {
