diff --git a/vendor/magento/module-swatches/view/frontend/web/js/swatch-renderer.js b/vendor/magento/module-swatches/view/frontend/web/js/swatch-renderer.js
index 2571c0385dab..b98776d97ef6 100644
--- a/vendor/magento/module-swatches/view/frontend/web/js/swatch-renderer.js
+++ b/vendor/magento/module-swatches/view/frontend/web/js/swatch-renderer.js
@@ -1199,17 +1199,11 @@ define([
          * @param {Array} images
          */
         _setImageType: function (images) {
-            var initial = this.options.mediaGalleryInitial[0].img;
-
-            if (images[0].img === initial) {
-                images = $.extend(true, [], this.options.mediaGalleryInitial);
-            } else {
-                images.map(function (img) {
-                    if (!img.type) {
-                        img.type = 'image';
-                    }
-                });
-            }
+            images.map(function (img) {
+                if (!img.type) {
+                    img.type = 'image';
+                }
+            });

             return images;
         },

