diff -Nuar a/vendor/magento/framework/Mail/Message.php b/vendor/magento/framework/Mail/Message.php
index 1519dc4f88f..d0f03be327e 100644
--- a/vendor/magento/framework/Mail/Message.php
+++ b/vendor/magento/framework/Mail/Message.php
@@ -168,7 +168,6 @@ class Message implements MailMessageInterface
         $part = new Part($body);
         $part->setCharset($this->zendMessage->getEncoding());
         $part->setEncoding(Mime::ENCODING_QUOTEDPRINTABLE);
-        $part->setDisposition(Mime::DISPOSITION_INLINE);
         $part->setType($messageType);
         $mimeMessage = new \Zend\Mime\Message();
         $mimeMessage->addPart($part);
diff -Nuar a/vendor/magento/framework/Mail/MimePart.php b/vendor/magento/framework/Mail/MimePart.php
index 9b9bb6eadbe..a43ed4b36e0 100644
--- a/vendor/magento/framework/Mail/MimePart.php
+++ b/vendor/magento/framework/Mail/MimePart.php
@@ -15,6 +15,9 @@ use Zend\Mime\Part as ZendMimePart;
  */
 class MimePart implements MimePartInterface
 {
+    /**
+     * UTF-8 charset
+     */
     public const CHARSET_UTF8 = 'utf-8';
 
     /**
@@ -47,7 +50,7 @@ class MimePart implements MimePartInterface
         $content,
         ?string $type = MimeInterface::TYPE_HTML,
         ?string $fileName = null,
-        ?string $disposition = MimeInterface::DISPOSITION_INLINE,
+        ?string $disposition = null,
         ?string $encoding = MimeInterface::ENCODING_QUOTED_PRINTABLE,
         ?string $description = null,
         ?array $filters = [],
