diff --git a/vendor/magento/module-paypal/Model/Payflow/Transparent.php b/vendor/magento/module-paypal/Model/Payflow/Transparent.php
index 87cb0df7b0d..19e2886a9f7 100644
--- a/vendor/magento/module-paypal/Model/Payflow/Transparent.php
+++ b/vendor/magento/module-paypal/Model/Payflow/Transparent.php
@@ -331,11 +331,9 @@ class Transparent extends Payflowpro implements TransparentInterface
         $zeroAmountAuthorizationId = $this->getZeroAmountAuthorizationId($payment);
         /** @var PaymentTokenInterface $vaultPaymentToken */
         $vaultPaymentToken = $payment->getExtensionAttributes()->getVaultPaymentToken();
-        if ($vaultPaymentToken && empty($zeroAmountAuthorizationId)) {
+        if ($vaultPaymentToken && empty($zeroAmountAuthorizationId) && empty($payment->getParentTransactionId())) {
             $payment->setAdditionalInformation(self::PNREF, $vaultPaymentToken->getGatewayToken());
-            if (!$payment->getParentTransactionId()) {
-                $payment->setParentTransactionId($vaultPaymentToken->getGatewayToken());
-            }
+            $payment->setParentTransactionId($vaultPaymentToken->getGatewayToken());
         }
         parent::capture($payment, $amount);

