diff -Nuar a/vendor/magento/module-tax/Model/Plugin/OrderSave.php b/vendor/magento/module-tax/Model/Plugin/OrderSave.php
index 38952eec02c..b46c5b51a9d 100644
--- a/vendor/magento/module-tax/Model/Plugin/OrderSave.php
+++ b/vendor/magento/module-tax/Model/Plugin/OrderSave.php
@@ -50,11 +50,14 @@ class OrderSave
     }
 
     /**
+     * Save order tax
+     *
      * @param \Magento\Sales\Api\Data\OrderInterface $order
      * @return $this
      * @SuppressWarnings(PHPMD.CyclomaticComplexity)
      * @SuppressWarnings(PHPMD.NPathComplexity)
      * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+     * phpcs:disable Generic.Metrics.NestingLevel.TooHigh
      */
     protected function saveOrderTax(\Magento\Sales\Api\Data\OrderInterface $order)
     {
@@ -176,7 +179,9 @@ class OrderSave
                                     } elseif (isset($quoteItemId['associated_item_id'])) {
                                         //This item is associated with a product item
                                         $item = $order->getItemByQuoteItemId($quoteItemId['associated_item_id']);
-                                        $associatedItemId = $item->getId();
+                                        if ($item !== null && $item->getId()) {
+                                            $associatedItemId = $item->getId();
+                                        }
                                     }
 
                                     $data = [
