diff --git a/vendor/magento/module-sales/Model/Order/Invoice/Total/Discount.php b/vendor/magento/module-sales/Model/Order/Invoice/Total/Discount.php
index ef7205b3744..f2a99a47659 100644
--- a/vendor/magento/module-sales/Model/Order/Invoice/Total/Discount.php
+++ b/vendor/magento/module-sales/Model/Order/Invoice/Total/Discount.php
@@ -71,9 +71,9 @@ class Discount extends AbstractTotal
         $invoice->setDiscountAmount(-$totalDiscountAmount);
         $invoice->setBaseDiscountAmount(-$baseTotalDiscountAmount);

-        $grandTotal = $invoice->getGrandTotal() - $totalDiscountAmount < 0.0001
+        $grandTotal = abs($invoice->getGrandTotal() - $totalDiscountAmount) < 0.0001
             ? 0 : $invoice->getGrandTotal() - $totalDiscountAmount;
-        $baseGrandTotal = $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount < 0.0001
+        $baseGrandTotal = abs($invoice->getBaseGrandTotal() - $baseTotalDiscountAmount) < 0.0001
             ? 0 : $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount;
         $invoice->setGrandTotal($grandTotal);
         $invoice->setBaseGrandTotal($baseGrandTotal);
