diff -Nuar a/vendor/magento/module-sales/Cron/CleanExpiredQuotes.php b/vendor/magento/module-sales/Cron/CleanExpiredQuotes.php
index 021e7b66cd1..a5c7f71df66 100644
--- a/vendor/magento/module-sales/Cron/CleanExpiredQuotes.php
+++ b/vendor/magento/module-sales/Cron/CleanExpiredQuotes.php
@@ -57,7 +57,6 @@ class CleanExpiredQuotes
 
             $quotes->addFieldToFilter('store_id', $storeId);
             $quotes->addFieldToFilter('updated_at', ['to' => date("Y-m-d", time() - $lifetime)]);
-            $quotes->addFieldToFilter('is_active', 0);
 
             foreach ($this->getExpireQuotesAdditionalFilterFields() as $field => $condition) {
                 $quotes->addFieldToFilter($field, $condition);
