diff --git a/vendor/magento/module-product-video/view/frontend/web/js/load-player.js b/vendor/magento/module-product-video/view/frontend/web/js/load-player.js
index aead9510434..1fd7523a237 100644
--- a/vendor/magento/module-product-video/view/frontend/web/js/load-player.js
+++ b/vendor/magento/module-product-video/view/frontend/web/js/load-player.js
@@ -247,6 +247,14 @@ define([

                 return;
             }
+
+            // if script already loaded by other library
+            if (window.YT) {
+                videoRegister.register('youtube', true);
+                $(window).trigger('youtubeapiready');
+
+                return;
+            }
             videoRegister.register('youtube');

             element = document.createElement('script');

