diff --git a/vendor/magento/module-paypal/Controller/Transparent/Response.php b/vendor/magento/module-paypal/Controller/Transparent/Response.php
index 05d11cedd03..efb38cd625c 100644
--- a/vendor/magento/module-paypal/Controller/Transparent/Response.php
+++ b/vendor/magento/module-paypal/Controller/Transparent/Response.php
@@ -18,7 +18,7 @@ use Magento\Paypal\Model\Payflow\Service\Response\Transaction;
 use Magento\Paypal\Model\Payflow\Service\Response\Validator\ResponseValidator;
 use Magento\Paypal\Model\Payflow\Transparent;
 use Magento\Sales\Api\PaymentFailuresInterface;
-use Magento\Framework\Session\Generic as Session;
+use Magento\Checkout\Model\Session;
 use Magento\Framework\App\Action\HttpPostActionInterface;

 /**
