diff --git a/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php b/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php
index d8d93167a3fe..eebf31c0e464 100644
--- a/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php
+++ b/vendor/magento/module-webapi-async/Controller/Rest/Asynchronous/InputParamsResolver.php
@@ -118,12 +118,13 @@ public function resolve()
         }
         $this->requestValidator->validate();
         $webapiResolvedParams = [];
+        $inputData = $this->getInputData();
         $route = $this->getRoute();
         $this->inputArraySizeLimitValue->set($route->getInputArraySizeLimit());
         $routeServiceClass = $route->getServiceClass();
         $routeServiceMethod = $route->getServiceMethod();

-        foreach ($this->getInputData() as $key => $singleEntityParams) {
+        foreach ($inputData as $key => $singleEntityParams) {
             $webapiResolvedParams[$key] = $this->resolveBulkItemParams(
                 $singleEntityParams,
                 $routeServiceClass,

